{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## Variant data selection and preprocessing (__ClinVar__)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load and filter ClinVar missense variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "# FTP site:              https://ftp.ncbi.nlm.nih.gov/pub/clinvar/tab_delimited/\n",
    "# Donwloaded file:       variant_summary.txt.gz\tlast modified: 2025-02-09\n",
    "\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import re\n",
    "import warnings\n",
    "from Bio.Data import IUPACData\n",
    "\n",
    "warnings.filterwarnings(\"ignore\")\n",
    "\n",
    "pd.set_option('display.max_rows', None)\n",
    "pd.set_option('display.max_columns', None)\n",
    "pd.set_option('display.expand_frame_repr', False)\n",
    "pd.set_option('display.max_colwidth', None)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Load data"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "data = pd.read_csv('../data/clinvar/variant_summary.txt', sep='\\t')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>ReferenceAllele</th>\n",
       "      <th>AlternateAllele</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>Guidelines</th>\n",
       "      <th>TestedInGTR</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>SomaticClinicalImpact</th>\n",
       "      <th>SomaticClinicalImpactLastEvaluated</th>\n",
       "      <th>ReviewStatusClinicalImpact</th>\n",
       "      <th>Oncogenicity</th>\n",
       "      <th>OncogenicityLastEvaluated</th>\n",
       "      <th>ReviewStatusOncogenicity</th>\n",
       "      <th>SCVsForAggregateGermlineClassification</th>\n",
       "      <th>SCVsForAggregateSomaticClinicalImpact</th>\n",
       "      <th>SCVsForAggregateOncogenicityClassification</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTAACTGTAAA (p.Arg27_Ile28delinsLeuLeuTer)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 25, 2024</td>\n",
       "      <td>397704705</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000012|RCV004998069</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511|MedGen:C3661900</td>\n",
       "      <td>Hereditary spastic paraplegia 48|not provided</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>NC_000007.13</td>\n",
       "      <td>7</td>\n",
       "      <td>4820844</td>\n",
       "      <td>4820847</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV001451119|SCV005622007</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>Indel</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTAACTGTAAA (p.Arg27_Ile28delinsLeuLeuTer)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 25, 2024</td>\n",
       "      <td>397704705</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000012|RCV004998069</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511|MedGen:C3661900</td>\n",
       "      <td>Hereditary spastic paraplegia 48|not provided</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7</td>\n",
       "      <td>4781213</td>\n",
       "      <td>4781216</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>3</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA215070,OMIM:613653.0001</td>\n",
       "      <td>3</td>\n",
       "      <td>2</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV001451119|SCV005622007</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000013</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511</td>\n",
       "      <td>Hereditary spastic paraplegia 48</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>NC_000007.13</td>\n",
       "      <td>7</td>\n",
       "      <td>4827361</td>\n",
       "      <td>4827374</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM:613653.0002,ClinGen:CA215072</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4827360</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV000020156</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>Deletion</td>\n",
       "      <td>NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)</td>\n",
       "      <td>9907</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>HGNC:22197</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>397704709</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000013</td>\n",
       "      <td>MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511</td>\n",
       "      <td>Hereditary spastic paraplegia 48</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000007.14</td>\n",
       "      <td>7</td>\n",
       "      <td>4787730</td>\n",
       "      <td>4787743</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>7p22.1</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>OMIM:613653.0002,ClinGen:CA215072</td>\n",
       "      <td>1</td>\n",
       "      <td>3</td>\n",
       "      <td>4787729</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV000020156</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_014630.3(ZNF592):c.3136G&gt;A (p.Gly1046Arg)</td>\n",
       "      <td>9640</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>HGNC:28986</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>150829393</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000000014</td>\n",
       "      <td>MONDO:MONDO:0033005,MedGen:C4551772,OMIM:251300,Orphanet:2065,Orphanet:83472</td>\n",
       "      <td>Galloway-Mowat syndrome 1</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh37</td>\n",
       "      <td>NC_000015.9</td>\n",
       "      <td>15</td>\n",
       "      <td>85342440</td>\n",
       "      <td>85342440</td>\n",
       "      <td>na</td>\n",
       "      <td>na</td>\n",
       "      <td>15q25.3</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>1</td>\n",
       "      <td>-</td>\n",
       "      <td>N</td>\n",
       "      <td>ClinGen:CA210674,UniProtKB:Q92610#VAR_064583,OMIM:613624.0001</td>\n",
       "      <td>1</td>\n",
       "      <td>4</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>SCV000020157</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type                                                                                   Name  GeneID GeneSymbol     HGNC_ID    ClinicalSignificance  ClinSigSimple LastEvaluated  RS# (dbSNP) nsv/esv (dbVar)               RCVaccession                                                                     PhenotypeIDS                                  PhenotypeList            Origin OriginSimple Assembly ChromosomeAccession Chromosome     Start      Stop ReferenceAllele AlternateAllele Cytogenetic                                          ReviewStatus  NumberSubmitters Guidelines TestedInGTR                                                       OtherIDs  SubmitterCategories  VariationID  PositionVCF ReferenceAlleleVCF      AlternateAlleleVCF SomaticClinicalImpact SomaticClinicalImpactLastEvaluated ReviewStatusClinicalImpact Oncogenicity OncogenicityLastEvaluated ReviewStatusOncogenicity SCVsForAggregateGermlineClassification SCVsForAggregateSomaticClinicalImpact SCVsForAggregateOncogenicityClassification\n",
       "0      15041                      Indel  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTAACTGTAAA (p.Arg27_Ile28delinsLeuLeuTer)    9907      AP5Z1  HGNC:22197              Pathogenic              1  Jun 25, 2024    397704705               -  RCV000000012|RCV004998069  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511|MedGen:C3661900  Hereditary spastic paraplegia 48|not provided  germline;unknown     germline   GRCh37        NC_000007.13          7   4820844   4820847              na              na      7p22.1  criteria provided, multiple submitters, no conflicts                 3          -           N                              ClinGen:CA215070,OMIM:613653.0001                    3            2      4820844               GGAT  TGCTGTAAACTGTAACTGTAAA                     -                                  -                          -            -                         -                        -              SCV001451119|SCV005622007                                     -                                          -\n",
       "1      15041                      Indel  NM_014855.3(AP5Z1):c.80_83delinsTGCTGTAAACTGTAACTGTAAA (p.Arg27_Ile28delinsLeuLeuTer)    9907      AP5Z1  HGNC:22197              Pathogenic              1  Jun 25, 2024    397704705               -  RCV000000012|RCV004998069  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511|MedGen:C3661900  Hereditary spastic paraplegia 48|not provided  germline;unknown     germline   GRCh38        NC_000007.14          7   4781213   4781216              na              na      7p22.1  criteria provided, multiple submitters, no conflicts                 3          -           N                              ClinGen:CA215070,OMIM:613653.0001                    3            2      4781213               GGAT  TGCTGTAAACTGTAACTGTAAA                     -                                  -                          -            -                         -                        -              SCV001451119|SCV005622007                                     -                                          -\n",
       "2      15042                   Deletion                                         NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1  HGNC:22197              Pathogenic              1  Jun 29, 2010    397704709               -               RCV000000013                  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511               Hereditary spastic paraplegia 48          germline     germline   GRCh37        NC_000007.13          7   4827361   4827374              na              na      7p22.1                        no assertion criteria provided                 1          -           N                              OMIM:613653.0002,ClinGen:CA215072                    1            3      4827360    GCTGCTGGACCTGCC                       G                     -                                  -                          -            -                         -                        -                           SCV000020156                                     -                                          -\n",
       "3      15042                   Deletion                                         NM_014855.3(AP5Z1):c.1413_1426del (p.Leu473fs)    9907      AP5Z1  HGNC:22197              Pathogenic              1  Jun 29, 2010    397704709               -               RCV000000013                  MONDO:MONDO:0013342,MedGen:C3150901,OMIM:613647,Orphanet:306511               Hereditary spastic paraplegia 48          germline     germline   GRCh38        NC_000007.14          7   4787730   4787743              na              na      7p22.1                        no assertion criteria provided                 1          -           N                              OMIM:613653.0002,ClinGen:CA215072                    1            3      4787729    GCTGCTGGACCTGCC                       G                     -                                  -                          -            -                         -                        -                           SCV000020156                                     -                                          -\n",
       "4      15043  single nucleotide variant                                           NM_014630.3(ZNF592):c.3136G>A (p.Gly1046Arg)    9640     ZNF592  HGNC:28986  Uncertain significance              0  Jun 29, 2015    150829393               -               RCV000000014     MONDO:MONDO:0033005,MedGen:C4551772,OMIM:251300,Orphanet:2065,Orphanet:83472                      Galloway-Mowat syndrome 1          germline     germline   GRCh37         NC_000015.9         15  85342440  85342440              na              na     15q25.3                        no assertion criteria provided                 1          -           N  ClinGen:CA210674,UniProtKB:Q92610#VAR_064583,OMIM:613624.0001                    1            4     85342440                  G                       A                     -                                  -                          -            -                         -                        -                           SCV000020157                                     -                                          -"
      ]
     },
     "execution_count": 86,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "38979"
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[\"GeneSymbol\"].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "6548628"
      ]
     },
     "execution_count": 88,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 #AlleleID\n",
      "2 Type\n",
      "3 Name\n",
      "4 GeneID\n",
      "5 GeneSymbol\n",
      "6 HGNC_ID\n",
      "7 ClinicalSignificance\n",
      "8 ClinSigSimple\n",
      "9 LastEvaluated\n",
      "10 RS# (dbSNP)\n",
      "11 nsv/esv (dbVar)\n",
      "12 RCVaccession\n",
      "13 PhenotypeIDS\n",
      "14 PhenotypeList\n",
      "15 Origin\n",
      "16 OriginSimple\n",
      "17 Assembly\n",
      "18 ChromosomeAccession\n",
      "19 Chromosome\n",
      "20 Start\n",
      "21 Stop\n",
      "22 ReferenceAllele\n",
      "23 AlternateAllele\n",
      "24 Cytogenetic\n",
      "25 ReviewStatus\n",
      "26 NumberSubmitters\n",
      "27 Guidelines\n",
      "28 TestedInGTR\n",
      "29 OtherIDs\n",
      "30 SubmitterCategories\n",
      "31 VariationID\n",
      "32 PositionVCF\n",
      "33 ReferenceAlleleVCF\n",
      "34 AlternateAlleleVCF\n",
      "35 SomaticClinicalImpact\n",
      "36 SomaticClinicalImpactLastEvaluated\n",
      "37 ReviewStatusClinicalImpact\n",
      "38 Oncogenicity\n",
      "39 OncogenicityLastEvaluated\n",
      "40 ReviewStatusOncogenicity\n",
      "41 SCVsForAggregateGermlineClassification\n",
      "42 SCVsForAggregateSomaticClinicalImpact\n",
      "43 SCVsForAggregateOncogenicityClassification\n"
     ]
    }
   ],
   "source": [
    "for count, col in enumerate(data.columns, start=1):\n",
    "    print(count, col)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We can highlight diverse columns that will be important for the next steps:\n",
    "- __AlleleID (1)__: Unique identifier for each allele in ClinVar.\n",
    "- __GeneSymbol (5)__: Identifies the affected gene. \n",
    "- __ClinicalSignificance (7)__: Defines pathogenicity classification (e.g., benign, pathogenic, etc).\n",
    "- __ClinSigSimple (8)__: Simplified numeric version of previous column (e.g., 0 and 1).\n",
    "- __LastEvaluated (9)__: Date of the last variant classification update.\n",
    "- __ReviewStatus (25)__: Indicates the confidence level of the variant classification.\n",
    "- __Chromosome (19)__ and __PositionVCF (32)__: Gives variant location.\n",
    "- __ReferenceAlleleVCF (33)__ and __AlternateAlleleVCF (34)__: Specifies the genetic variant."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15041</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 25, 2024</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>7</td>\n",
       "      <td>4820844</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15041</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 25, 2024</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>7</td>\n",
       "      <td>4781213</td>\n",
       "      <td>GGAT</td>\n",
       "      <td>TGCTGTAAACTGTAACTGTAAA</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15042</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>7</td>\n",
       "      <td>4827360</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15042</td>\n",
       "      <td>AP5Z1</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>Jun 29, 2010</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>7</td>\n",
       "      <td>4787729</td>\n",
       "      <td>GCTGCTGGACCTGCC</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15043</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Uncertain significance</td>\n",
       "      <td>0</td>\n",
       "      <td>Jun 29, 2015</td>\n",
       "      <td>no assertion criteria provided</td>\n",
       "      <td>15</td>\n",
       "      <td>85342440</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID GeneSymbol    ClinicalSignificance  ClinSigSimple LastEvaluated                                          ReviewStatus Chromosome  PositionVCF ReferenceAlleleVCF      AlternateAlleleVCF\n",
       "0      15041      AP5Z1              Pathogenic              1  Jun 25, 2024  criteria provided, multiple submitters, no conflicts          7      4820844               GGAT  TGCTGTAAACTGTAACTGTAAA\n",
       "1      15041      AP5Z1              Pathogenic              1  Jun 25, 2024  criteria provided, multiple submitters, no conflicts          7      4781213               GGAT  TGCTGTAAACTGTAACTGTAAA\n",
       "2      15042      AP5Z1              Pathogenic              1  Jun 29, 2010                        no assertion criteria provided          7      4827360    GCTGCTGGACCTGCC                       G\n",
       "3      15042      AP5Z1              Pathogenic              1  Jun 29, 2010                        no assertion criteria provided          7      4787729    GCTGCTGGACCTGCC                       G\n",
       "4      15043     ZNF592  Uncertain significance              0  Jun 29, 2015                        no assertion criteria provided         15     85342440                  G                       A"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data[[\"#AlleleID\", \"GeneSymbol\", \"ClinicalSignificance\", \"ClinSigSimple\", \n",
    "      \"LastEvaluated\", \"ReviewStatus\", \"Chromosome\", \"PositionVCF\", \n",
    "      \"ReferenceAlleleVCF\", \"AlternateAlleleVCF\"]].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Filter by criteria"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 1. Keep only SNVs (remove insertions, deletions, complex variants)."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Indel', 'Deletion', 'single nucleotide variant', 'Duplication',\n",
       "       'Microsatellite', 'Insertion', 'Variation', 'Complex',\n",
       "       'Translocation', 'Inversion', 'copy number gain', 'fusion',\n",
       "       'copy number loss', 'protein only', 'Tandem duplication'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data['Type'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "5944878"
      ]
     },
     "execution_count": 92,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter1 = data[data.Type =='single nucleotide variant'].reset_index(drop=True)\n",
    "len(data_filter1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### --> Parse \"Name\" column and filter variants"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "dict_keys(['Ala', 'Cys', 'Asp', 'Glu', 'Phe', 'Gly', 'His', 'Ile', 'Lys', 'Leu', 'Met', 'Asn', 'Pro', 'Gln', 'Arg', 'Ser', 'Thr', 'Val', 'Trp', 'Tyr'])\n",
      "dict_values(['A', 'C', 'D', 'E', 'F', 'G', 'H', 'I', 'K', 'L', 'M', 'N', 'P', 'Q', 'R', 'S', 'T', 'V', 'W', 'Y'])\n"
     ]
    }
   ],
   "source": [
    "# convert the 3-letter Aa codes to 1-letter codes\n",
    "three_to_one = IUPACData.protein_letters_3to1\n",
    "\n",
    "print(three_to_one.keys())\n",
    "print(three_to_one.values())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to extract the variant in 3-letter and 1-letter format\n",
    "def extract_variant(name):\n",
    "    # find protein variant descriptions in the Name column\n",
    "    match = re.search(r'p\\.([A-Za-z]+[0-9]+[A-Za-z]+)', name)    # e.g., p.Arg123His\n",
    "    if match:\n",
    "        # extract full variant string (e.g., \"Arg123His\")\n",
    "        three_letter_variant = match.group(1)\n",
    "        # convert the 3-letter variant to 1-letter variant\n",
    "        variant_3letter = three_letter_variant\n",
    "\n",
    "        # first 3 letters (original Aa)\n",
    "        three_letter_from = three_letter_variant[:3]\n",
    "        # extract the numeric position\n",
    "        position = re.search(r'\\d+', three_letter_variant).group()\n",
    "        # last 3 letters (mutated Aa)\n",
    "        three_letter_to = three_letter_variant[-3:]\n",
    "\n",
    "        # convert 3-letter to 1-letter (e.g., \"Arg123His\" to \"R123H\")\n",
    "        one_letter_from = three_to_one.get(three_letter_from.capitalize(), three_letter_from)\n",
    "        one_letter_to = three_to_one.get(three_letter_to.capitalize(), three_letter_to)\n",
    "        variant = f\"{one_letter_from}{position}{one_letter_to}\" \n",
    "\n",
    "        return variant_3letter, variant   # e.g., \"Arg123His\", \"R123H\"\n",
    "\n",
    "    else:\n",
    "        return None, None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# addition of 2 new cols\n",
    "data_filter1[['Variant (3-letter)', 'Variant']] = data_filter1['Name'].apply(\n",
    "    lambda x: pd.Series(extract_variant(x)))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>Variant (3-letter)</th>\n",
       "      <th>Variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15043</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Gly1046Arg</td>\n",
       "      <td>G1046R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15043</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Gly1046Arg</td>\n",
       "      <td>G1046R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>15044</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>Gln232Ter</td>\n",
       "      <td>Q232Ter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>15044</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>Gln232Ter</td>\n",
       "      <td>Q232Ter</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15045</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>Asn430Ser</td>\n",
       "      <td>N430S</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID GeneSymbol Variant (3-letter)  Variant\n",
       "0      15043     ZNF592         Gly1046Arg   G1046R\n",
       "1      15043     ZNF592         Gly1046Arg   G1046R\n",
       "2      15044    FOXRED1          Gln232Ter  Q232Ter\n",
       "3      15044    FOXRED1          Gln232Ter  Q232Ter\n",
       "4      15045    FOXRED1          Asn430Ser    N430S"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter1[[\"#AlleleID\", \"GeneSymbol\", \"Variant (3-letter)\", \"Variant\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# function to check if both parts of the variant are valid Aas\n",
    "def is_valid_variant(variant_3letter):\n",
    "    valid_amino_acids = set(three_to_one.keys())  # set of valid 3-letter Aas\n",
    "\n",
    "    if pd.isna(variant_3letter):\n",
    "        return False\n",
    "\n",
    "    # extract the 3-letter Aa codes (first 3 and last 3 chars)\n",
    "    three_letter_from = variant_3letter[:3]\n",
    "    three_letter_to = variant_3letter[-3:]\n",
    "    \n",
    "    # check if both are in the valid set and no \"Ter\" (stop codon)\n",
    "    if (three_letter_from in valid_amino_acids and three_letter_to in valid_amino_acids \n",
    "        and 'Ter' not in variant_3letter):\n",
    "        return True\n",
    "    else:\n",
    "        return False"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter1_1 = data_filter1[data_filter1['Variant (3-letter)'].apply(is_valid_variant)]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>Variant (3-letter)</th>\n",
       "      <th>Variant</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>15043</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Gly1046Arg</td>\n",
       "      <td>G1046R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>15043</td>\n",
       "      <td>ZNF592</td>\n",
       "      <td>Gly1046Arg</td>\n",
       "      <td>G1046R</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>15045</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>Asn430Ser</td>\n",
       "      <td>N430S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>15045</td>\n",
       "      <td>FOXRED1</td>\n",
       "      <td>Asn430Ser</td>\n",
       "      <td>N430S</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>15046</td>\n",
       "      <td>NUBPL</td>\n",
       "      <td>Gly56Arg</td>\n",
       "      <td>G56R</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID GeneSymbol Variant (3-letter) Variant\n",
       "0      15043     ZNF592         Gly1046Arg  G1046R\n",
       "1      15043     ZNF592         Gly1046Arg  G1046R\n",
       "4      15045    FOXRED1          Asn430Ser   N430S\n",
       "5      15045    FOXRED1          Asn430Ser   N430S\n",
       "6      15046      NUBPL           Gly56Arg    G56R"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter1_1[[\"#AlleleID\", \"GeneSymbol\", \"Variant (3-letter)\", \"Variant\"]].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 100,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3426329"
      ]
     },
     "execution_count": 100,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# invalid or ambiguous variants were removed\n",
    "len(data_filter1_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 2. Retain submissions from 2021 and later."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LastEvaluated</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Jun 29, 2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Jun 29, 2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Jun 06, 2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Jun 06, 2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Jul 05, 2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  LastEvaluated\n",
       "0  Jun 29, 2015\n",
       "1  Jun 29, 2015\n",
       "4  Jun 06, 2024\n",
       "5  Jun 06, 2024\n",
       "6  Jul 05, 2022"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter1_1[['LastEvaluated']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "metadata": {},
   "outputs": [],
   "source": [
    "# convert LastEvaluated column to datetime format\n",
    "data_filter1_1['LastEvaluated'] = pd.to_datetime(data_filter1_1['LastEvaluated'], errors='coerce')\n",
    "\n",
    "# new column extracting the year and converting it to integer\n",
    "data_filter1_1['LastEvaluated (Year)'] = data_filter1_1['LastEvaluated'].dt.year.astype('Int64')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 103,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>LastEvaluated (Year)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>2015-06-29</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2015-06-29</td>\n",
       "      <td>2015</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2024-06-06</td>\n",
       "      <td>2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2024-06-06</td>\n",
       "      <td>2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2022-07-05</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  LastEvaluated  LastEvaluated (Year)\n",
       "0    2015-06-29                  2015\n",
       "1    2015-06-29                  2015\n",
       "4    2024-06-06                  2024\n",
       "5    2024-06-06                  2024\n",
       "6    2022-07-05                  2022"
      ]
     },
     "execution_count": 103,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter1_1[['LastEvaluated', 'LastEvaluated (Year)']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 104,
   "metadata": {},
   "outputs": [],
   "source": [
    "# filter variants where year is 2021 or later\n",
    "data_filter2 = data_filter1_1[data_filter1_1['LastEvaluated (Year)'] >= 2021]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>LastEvaluated</th>\n",
       "      <th>LastEvaluated (Year)</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2024-06-06</td>\n",
       "      <td>2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>2024-06-06</td>\n",
       "      <td>2024</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2022-07-05</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>2022-07-05</td>\n",
       "      <td>2022</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>2024-11-01</td>\n",
       "      <td>2024</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  LastEvaluated  LastEvaluated (Year)\n",
       "4    2024-06-06                  2024\n",
       "5    2024-06-06                  2024\n",
       "6    2022-07-05                  2022\n",
       "7    2022-07-05                  2022\n",
       "8    2024-11-01                  2024"
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter2[['LastEvaluated', 'LastEvaluated (Year)']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "3112595"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 107,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LastEvaluated (Year)\n",
       "1965          2\n",
       "1973          2\n",
       "1976          2\n",
       "1977          2\n",
       "1979          2\n",
       "1980          4\n",
       "1981          4\n",
       "1982          2\n",
       "1983          8\n",
       "1984         10\n",
       "1985          4\n",
       "1986         10\n",
       "1987         14\n",
       "1988         16\n",
       "1989         60\n",
       "1990         80\n",
       "1991        114\n",
       "1992        192\n",
       "1993        176\n",
       "1994        176\n",
       "1995        240\n",
       "1996        166\n",
       "1997        226\n",
       "1998        282\n",
       "1999        294\n",
       "2000        314\n",
       "2001        494\n",
       "2002        393\n",
       "2003        374\n",
       "2004        384\n",
       "2005        288\n",
       "2006        392\n",
       "2007        442\n",
       "2008        424\n",
       "2009        494\n",
       "2010        540\n",
       "2011        984\n",
       "2012       1620\n",
       "2013       4998\n",
       "2014       4900\n",
       "2015       7295\n",
       "2016      16063\n",
       "2017      28332\n",
       "2018      61288\n",
       "2019      76755\n",
       "2020      71901\n",
       "2021     344526\n",
       "2022     766916\n",
       "2023     963595\n",
       "2024    1034214\n",
       "2025       3344\n",
       "dtype: int64"
      ]
     },
     "execution_count": 107,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter1_1.groupby('LastEvaluated (Year)').size()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 108,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "LastEvaluated (Year)\n",
       "2021     344526\n",
       "2022     766916\n",
       "2023     963595\n",
       "2024    1034214\n",
       "2025       3344\n",
       "dtype: int64"
      ]
     },
     "execution_count": 108,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter2.groupby('LastEvaluated (Year)').size()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 3. Exclude variants with zero-star review status, VUS and conflicting classifications.\n",
    "\n",
    "##### --> Remove VUS and conflicting classifications"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 109,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Likely pathogenic',\n",
       "       'Conflicting classifications of pathogenicity',\n",
       "       'Pathogenic/Pathogenic, low penetrance; other; risk factor',\n",
       "       'Pathogenic/Likely pathogenic/Pathogenic, low penetrance; other',\n",
       "       'Uncertain significance', 'Pathogenic/Likely pathogenic',\n",
       "       'Pathogenic', 'Likely benign', 'Benign', 'Benign/Likely benign',\n",
       "       'Conflicting classifications of pathogenicity; risk factor',\n",
       "       'drug response', 'Benign; drug response',\n",
       "       'Conflicting classifications of pathogenicity; association; risk factor',\n",
       "       'Benign/Likely benign; other',\n",
       "       'Conflicting classifications of pathogenicity; other',\n",
       "       'Pathogenic/Likely pathogenic; risk factor',\n",
       "       'no classifications from unflagged records', 'not provided',\n",
       "       'Likely benign; other', 'Benign; other', 'Pathogenic; risk factor',\n",
       "       'Conflicting classifications of pathogenicity; association',\n",
       "       'Benign/Likely benign; association', 'Benign; risk factor',\n",
       "       'Benign/Likely benign; other; risk factor',\n",
       "       'Pathogenic; drug response',\n",
       "       'Pathogenic/Likely pathogenic/Pathogenic, low penetrance',\n",
       "       'Uncertain significance/Uncertain risk allele',\n",
       "       'Conflicting classifications of pathogenicity; protective',\n",
       "       'Uncertain significance; risk factor', 'Likely pathogenic; other',\n",
       "       'Likely pathogenic; association',\n",
       "       'Uncertain significance; drug response',\n",
       "       'Likely pathogenic; drug response',\n",
       "       'Conflicting classifications of pathogenicity; drug response',\n",
       "       'Pathogenic/Likely risk allele',\n",
       "       'Pathogenic/Likely pathogenic/Likely risk allele',\n",
       "       'Conflicting classifications of pathogenicity; other; risk factor',\n",
       "       'Pathogenic; other', 'other', 'Uncertain significance; other',\n",
       "       'Benign; association',\n",
       "       'Pathogenic/Pathogenic, low penetrance; other',\n",
       "       'Pathogenic/Likely pathogenic; other',\n",
       "       'Likely benign; Affects; association',\n",
       "       'Likely pathogenic/Likely risk allele', 'Likely risk allele',\n",
       "       'Uncertain risk allele; risk factor',\n",
       "       'Pathogenic/Likely pathogenic; association',\n",
       "       'Likely benign; risk factor', 'risk factor',\n",
       "       'Likely pathogenic; Affects', 'Pathogenic; association',\n",
       "       'Benign; Affects', 'Affects', 'Likely pathogenic; risk factor',\n",
       "       'Likely benign; association', 'Likely benign; drug response',\n",
       "       'Uncertain significance; association', 'association', 'protective',\n",
       "       'Likely pathogenic, low penetrance', 'Pathogenic; Affects'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 109,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter2.ClinicalSignificance.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 110,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter3 = data_filter2[data_filter2['ClinicalSignificance'].isin(['Pathogenic','Likely pathogenic',\n",
    "                                                                       'Pathogenic/Likely pathogenic', \n",
    "                                                                       'Benign', 'Likely benign', \n",
    "                                                                       'Benign/Likely benign'])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 111,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Likely pathogenic', 'Pathogenic/Likely pathogenic', 'Pathogenic',\n",
       "       'Likely benign', 'Benign', 'Benign/Likely benign'], dtype=object)"
      ]
     },
     "execution_count": 111,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3.ClinicalSignificance.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 112,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "277997"
      ]
     },
     "execution_count": 112,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter3)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### --> Filter out variants with zero-star review status"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 113,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['-', 'criteria provided, multiple submitters',\n",
       "       'no assertion criteria provided',\n",
       "       'criteria provided, single submitter'], dtype=object)"
      ]
     },
     "execution_count": 113,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# somatic classification\n",
    "data_filter3.ReviewStatusClinicalImpact.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 114,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['criteria provided, single submitter',\n",
       "       'criteria provided, multiple submitters, no conflicts',\n",
       "       'reviewed by expert panel', 'no assertion criteria provided'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 114,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# germline classification\n",
    "data_filter3.ReviewStatus.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 115,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ReviewStatus\n",
       "criteria provided, single submitter                     176317\n",
       "criteria provided, multiple submitters, no conflicts     86823\n",
       "no assertion criteria provided                           10115\n",
       "reviewed by expert panel                                  4742\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 115,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3.ReviewStatus.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 116,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter3_1 = data_filter3[data_filter3['ReviewStatus'] != 'no assertion criteria provided'].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 117,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "ReviewStatus\n",
       "criteria provided, single submitter                     176317\n",
       "criteria provided, multiple submitters, no conflicts     86823\n",
       "reviewed by expert panel                                  4742\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 117,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_1.ReviewStatus.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 118,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "267882"
      ]
     },
     "execution_count": 118,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter3_1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### --> Parse Chromosome column"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 119,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['11', '6', '10', '16', '22', '15', '7', '1', '8', '21', '5', '19',\n",
       "       '4', '3', '17', '12', '20', '9', '18', '2', '14', '13', 'MT', 'Y',\n",
       "       'X', 'na', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16,\n",
       "       17, 18, 19, 20, 21, 22, 'Un'], dtype=object)"
      ]
     },
     "execution_count": 119,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_1.Chromosome.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 120,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter3_2=data_filter3_1[~data_filter3_1.Chromosome.isin(['na','Un'])].reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 121,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "267735"
      ]
     },
     "execution_count": 121,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# after removing invalid or missing chromosome data\n",
    "len(data_filter3_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 122,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['11', '6', '10', '16', '22', '15', '7', '1', '8', '21', '5', '19',\n",
       "       '4', '3', '17', '12', '20', '9', '18', '2', '14', '13', 'MT', 'Y',\n",
       "       'X', 1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11, 12, 13, 14, 15, 16, 17, 18,\n",
       "       19, 20, 21, 22], dtype=object)"
      ]
     },
     "execution_count": 122,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_2.Chromosome.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 123,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "25"
      ]
     },
     "execution_count": 123,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_2['Chromosome'] = data_filter3_2['Chromosome'].astype(str)\n",
    "data_filter3_2['Chromosome'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 124,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['11', '6', '10', '16', '22', '15', '7', '1', '8', '21', '5', '19',\n",
       "       '4', '3', '17', '12', '20', '9', '18', '2', '14', '13', 'MT', 'Y',\n",
       "       'X'], dtype=object)"
      ]
     },
     "execution_count": 124,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_2['Chromosome'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now the 'Chromosome' column contains the expected chromosome identifiers:\n",
    "- autosomes (1-22)\n",
    "- sex chromosomes (X, Y)\n",
    "- mitochondrial DNA (MT)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 4. Keep variants with AF<0.01 in the gnomAD v.2.1.\n",
    "\n",
    "At this point, we need to pass the data through VEP in order to get gnomAD information.\n",
    "\n",
    "But first, the __assembly issue__ must be taken into account."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 125,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GRCh37', 'GRCh38'], dtype=object)"
      ]
     },
     "execution_count": 125,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_2.Assembly.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 126,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assembly\n",
       "GRCh37    133870\n",
       "GRCh38    133865\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 126,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter3_2 = data_filter3_2.sort_values(by='Assembly', ascending=False).reset_index(drop=True)\n",
    "data_filter3_2.Assembly.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Here, we notice that several variants have the same Allele ID and the only difference is the Assembly. \n",
    "\n",
    "Thus, we make sure this is the case and then, drop duplicates."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 127,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Function to identify Allele ID groups with inconsistent data\n",
    "def check_duplicates(df):\n",
    "    cols_to_check=['#AlleleID', 'Type', 'Name', 'GeneID', 'GeneSymbol', 'HGNC_ID',\n",
    "       'ClinicalSignificance', 'ClinSigSimple', 'LastEvaluated', 'RS# (dbSNP)',\n",
    "       'nsv/esv (dbVar)', 'RCVaccession', 'PhenotypeIDS', 'PhenotypeList',\n",
    "       'Origin', 'OriginSimple', 'Cytogenetic', 'ReviewStatus', \n",
    "       'NumberSubmitters', 'Guidelines', 'TestedInGTR', 'OtherIDs', \n",
    "       'SubmitterCategories', 'VariationID', 'Variant (3-letter)', 'Variant']\n",
    "\n",
    "    grouped = df.groupby('#AlleleID')\n",
    "    \n",
    "    # find groups where any column (except Assembly) has different values\n",
    "    different_values = {}\n",
    "    for name, group in grouped:\n",
    "        if len(group) > 1:  # only check groups with multiple variants\n",
    "            for col in cols_to_check:\n",
    "                unique_values = group[col].nunique()\n",
    "                if unique_values > 1:\n",
    "                    if name not in different_values:\n",
    "                        different_values[name] = []\n",
    "                    different_values[name].append(col)\n",
    "\n",
    "    if different_values:\n",
    "        print(\"Found AlleleIDs with different values in columns other than Assembly:\")\n",
    "        for allele_id, columns in different_values.items():\n",
    "            print(f\"\\nAlleleID {allele_id} has different values in columns: {columns}\")\n",
    "    else:\n",
    "        print(\"All rows with the same #AlleleID have identical values (except possibly Assembly)\")\n",
    "\n",
    "    return df['#AlleleID'].duplicated().sum()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 128,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "All rows with the same #AlleleID have identical values (except possibly Assembly)\n"
     ]
    }
   ],
   "source": [
    "total_duplicates= check_duplicates(data_filter3_2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 129,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assembly\n",
       "GRCh38    133845\n",
       "GRCh37        10\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 129,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4 = data_filter3_2.drop_duplicates(subset=['#AlleleID'],keep='first')\n",
    "data_filter4.Assembly.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Next, we must separate data by Assembly before using VEP. \n",
    "\n",
    "This is done because coordinates info should __not__ be mixed."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 130,
   "metadata": {},
   "outputs": [],
   "source": [
    "for assembly in ['GRCh37', 'GRCh38']:\n",
    "    file = data_filter4[data_filter4['Assembly'] == assembly].copy()\n",
    "    file.to_csv(f'../data/clinvar/clinvar_data_preVEP_{assembly.lower()}.csv', index=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "VEP input must be in VCF format, thus the following function."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "def create_vcf(df, outputfile):\n",
    "    # necessary columns for VCF format\n",
    "    vcf_columns = [\n",
    "        'Chromosome',\n",
    "        'PositionVCF',\n",
    "        'RS# (dbSNP)',\n",
    "        'ReferenceAlleleVCF',\n",
    "        'AlternateAlleleVCF'\n",
    "    ]\n",
    "\n",
    "    vcf_df = df[vcf_columns].copy()\n",
    "    vcf_df['Chromosome'] = vcf_df['Chromosome'].astype(str)\n",
    "    vcf_df['Chromosome'] = vcf_df['Chromosome'].str.replace('chr', '', case=False)\n",
    "\n",
    "    chrom_order = ([str(i) for i in range(1, 23)] + ['X', 'Y', 'MT'])\n",
    "\n",
    "    vcf_df['Chromosome'] = pd.Categorical(vcf_df['Chromosome'], categories=chrom_order, ordered=True)\n",
    "    vcf_df['PositionVCF'] = pd.to_numeric(vcf_df['PositionVCF'])\n",
    "    vcf_df = vcf_df.sort_values(['Chromosome', 'PositionVCF'])\n",
    "    \n",
    "    # add remaining required VCF columns\n",
    "    vcf_df['ID'] = vcf_df['RS# (dbSNP)'].fillna('.')\n",
    "    vcf_df['QUAL'] = '.'  \n",
    "    vcf_df['FILTER'] = '.'  \n",
    "    vcf_df['INFO'] = '.'  \n",
    "    \n",
    "    # reorder and rename columns to match VCF format\n",
    "    vcf_df = vcf_df[['Chromosome', 'PositionVCF', 'ID', 'ReferenceAlleleVCF', 'AlternateAlleleVCF', 'QUAL', 'FILTER', 'INFO']]\n",
    "    vcf_df.columns = ['#CHROM', 'POS', 'ID', 'REF', 'ALT', 'QUAL', 'FILTER', 'INFO']\n",
    "\n",
    "    vcf_df.to_csv(outputfile, sep='\\t', index=False)\n",
    "    print(f\"VCF file created at: {outputfile}\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After running VEP for the third filter criteria (gnomAD), we should rearrange the output and continue cleaning the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "def parse_output(file_path):\n",
    "    with open(file_path, 'r') as file:\n",
    "        for i, line in enumerate(file):\n",
    "            if line.startswith(\"#Uploaded_variation\"):\n",
    "                header_line = i\n",
    "                break\n",
    "\n",
    "    df = pd.read_csv(file_path, delimiter='\\t', skiprows=header_line,low_memory=False)\n",
    "    #print(df.columns)\n",
    "    print('initial file length:', len(df))\n",
    "\n",
    "    # rename the first column (Uploaded_variation)\n",
    "    df.columns = df.columns.str.replace('#', '')\n",
    "    \n",
    "    df2= df[df.Protein_position != '-'].reset_index(drop=True).copy()\n",
    "\n",
    "    df_filtered = df2[df2['Amino_acids'].str.contains('/')]\n",
    "    df_filtered = df_filtered[~df_filtered['Amino_acids'].str.contains(r'\\*')]\n",
    "\n",
    "    # create Chromosome, PositionVCF, variant columns before merging with the original clinvar df\n",
    "    df_filtered[['Chromosome', 'PositionVCF_dashed']] = df_filtered['Location'].str.split(':', expand=True)\n",
    "    df_filtered['PositionVCF'] = df_filtered.apply(lambda x: x['PositionVCF_dashed'].split('-')[0], axis=1)\n",
    "    \n",
    "    df_filtered['PositionVCF'] = df_filtered['PositionVCF'].astype(int)\n",
    "    df_filtered['Variant'] = df_filtered.apply(lambda x: x['Amino_acids'].split('/')[0] + str(x['Protein_position']) + x['Amino_acids'].split('/')[1], axis=1)\n",
    "    \n",
    "    # the Allele column represents the alternate allele for a variant. \n",
    "    # the Reference allele is not explicitly listed here because VCF files generally only list \n",
    "        # the alternate allele (the variation from the reference genome). \n",
    "    # the Reference allele would typically be implied based on the genomic position and reference sequence.\n",
    "    \n",
    "    df_filtered['AlternateAlleleVCF'] = df_filtered['Allele']\n",
    "\n",
    "    df_filtered2 = df_filtered[df_filtered['Consequence'].str.contains('missense_variant', na=False)].copy()\n",
    "    \n",
    "    # duplicated lines, due to canonical + isoform. we drop duplicates based on \n",
    "    df_filtered3 = df_filtered2.drop_duplicates(subset=['Chromosome','PositionVCF','Allele','Gene','Feature_type','CDS_position','Protein_position','Variant'], keep='first')\n",
    "    df_filtered3['GeneSymbol'] = df_filtered3['Extra'].str.extract(r'SYMBOL=([^;]+)')\n",
    "    df_filtered3['HGNC_ID'] = df_filtered3['Extra'].str.extract(r'HGNC_ID=([^;]+)')\n",
    "    \n",
    "    print('final file length:',len(df_filtered3))\n",
    "    \n",
    "    return df_filtered3\n",
    "\n",
    "\n",
    "def merge_original_and_vepout(df1,df2):\n",
    "    df1['Chromosome'] = df1['Chromosome'].astype(str)\n",
    "    df2['Chromosome'] = df2['Chromosome'].astype(str)\n",
    "    df1=df1.reset_index(drop=True)\n",
    "    df2=df2.reset_index(drop=True)\n",
    "    merged_df = df1.merge(df2, on =['Chromosome','PositionVCF','AlternateAlleleVCF','Variant','GeneSymbol','HGNC_ID'], how='left')\n",
    "    return merged_df    "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "First, for GRCh38:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF file created at: ../data/clinvar/clinvar_data_inputVEP_grch38.vcf\n"
     ]
    }
   ],
   "source": [
    "data_grch38 = pd.read_csv('../data/clinvar/clinvar_data_preVEP_grch38.csv')\n",
    "create_vcf(data_grch38, '../data/clinvar/clinvar_data_inputVEP_grch38.vcf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To run VEP, the following command was executed:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 134,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#  ./vep -i clinvar_data_inputVEP_grch38.vcf -o clinvar_data_outputVEP_grch38.txt --offline \\\n",
    "#      --assembly GRCh38 \\\n",
    "#      --symbol --transcript_version --ccds --protein --uniprot \\\n",
    "#      --hgvs --fasta Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz \\\n",
    "#      --af --af_1kg --af_gnomade --af_gnomadg --max_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial file length: 1613017\n",
      "final file length: 339455\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_grch38 = \"../data/clinvar/clinvar_data_outputVEP_grch38.txt\"\n",
    "df = parse_output(output_grch38)\n",
    "\n",
    "data_filter4_1_grch38 = merge_original_and_vepout(data_grch38, df)\n",
    "\n",
    "# same rows, but more columns after merging\n",
    "len(data_filter4_1_grch38) == len(data_grch38)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['missense_variant', nan, 'missense_variant,splice_region_variant',\n",
       "       'missense_variant,NMD_transcript_variant',\n",
       "       'missense_variant,splice_region_variant,NMD_transcript_variant'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_1_grch38.Consequence.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>gnomADg_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.001133</td>\n",
       "      <td>0.0009133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gnomADe_AF gnomADg_AF\n",
       "0   0.001133  0.0009133\n",
       "1        NaN        NaN\n",
       "2        NaN        NaN\n",
       "3        NaN        NaN\n",
       "4        NaN        NaN"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_1_grch38['gnomADe_AF'] = data_filter4_1_grch38['Extra'].str.extract(r'gnomADe_AF=([^;]+)')\n",
    "data_filter4_1_grch38['gnomADg_AF'] = data_filter4_1_grch38['Extra'].str.extract(r'gnomADg_AF=([^;]+)')\n",
    "\n",
    "data_filter4_1_grch38[['gnomADe_AF','gnomADg_AF']].head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "101504"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter4_1_grch38[data_filter4_1_grch38.gnomADe_AF.notna()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Then, for GRCh37:"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF file created at: ../data/clinvar/clinvar_data_inputVEP_grch37.vcf\n"
     ]
    }
   ],
   "source": [
    "data_grch37 = pd.read_csv('../data/clinvar/clinvar_data_preVEP_grch37.csv')\n",
    "create_vcf(data_grch37, '../data/clinvar/clinvar_data_inputVEP_grch37.vcf')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "To run VEP, the following command was executed:\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#  ./vep -i clinvar_data_inputVEP_grch37.vcf -o clinvar_data_outputVEP_grch37.txt --offline \\\n",
    "#      --assembly GRCh37 \\\n",
    "#      --symbol --transcript_version --ccds --protein --uniprot \\\n",
    "#      --hgvs --fasta Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz \\\n",
    "#      --af --af_1kg --af_gnomade --af_gnomadg --max_af"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "initial file length: 20\n",
      "final file length: 12\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "output_grch37 = \"../data/clinvar/clinvar_data_outputVEP_grch37.txt\"\n",
    "df = parse_output(output_grch37)\n",
    "df['HGNC_ID'] = df['HGNC_ID'].apply(lambda x: 'HGNC:' + str(x) if pd.notna(x) else x)\n",
    "\n",
    "data_filter4_1_grch37 = merge_original_and_vepout(data_grch37, df)\n",
    "\n",
    "# same rows, but more columns after merging\n",
    "len(data_filter4_1_grch37) == len(data_grch37)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['missense_variant'], dtype=object)"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_1_grch37.Consequence.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "10"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_1_grch37['gnomADe_AF'] = data_filter4_1_grch37['Extra'].str.extract(r'gnomADe_AF=([^;]+)')\n",
    "\n",
    "len(data_filter4_1_grch37[data_filter4_1_grch37.gnomADe_AF.notna()])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Finally, we can merge again GRCh37 and GRCh38 files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "48"
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# cols to mantain, from original dataset\n",
    "cols = ['#AlleleID', 'Type', 'Name', 'GeneID', 'GeneSymbol', 'HGNC_ID', \n",
    "        'ClinicalSignificance', 'ClinSigSimple', 'RS# (dbSNP)', 'nsv/esv (dbVar)', \n",
    "        'RCVaccession', 'PhenotypeIDS', 'PhenotypeList', 'Origin', 'OriginSimple', \n",
    "        'Assembly', 'ChromosomeAccession', 'Chromosome', 'Start', 'Stop', 'Cytogenetic', \n",
    "        'ReviewStatus', 'NumberSubmitters', 'OtherIDs', 'SubmitterCategories', \n",
    "        'VariationID', 'PositionVCF', 'ReferenceAlleleVCF', 'AlternateAlleleVCF', \n",
    "        'Variant (3-letter)', 'Variant', 'LastEvaluated (Year)']\n",
    "\n",
    "# cols to add, after obtaining gnomAD info\n",
    "cols_to_take = ['Uploaded_variation', 'Location', 'Allele', 'Gene', 'Feature', \n",
    "                'Feature_type', 'Consequence', 'cDNA_position', 'CDS_position', \n",
    "                'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation', \n",
    "                'Extra', 'PositionVCF_dashed', 'gnomADe_AF']\n",
    "\n",
    "for col in cols_to_take:\n",
    "    cols.append(col)\n",
    "\n",
    "len(cols)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>Variant (3-letter)</th>\n",
       "      <th>Variant</th>\n",
       "      <th>LastEvaluated (Year)</th>\n",
       "      <th>Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "      <th>PositionVCF_dashed</th>\n",
       "      <th>gnomADe_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1809239</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001386125.1(OBSCN):c.496G&gt;A (p.Ala166Thr)</td>\n",
       "      <td>84033</td>\n",
       "      <td>OBSCN</td>\n",
       "      <td>HGNC:15719</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV003418466|RCV004050307</td>\n",
       "      <td>MedGen:C3661900|MedGen:CN169374</td>\n",
       "      <td>not provided|not specified</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>228212279</td>\n",
       "      <td>228212279</td>\n",
       "      <td>1q42.13</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>2</td>\n",
       "      <td>ClinGen:CA1431438</td>\n",
       "      <td>2</td>\n",
       "      <td>1744457</td>\n",
       "      <td>228212279</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Ala166Thr</td>\n",
       "      <td>A166T</td>\n",
       "      <td>2023</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1:228212279</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000154358</td>\n",
       "      <td>ENST00000284548.16</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>637</td>\n",
       "      <td>496</td>\n",
       "      <td>166</td>\n",
       "      <td>A/T</td>\n",
       "      <td>Gca/Aca</td>\n",
       "      <td>rs555146765</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=OBSCN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:15719;CCDS=CCDS1570.2;ENSP=ENSP00000284548;SWISSPROT=Q5VST9.202;UNIPARC=UPI0000425971;UNIPROT_ISOFORM=Q5VST9-3;HGVSc=ENST00000284548.16:c.496G&gt;A;HGVSp=ENSP00000284548.11:p.Ala166Thr;AF=0.0004;AFR_AF=0.0008;AMR_AF=0.0014;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.001133;gnomADe_AFR_AF=0.0002016;gnomADe_AMR_AF=0.0004112;gnomADe_ASJ_AF=9.352e-05;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0.0004933;gnomADe_MID_AF=0.0005405;gnomADe_NFE_AF=0.001353;gnomADe_REMAINING_AF=0.0005889;gnomADe_SAS_AF=0;gnomADg_AF=0.0009133;gnomADg_AFR_AF=0.0003389;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.001063;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0.0001037;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.00156;gnomADg_REMAINING_AF=0.0004812;gnomADg_SAS_AF=0;MAX_AF=0.00156;MAX_AF_POPS=gnomADg_NFE;CLIN_SIG=likely_benign,benign;PHENO=1</td>\n",
       "      <td>228212279</td>\n",
       "      <td>0.001133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>2059503</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001205293.3(CACNA1E):c.3965C&gt;T (p.Ser1322Phe)</td>\n",
       "      <td>777</td>\n",
       "      <td>CACNA1E</td>\n",
       "      <td>HGNC:1392</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV002806411</td>\n",
       "      <td>MedGen:C3661900</td>\n",
       "      <td>not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>181755373</td>\n",
       "      <td>181755373</td>\n",
       "      <td>1q25.3</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>ClinGen:CA343623382</td>\n",
       "      <td>2</td>\n",
       "      <td>1993841</td>\n",
       "      <td>181755373</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Ser1322Phe</td>\n",
       "      <td>S1322F</td>\n",
       "      <td>2022</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1:181755373</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000198216</td>\n",
       "      <td>ENST00000367570.6</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>4856</td>\n",
       "      <td>3965</td>\n",
       "      <td>1322</td>\n",
       "      <td>S/F</td>\n",
       "      <td>tCc/tTc</td>\n",
       "      <td>COSV62407041</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=CACNA1E;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:1392;CCDS=CCDS53443.1;ENSP=ENSP00000356542;SWISSPROT=Q15878.209;UNIPARC=UPI000044D37D;UNIPROT_ISOFORM=Q15878-3;HGVSc=ENST00000367570.6:c.3965C&gt;T;HGVSp=ENSP00000356542.1:p.Ser1322Phe;SOMATIC=1;PHENO=1</td>\n",
       "      <td>181755373</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>2058909</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_002397.5(MEF2C):c.439A&gt;G (p.Ile147Val)</td>\n",
       "      <td>4208</td>\n",
       "      <td>MEF2C</td>\n",
       "      <td>HGNC:6996</td>\n",
       "      <td>Benign</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV002828380</td>\n",
       "      <td>MONDO:MONDO:0013266,MedGen:C3150700,OMIM:613443,Orphanet:228384,Orphanet:664410</td>\n",
       "      <td>Intellectual disability, autosomal dominant 20</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000005.10</td>\n",
       "      <td>5</td>\n",
       "      <td>88752007</td>\n",
       "      <td>88752007</td>\n",
       "      <td>5q14.3</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>ClinGen:CA360423982</td>\n",
       "      <td>2</td>\n",
       "      <td>2004843</td>\n",
       "      <td>88752007</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Ile147Val</td>\n",
       "      <td>I147V</td>\n",
       "      <td>2022</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>5:88752007</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000081189</td>\n",
       "      <td>ENST00000437473.6</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1090</td>\n",
       "      <td>439</td>\n",
       "      <td>147</td>\n",
       "      <td>I/V</td>\n",
       "      <td>Atc/Gtc</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=MEF2C;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:6996;CCDS=CCDS47245.1;ENSP=ENSP00000396219;SWISSPROT=Q06413.210;UNIPARC=UPI0000040635;UNIPROT_ISOFORM=Q06413-1;HGVSc=ENST00000437473.6:c.439A&gt;G;HGVSp=ENSP00000396219.2:p.Ile147Val</td>\n",
       "      <td>88752007</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>2058962</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_152296.5(ATP1A3):c.281T&gt;C (p.Leu94Pro)</td>\n",
       "      <td>478</td>\n",
       "      <td>ATP1A3</td>\n",
       "      <td>HGNC:801</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV002795956</td>\n",
       "      <td>MONDO:MONDO:0007496,MedGen:C1868681,OMIM:128235,Orphanet:71517</td>\n",
       "      <td>Dystonia 12</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000019.10</td>\n",
       "      <td>19</td>\n",
       "      <td>41988012</td>\n",
       "      <td>41988012</td>\n",
       "      <td>19q13.2</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>ClinGen:CA406056371</td>\n",
       "      <td>2</td>\n",
       "      <td>1992440</td>\n",
       "      <td>41988012</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Leu94Pro</td>\n",
       "      <td>L94P</td>\n",
       "      <td>2022</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>19:41988012</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000105409</td>\n",
       "      <td>ENST00000441343.5</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>419</td>\n",
       "      <td>281</td>\n",
       "      <td>94</td>\n",
       "      <td>L/P</td>\n",
       "      <td>cTg/cCg</td>\n",
       "      <td>-</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=ATP1A3;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:801;ENSP=ENSP00000411503;TREMBL=A0A0A0MT26.64;UNIPARC=UPI0000E5A1FE;HGVSc=ENST00000441343.5:c.281T&gt;C;HGVSp=ENSP00000411503.1:p.Leu94Pro</td>\n",
       "      <td>41988012</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>2059150</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001100.4(ACTA1):c.794A&gt;G (p.Gln265Arg)</td>\n",
       "      <td>58</td>\n",
       "      <td>ACTA1</td>\n",
       "      <td>HGNC:129</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV002796062</td>\n",
       "      <td>MONDO:MONDO:0008070,MedGen:C3711389,OMIM:161800,Orphanet:98904</td>\n",
       "      <td>Actin accumulation myopathy</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000001.11</td>\n",
       "      <td>1</td>\n",
       "      <td>229432008</td>\n",
       "      <td>229432008</td>\n",
       "      <td>1q42.13</td>\n",
       "      <td>criteria provided, single submitter</td>\n",
       "      <td>1</td>\n",
       "      <td>ClinGen:CA345146455</td>\n",
       "      <td>2</td>\n",
       "      <td>1992629</td>\n",
       "      <td>229432008</td>\n",
       "      <td>T</td>\n",
       "      <td>C</td>\n",
       "      <td>Gln265Arg</td>\n",
       "      <td>Q265R</td>\n",
       "      <td>2023</td>\n",
       "      <td>-1.0</td>\n",
       "      <td>1:229432008</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000143632</td>\n",
       "      <td>ENST00000366683.4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>907</td>\n",
       "      <td>794</td>\n",
       "      <td>265</td>\n",
       "      <td>Q/R</td>\n",
       "      <td>cAg/cGg</td>\n",
       "      <td>CM992127</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=ACTA1;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:129;ENSP=ENSP00000355644;TREMBL=A6NL76.108;UNIPARC=UPI000C755200;HGVSc=ENST00000366683.4:c.794A&gt;G;HGVSp=ENSP00000355644.4:p.Gln265Arg;SOMATIC=1;PHENO=1</td>\n",
       "      <td>229432008</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type                                              Name  GeneID GeneSymbol     HGNC_ID  ClinicalSignificance  ClinSigSimple  RS# (dbSNP) nsv/esv (dbVar)               RCVaccession                                                                     PhenotypeIDS                                   PhenotypeList    Origin OriginSimple Assembly ChromosomeAccession Chromosome      Start       Stop Cytogenetic                                          ReviewStatus  NumberSubmitters             OtherIDs  SubmitterCategories  VariationID  PositionVCF ReferenceAlleleVCF AlternateAlleleVCF Variant (3-letter) Variant  LastEvaluated (Year)  Uploaded_variation     Location Allele             Gene             Feature Feature_type       Consequence cDNA_position CDS_position Protein_position Amino_acids   Codons Existing_variation                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                          Extra PositionVCF_dashed gnomADe_AF\n",
       "0    1809239  single nucleotide variant      NM_001386125.1(OBSCN):c.496G>A (p.Ala166Thr)   84033      OBSCN  HGNC:15719  Benign/Likely benign              0           -1               -  RCV003418466|RCV004050307                                                  MedGen:C3661900|MedGen:CN169374                      not provided|not specified  germline     germline   GRCh38        NC_000001.11          1  228212279  228212279     1q42.13  criteria provided, multiple submitters, no conflicts                 2    ClinGen:CA1431438                    2      1744457    228212279                  G                  A          Ala166Thr   A166T                  2023                -1.0  1:228212279      A  ENSG00000154358  ENST00000284548.16   Transcript  missense_variant           637          496              166         A/T  Gca/Aca        rs555146765  IMPACT=MODERATE;STRAND=1;SYMBOL=OBSCN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:15719;CCDS=CCDS1570.2;ENSP=ENSP00000284548;SWISSPROT=Q5VST9.202;UNIPARC=UPI0000425971;UNIPROT_ISOFORM=Q5VST9-3;HGVSc=ENST00000284548.16:c.496G>A;HGVSp=ENSP00000284548.11:p.Ala166Thr;AF=0.0004;AFR_AF=0.0008;AMR_AF=0.0014;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.001133;gnomADe_AFR_AF=0.0002016;gnomADe_AMR_AF=0.0004112;gnomADe_ASJ_AF=9.352e-05;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0.0004933;gnomADe_MID_AF=0.0005405;gnomADe_NFE_AF=0.001353;gnomADe_REMAINING_AF=0.0005889;gnomADe_SAS_AF=0;gnomADg_AF=0.0009133;gnomADg_AFR_AF=0.0003389;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.001063;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0.0001037;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.00156;gnomADg_REMAINING_AF=0.0004812;gnomADg_SAS_AF=0;MAX_AF=0.00156;MAX_AF_POPS=gnomADg_NFE;CLIN_SIG=likely_benign,benign;PHENO=1          228212279   0.001133\n",
       "1    2059503  single nucleotide variant  NM_001205293.3(CACNA1E):c.3965C>T (p.Ser1322Phe)     777    CACNA1E   HGNC:1392         Likely benign              0           -1               -               RCV002806411                                                                  MedGen:C3661900                                    not provided  germline     germline   GRCh38        NC_000001.11          1  181755373  181755373      1q25.3                   criteria provided, single submitter                 1  ClinGen:CA343623382                    2      1993841    181755373                  C                  T         Ser1322Phe  S1322F                  2022                -1.0  1:181755373      T  ENSG00000198216   ENST00000367570.6   Transcript  missense_variant          4856         3965             1322         S/F  tCc/tTc       COSV62407041                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                IMPACT=MODERATE;STRAND=1;SYMBOL=CACNA1E;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:1392;CCDS=CCDS53443.1;ENSP=ENSP00000356542;SWISSPROT=Q15878.209;UNIPARC=UPI000044D37D;UNIPROT_ISOFORM=Q15878-3;HGVSc=ENST00000367570.6:c.3965C>T;HGVSp=ENSP00000356542.1:p.Ser1322Phe;SOMATIC=1;PHENO=1          181755373        NaN\n",
       "2    2058909  single nucleotide variant         NM_002397.5(MEF2C):c.439A>G (p.Ile147Val)    4208      MEF2C   HGNC:6996                Benign              0           -1               -               RCV002828380  MONDO:MONDO:0013266,MedGen:C3150700,OMIM:613443,Orphanet:228384,Orphanet:664410  Intellectual disability, autosomal dominant 20  germline     germline   GRCh38        NC_000005.10          5   88752007   88752007      5q14.3                   criteria provided, single submitter                 1  ClinGen:CA360423982                    2      2004843     88752007                  T                  C          Ile147Val   I147V                  2022                -1.0   5:88752007      C  ENSG00000081189   ENST00000437473.6   Transcript  missense_variant          1090          439              147         I/V  Atc/Gtc                  -                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     IMPACT=MODERATE;STRAND=-1;SYMBOL=MEF2C;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:6996;CCDS=CCDS47245.1;ENSP=ENSP00000396219;SWISSPROT=Q06413.210;UNIPARC=UPI0000040635;UNIPROT_ISOFORM=Q06413-1;HGVSc=ENST00000437473.6:c.439A>G;HGVSp=ENSP00000396219.2:p.Ile147Val           88752007        NaN\n",
       "3    2058962  single nucleotide variant         NM_152296.5(ATP1A3):c.281T>C (p.Leu94Pro)     478     ATP1A3    HGNC:801            Pathogenic              1           -1               -               RCV002795956                   MONDO:MONDO:0007496,MedGen:C1868681,OMIM:128235,Orphanet:71517                                     Dystonia 12  germline     germline   GRCh38        NC_000019.10         19   41988012   41988012     19q13.2                   criteria provided, single submitter                 1  ClinGen:CA406056371                    2      1992440     41988012                  A                  G           Leu94Pro    L94P                  2022                -1.0  19:41988012      G  ENSG00000105409   ENST00000441343.5   Transcript  missense_variant           419          281               94         L/P  cTg/cCg                  -                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                IMPACT=MODERATE;STRAND=-1;SYMBOL=ATP1A3;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:801;ENSP=ENSP00000411503;TREMBL=A0A0A0MT26.64;UNIPARC=UPI0000E5A1FE;HGVSc=ENST00000441343.5:c.281T>C;HGVSp=ENSP00000411503.1:p.Leu94Pro           41988012        NaN\n",
       "4    2059150  single nucleotide variant         NM_001100.4(ACTA1):c.794A>G (p.Gln265Arg)      58      ACTA1    HGNC:129            Pathogenic              1           -1               -               RCV002796062                   MONDO:MONDO:0008070,MedGen:C3711389,OMIM:161800,Orphanet:98904                     Actin accumulation myopathy  germline     germline   GRCh38        NC_000001.11          1  229432008  229432008     1q42.13                   criteria provided, single submitter                 1  ClinGen:CA345146455                    2      1992629    229432008                  T                  C          Gln265Arg   Q265R                  2023                -1.0  1:229432008      C  ENSG00000143632   ENST00000366683.4   Transcript  missense_variant           907          794              265         Q/R  cAg/cGg           CM992127                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                 IMPACT=MODERATE;STRAND=-1;SYMBOL=ACTA1;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:129;ENSP=ENSP00000355644;TREMBL=A6NL76.108;UNIPARC=UPI000C755200;HGVSc=ENST00000366683.4:c.794A>G;HGVSp=ENSP00000355644.4:p.Gln265Arg;SOMATIC=1;PHENO=1          229432008        NaN"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_2 = pd.concat([data_filter4_1_grch38,data_filter4_1_grch37])\n",
    "data_filter4_2 = data_filter4_2[cols]\n",
    "len(data_filter4_2)\n",
    "data_filter4_2.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IMPACT=MODERATE;STRAND=1;SYMBOL=OBSCN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:15719;CCDS=CCDS1570.2;ENSP=ENSP00000284548;SWISSPROT=Q5VST9.202;UNIPARC=UPI0000425971;UNIPROT_ISOFORM=Q5VST9-3;HGVSc=ENST00000284548.16:c.496G>A;HGVSp=ENSP00000284548.11:p.Ala166Thr;AF=0.0004;AFR_AF=0.0008;AMR_AF=0.0014;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.001133;gnomADe_AFR_AF=0.0002016;gnomADe_AMR_AF=0.0004112;gnomADe_ASJ_AF=9.352e-05;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0.0004933;gnomADe_MID_AF=0.0005405;gnomADe_NFE_AF=0.001353;gnomADe_REMAINING_AF=0.0005889;gnomADe_SAS_AF=0;gnomADg_AF=0.0009133;gnomADg_AFR_AF=0.0003389;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.001063;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0.0001037;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.00156;gnomADg_REMAINING_AF=0.0004812;gnomADg_SAS_AF=0;MAX_AF=0.00156;MAX_AF_POPS=gnomADg_NFE;CLIN_SIG=likely_benign,benign;PHENO=1\n"
     ]
    }
   ],
   "source": [
    "print(data_filter4_2['Extra'].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "1619"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter4_2[data_filter4_2.Consequence.isna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Consequence\n",
       "missense_variant                                                 127328\n",
       "missense_variant,splice_region_variant                             3742\n",
       "missense_variant,NMD_transcript_variant                            1120\n",
       "missense_variant,splice_region_variant,NMD_transcript_variant        46\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_2.Consequence.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# data_filter4 is the cleaned dataset before splitting by Assembly\n",
    "# data_filter4_2 is the cleaned dataset after splitting, running VEP and merging again\n",
    "len(data_filter4_2)==len(data_filter4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GRCh38', 'GRCh37'], dtype=object)"
      ]
     },
     "execution_count": 99,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_2.Assembly.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>gnomADg_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>0.001133</td>\n",
       "      <td>0.0009133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  gnomADe_AF gnomADg_AF\n",
       "0   0.001133  0.0009133\n",
       "1        NaN        NaN\n",
       "2        NaN        NaN\n",
       "3        NaN        NaN\n",
       "4        NaN        NaN"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_2['gnomADe_AF'] = data_filter4_2['Extra'].str.extract(r'gnomADe_AF=([^;]+)')\n",
    "data_filter4_2['gnomADg_AF'] = data_filter4_2['Extra'].str.extract(r'gnomADg_AF=([^;]+)')\n",
    "\n",
    "data_filter4_2[['gnomADe_AF','gnomADg_AF']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The columns 'gnomADe_AF' and 'gnomADg_AF' represent the AFs of the variant \n",
    "in the gnomAD Exomes (gnomADe_AF) and gnomAD Genomes (gnomADg_AF) datasets, respectively. \n",
    "\n",
    "The new column 'gnomAD_AF' combines these frequencies, prioritizing Exomes (gnomADe_AF) \n",
    "and using Genomes (gnomADg_AF) when Exomes data is missing. \n",
    "\n",
    "The final filtering step selects rare variants with a combined gnomAD_AF < 0.01.\n",
    "\n",
    "With this we make sure we have a frequency value for every variant"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter4_2['gnomADe_AF'] = pd.to_numeric(data_filter4_2['gnomADe_AF'], errors='coerce')\n",
    "data_filter4_2['gnomADg_AF'] = pd.to_numeric(data_filter4_2['gnomADg_AF'], errors='coerce')\n",
    "\n",
    "# new column 'gnomAD_AF' that fills missing exomes AF with genomes AF\n",
    "data_filter4_2['gnomAD_AF'] = data_filter4_2['gnomADe_AF'].fillna(data_filter4_2['gnomADg_AF'])\n",
    "\n",
    "# filter for variants where gnomAD_AF < 0.01 and drop variants where gnomAD_AF is NaN\n",
    "data_filter4_3 = data_filter4_2[data_filter4_2['gnomAD_AF'] < 0.01].dropna(subset=['gnomAD_AF'])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "96686\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>...</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "      <th>PositionVCF_dashed</th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>gnomADg_AF</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>1809239</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_001386125.1(OBSCN):c.496G&gt;A (p.Ala166Thr)</td>\n",
       "      <td>84033</td>\n",
       "      <td>OBSCN</td>\n",
       "      <td>HGNC:15719</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>496</td>\n",
       "      <td>166</td>\n",
       "      <td>A/T</td>\n",
       "      <td>Gca/Aca</td>\n",
       "      <td>rs555146765</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=OBSCN;SYMBOL_S...</td>\n",
       "      <td>228212279</td>\n",
       "      <td>0.001133</td>\n",
       "      <td>0.000913</td>\n",
       "      <td>0.001133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>2059238</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_006734.4(HIVEP2):c.1529C&gt;T (p.Ser510Leu)</td>\n",
       "      <td>3097</td>\n",
       "      <td>HIVEP2</td>\n",
       "      <td>HGNC:4921</td>\n",
       "      <td>Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>1529</td>\n",
       "      <td>510</td>\n",
       "      <td>S/L</td>\n",
       "      <td>tCa/tTa</td>\n",
       "      <td>rs531761193</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=HIVEP2;SYMBOL...</td>\n",
       "      <td>142773210</td>\n",
       "      <td>0.000031</td>\n",
       "      <td>0.000013</td>\n",
       "      <td>0.000031</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2 rows  50 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type  \\\n",
       "0    1809239  single nucleotide variant   \n",
       "6    2059238  single nucleotide variant   \n",
       "\n",
       "                                           Name  GeneID GeneSymbol  \\\n",
       "0  NM_001386125.1(OBSCN):c.496G>A (p.Ala166Thr)   84033      OBSCN   \n",
       "6   NM_006734.4(HIVEP2):c.1529C>T (p.Ser510Leu)    3097     HIVEP2   \n",
       "\n",
       "      HGNC_ID  ClinicalSignificance  ClinSigSimple  RS# (dbSNP)  \\\n",
       "0  HGNC:15719  Benign/Likely benign              0           -1   \n",
       "6   HGNC:4921         Likely benign              0           -1   \n",
       "\n",
       "  nsv/esv (dbVar)  ... CDS_position Protein_position Amino_acids   Codons  \\\n",
       "0               -  ...          496              166         A/T  Gca/Aca   \n",
       "6               -  ...         1529              510         S/L  tCa/tTa   \n",
       "\n",
       "  Existing_variation                                              Extra  \\\n",
       "0        rs555146765  IMPACT=MODERATE;STRAND=1;SYMBOL=OBSCN;SYMBOL_S...   \n",
       "6        rs531761193  IMPACT=MODERATE;STRAND=-1;SYMBOL=HIVEP2;SYMBOL...   \n",
       "\n",
       "  PositionVCF_dashed gnomADe_AF  gnomADg_AF  gnomAD_AF  \n",
       "0          228212279   0.001133    0.000913   0.001133  \n",
       "6          142773210   0.000031    0.000013   0.000031  \n",
       "\n",
       "[2 rows x 50 columns]"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(len(data_filter4_3))\n",
    "data_filter4_3.head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "96686"
      ]
     },
     "execution_count": 69,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter4_3[data_filter4_3.gnomAD_AF.notna()])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "IMPACT=MODERATE;STRAND=1;SYMBOL=OBSCN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:15719;CCDS=CCDS1570.2;ENSP=ENSP00000284548;SWISSPROT=Q5VST9.202;UNIPARC=UPI0000425971;UNIPROT_ISOFORM=Q5VST9-3;HGVSc=ENST00000284548.16:c.496G>A;HGVSp=ENSP00000284548.11:p.Ala166Thr;AF=0.0004;AFR_AF=0.0008;AMR_AF=0.0014;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.001133;gnomADe_AFR_AF=0.0002016;gnomADe_AMR_AF=0.0004112;gnomADe_ASJ_AF=9.352e-05;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0.0004933;gnomADe_MID_AF=0.0005405;gnomADe_NFE_AF=0.001353;gnomADe_REMAINING_AF=0.0005889;gnomADe_SAS_AF=0;gnomADg_AF=0.0009133;gnomADg_AFR_AF=0.0003389;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.001063;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0.0001037;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.00156;gnomADg_REMAINING_AF=0.0004812;gnomADg_SAS_AF=0;MAX_AF=0.00156;MAX_AF_POPS=gnomADg_NFE;CLIN_SIG=likely_benign,benign;PHENO=1\n"
     ]
    }
   ],
   "source": [
    "print(data_filter4_3[\"Extra\"].iloc[0])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "96686\n",
      "96544\n"
     ]
    }
   ],
   "source": [
    "data_filter4_3 = data_filter4_3.sort_values(by='LastEvaluated (Year)', ascending=False)\n",
    "data_filter4_3['HGVSp'] = data_filter4_3['Extra'].str.extract(r'HGVSp=([^;]+)')\n",
    "\n",
    "cols_check = ['Type', 'GeneID','Gene', 'GeneSymbol', 'Feature','HGNC_ID',\n",
    "      'Assembly','ChromosomeAccession', 'Chromosome', \n",
    "      'HGVSp','Protein_position', 'Amino_acids','Variant']\n",
    "\n",
    "data_filter4_4 = data_filter4_3.drop_duplicates(subset=cols_check, keep='first')\n",
    "\n",
    "print(len(data_filter4_3))\n",
    "print(len(data_filter4_4))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### 5. Keep genes with at least one pathogenic variant of any type."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 72,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['Pathogenic/Likely pathogenic', 'Likely pathogenic', 'Pathogenic',\n",
       "       'Likely benign', 'Benign', 'Benign/Likely benign'], dtype=object)"
      ]
     },
     "execution_count": 72,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter4_3.ClinicalSignificance.unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 73,
   "metadata": {},
   "outputs": [],
   "source": [
    "def filter_pathogenic_genes(df):\n",
    "    # pathogenic terms\n",
    "    pathogenic_terms = ['Pathogenic', 'Likely pathogenic', 'Pathogenic/Likely pathogenic']\n",
    "    \n",
    "    # filter for genes that have at least one pathogenic variant\n",
    "    pathogenic_genes = df[df['ClinicalSignificance'].isin(pathogenic_terms)]['GeneSymbol'].unique()\n",
    "    \n",
    "    # filter df to retain only variants for those genes\n",
    "    df_filtered = df[df['GeneSymbol'].isin(pathogenic_genes)]\n",
    "    \n",
    "    return df_filtered"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "data_filter5 = filter_pathogenic_genes(data_filter4_4)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 75,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49187"
      ]
     },
     "execution_count": 75,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(data_filter5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice that only GRCh38 remains. All the pathogenic variants in the filtered dataset are from this assembly."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 106,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Assembly\n",
       "GRCh38    49187\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 106,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "data_filter5[\"Assembly\"].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Some parsing final steps..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 76,
   "metadata": {},
   "outputs": [],
   "source": [
    "def classify_significance(clinical_significance):\n",
    "    if any(term in clinical_significance for term in ['Pathogenic', 'Likely pathogenic']):\n",
    "        return 'P'\n",
    "    elif any(term in clinical_significance for term in ['Benign', 'Likely benign']):\n",
    "        return 'B'\n",
    "    return 'Other'"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 77,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_ClinVar_dataset = data_filter5.copy()\n",
    "cleaned_ClinVar_dataset['BinaryClinicalSignificance'] = data_filter5['ClinicalSignificance'].apply(classify_significance)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BinaryClinicalSignificance\n",
       "B    34405\n",
       "P    14782\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 78,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_ClinVar_dataset.BinaryClinicalSignificance.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2156"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cleaned_ClinVar_dataset.GeneSymbol.unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49187"
      ]
     },
     "execution_count": 93,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(cleaned_ClinVar_dataset)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>...</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "      <th>PositionVCF_dashed</th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>gnomADg_AF</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>BinaryClinicalSignificance</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>119639</th>\n",
       "      <td>29901</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_182894.3(VSX2):c.679C&gt;T (p.Arg227Trp)</td>\n",
       "      <td>338917</td>\n",
       "      <td>VSX2</td>\n",
       "      <td>HGNC:1975</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>121912545</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>R/W</td>\n",
       "      <td>Cgg/Tgg</td>\n",
       "      <td>rs121912545,CM042327</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=VSX2;SYMBOL_SO...</td>\n",
       "      <td>74259701</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>0.000020</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSP00000261980.2:p.Arg227Trp</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>93641</th>\n",
       "      <td>104480</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000180.4(GUCY2D):c.307G&gt;A (p.Glu103Lys)</td>\n",
       "      <td>3000</td>\n",
       "      <td>GUCY2D</td>\n",
       "      <td>HGNC:4689</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>61749668</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>E/K</td>\n",
       "      <td>Gag/Aag</td>\n",
       "      <td>rs61749668,CM077936</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=GUCY2D;SYMBOL_...</td>\n",
       "      <td>8003354</td>\n",
       "      <td>0.000191</td>\n",
       "      <td>0.000033</td>\n",
       "      <td>0.000191</td>\n",
       "      <td>ENSP00000254854.4:p.Glu103Lys</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59576</th>\n",
       "      <td>3734350</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000441.2(SLC26A4):c.1335G&gt;C (p.Leu445Phe)</td>\n",
       "      <td>5172</td>\n",
       "      <td>SLC26A4</td>\n",
       "      <td>HGNC:8818</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>L/F</td>\n",
       "      <td>ttG/ttC</td>\n",
       "      <td>rs1355468475</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=SLC26A4;SYMBOL...</td>\n",
       "      <td>107694474</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>ENSP00000494017.1:p.Leu445Phe</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59575</th>\n",
       "      <td>3734348</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000441.2(SLC26A4):c.1279T&gt;C (p.Ser427Pro)</td>\n",
       "      <td>5172</td>\n",
       "      <td>SLC26A4</td>\n",
       "      <td>HGNC:8818</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>S/P</td>\n",
       "      <td>Tct/Cct</td>\n",
       "      <td>rs758015694</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=SLC26A4;SYMBOL...</td>\n",
       "      <td>107694418</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000002</td>\n",
       "      <td>ENSP00000494017.1:p.Ser427Pro</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>59572</th>\n",
       "      <td>3734343</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_000441.2(SLC26A4):c.1207G&gt;T (p.Ala403Ser)</td>\n",
       "      <td>5172</td>\n",
       "      <td>SLC26A4</td>\n",
       "      <td>HGNC:8818</td>\n",
       "      <td>Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>-1</td>\n",
       "      <td>-</td>\n",
       "      <td>...</td>\n",
       "      <td>A/S</td>\n",
       "      <td>Gcc/Tcc</td>\n",
       "      <td>rs1791527351,COSV107219136</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=SLC26A4;SYMBOL...</td>\n",
       "      <td>107690181</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000001</td>\n",
       "      <td>ENSP00000494017.1:p.Ala403Ser</td>\n",
       "      <td>P</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows  52 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "        #AlleleID                       Type  \\\n",
       "119639      29901  single nucleotide variant   \n",
       "93641      104480  single nucleotide variant   \n",
       "59576     3734350  single nucleotide variant   \n",
       "59575     3734348  single nucleotide variant   \n",
       "59572     3734343  single nucleotide variant   \n",
       "\n",
       "                                                Name  GeneID GeneSymbol  \\\n",
       "119639      NM_182894.3(VSX2):c.679C>T (p.Arg227Trp)  338917       VSX2   \n",
       "93641     NM_000180.4(GUCY2D):c.307G>A (p.Glu103Lys)    3000     GUCY2D   \n",
       "59576   NM_000441.2(SLC26A4):c.1335G>C (p.Leu445Phe)    5172    SLC26A4   \n",
       "59575   NM_000441.2(SLC26A4):c.1279T>C (p.Ser427Pro)    5172    SLC26A4   \n",
       "59572   NM_000441.2(SLC26A4):c.1207G>T (p.Ala403Ser)    5172    SLC26A4   \n",
       "\n",
       "          HGNC_ID          ClinicalSignificance  ClinSigSimple  RS# (dbSNP)  \\\n",
       "119639  HGNC:1975  Pathogenic/Likely pathogenic              1    121912545   \n",
       "93641   HGNC:4689             Likely pathogenic              1     61749668   \n",
       "59576   HGNC:8818             Likely pathogenic              1           -1   \n",
       "59575   HGNC:8818                    Pathogenic              1           -1   \n",
       "59572   HGNC:8818             Likely pathogenic              1           -1   \n",
       "\n",
       "       nsv/esv (dbVar)  ... Amino_acids   Codons          Existing_variation  \\\n",
       "119639               -  ...         R/W  Cgg/Tgg        rs121912545,CM042327   \n",
       "93641                -  ...         E/K  Gag/Aag         rs61749668,CM077936   \n",
       "59576                -  ...         L/F  ttG/ttC                rs1355468475   \n",
       "59575                -  ...         S/P  Tct/Cct                 rs758015694   \n",
       "59572                -  ...         A/S  Gcc/Tcc  rs1791527351,COSV107219136   \n",
       "\n",
       "                                                    Extra PositionVCF_dashed  \\\n",
       "119639  IMPACT=MODERATE;STRAND=1;SYMBOL=VSX2;SYMBOL_SO...           74259701   \n",
       "93641   IMPACT=MODERATE;STRAND=1;SYMBOL=GUCY2D;SYMBOL_...            8003354   \n",
       "59576   IMPACT=MODERATE;STRAND=1;SYMBOL=SLC26A4;SYMBOL...          107694474   \n",
       "59575   IMPACT=MODERATE;STRAND=1;SYMBOL=SLC26A4;SYMBOL...          107694418   \n",
       "59572   IMPACT=MODERATE;STRAND=1;SYMBOL=SLC26A4;SYMBOL...          107690181   \n",
       "\n",
       "       gnomADe_AF gnomADg_AF gnomAD_AF                          HGVSp  \\\n",
       "119639   0.000003   0.000020  0.000003  ENSP00000261980.2:p.Arg227Trp   \n",
       "93641    0.000191   0.000033  0.000191  ENSP00000254854.4:p.Glu103Lys   \n",
       "59576         NaN   0.000007  0.000007  ENSP00000494017.1:p.Leu445Phe   \n",
       "59575    0.000002        NaN  0.000002  ENSP00000494017.1:p.Ser427Pro   \n",
       "59572    0.000001        NaN  0.000001  ENSP00000494017.1:p.Ala403Ser   \n",
       "\n",
       "        BinaryClinicalSignificance  \n",
       "119639                           P  \n",
       "93641                            P  \n",
       "59576                            P  \n",
       "59575                            P  \n",
       "59572                            P  \n",
       "\n",
       "[5 rows x 52 columns]"
      ]
     },
     "execution_count": 81,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_ClinVar_dataset.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### And some statistics"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "def count_p_and_b_per_gene(df):\n",
    "    # group by 'GeneSymbol' and 'BinaryClinicalSignificance', then count\n",
    "    counts = df.groupby(['GeneSymbol', 'BinaryClinicalSignificance']).size().unstack(fill_value=0)\n",
    "\n",
    "    if 'B' in counts.columns and 'P' in counts.columns:\n",
    "        counts.columns = ['B_count', 'P_count']\n",
    "    elif 'P' in counts.columns:\n",
    "        counts = counts.rename(columns={'P': 'P_count'})\n",
    "    elif 'B' in counts.columns:\n",
    "        counts = counts.rename(columns={'B': 'B_count'})\n",
    "    \n",
    "    return counts"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>B_count</th>\n",
       "      <th>P_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAS</th>\n",
       "      <td>2</td>\n",
       "      <td>8</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS1</th>\n",
       "      <td>21</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AARS2</th>\n",
       "      <td>19</td>\n",
       "      <td>6</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AASS</th>\n",
       "      <td>9</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABAT</th>\n",
       "      <td>2</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMYND10</th>\n",
       "      <td>9</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZMYND11</th>\n",
       "      <td>12</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF341</th>\n",
       "      <td>21</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZNF408</th>\n",
       "      <td>7</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ZP1</th>\n",
       "      <td>4</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>2156 rows  2 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "            B_count  P_count\n",
       "GeneSymbol                  \n",
       "AAAS              2        8\n",
       "AARS1            21        4\n",
       "AARS2            19        6\n",
       "AASS              9        1\n",
       "ABAT              2        3\n",
       "...             ...      ...\n",
       "ZMYND10           9        4\n",
       "ZMYND11          12        2\n",
       "ZNF341           21        1\n",
       "ZNF408            7        3\n",
       "ZP1               4        1\n",
       "\n",
       "[2156 rows x 2 columns]"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_per_gene = count_p_and_b_per_gene(cleaned_ClinVar_dataset)\n",
    "counts_per_gene"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>B_count</th>\n",
       "      <th>P_count</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>USH2A</th>\n",
       "      <td>90</td>\n",
       "      <td>133</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>TP53</th>\n",
       "      <td>65</td>\n",
       "      <td>86</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A3</th>\n",
       "      <td>36</td>\n",
       "      <td>83</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>PKHD1</th>\n",
       "      <td>38</td>\n",
       "      <td>79</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL7A1</th>\n",
       "      <td>237</td>\n",
       "      <td>76</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>FBN1</th>\n",
       "      <td>48</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A4</th>\n",
       "      <td>47</td>\n",
       "      <td>68</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>RYR1</th>\n",
       "      <td>47</td>\n",
       "      <td>62</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL4A5</th>\n",
       "      <td>113</td>\n",
       "      <td>60</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>SCN1A</th>\n",
       "      <td>44</td>\n",
       "      <td>56</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NOTCH3</th>\n",
       "      <td>77</td>\n",
       "      <td>50</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>DYSF</th>\n",
       "      <td>69</td>\n",
       "      <td>48</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCB11</th>\n",
       "      <td>31</td>\n",
       "      <td>42</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>ABCD1</th>\n",
       "      <td>61</td>\n",
       "      <td>36</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>MYO15A</th>\n",
       "      <td>66</td>\n",
       "      <td>35</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>BRCA2</th>\n",
       "      <td>64</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CDH23</th>\n",
       "      <td>32</td>\n",
       "      <td>34</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>NPHS1</th>\n",
       "      <td>38</td>\n",
       "      <td>33</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL2A1</th>\n",
       "      <td>105</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>CPS1</th>\n",
       "      <td>46</td>\n",
       "      <td>31</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>COL1A1</th>\n",
       "      <td>78</td>\n",
       "      <td>30</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            B_count  P_count\n",
       "GeneSymbol                  \n",
       "USH2A            90      133\n",
       "TP53             65       86\n",
       "COL4A3           36       83\n",
       "PKHD1            38       79\n",
       "COL7A1          237       76\n",
       "FBN1             48       68\n",
       "COL4A4           47       68\n",
       "RYR1             47       62\n",
       "COL4A5          113       60\n",
       "SCN1A            44       56\n",
       "NOTCH3           77       50\n",
       "DYSF             69       48\n",
       "ABCB11           31       42\n",
       "ABCD1            61       36\n",
       "MYO15A           66       35\n",
       "BRCA2            64       34\n",
       "CDH23            32       34\n",
       "NPHS1            38       33\n",
       "COL2A1          105       31\n",
       "CPS1             46       31\n",
       "COL1A1           78       30"
      ]
     },
     "execution_count": 84,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "counts_per_gene[(counts_per_gene.P_count>=30) & (counts_per_gene.B_count>= 30)].sort_values(by=['P_count','B_count'], ascending=False)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "metadata": {},
   "outputs": [],
   "source": [
    "cleaned_ClinVar_dataset['PositionVCF'] = pd.to_numeric(cleaned_ClinVar_dataset['PositionVCF'])\n",
    "\n",
    "# sort by chromosome and position\n",
    "cleaned_ClinVar_dataset = cleaned_ClinVar_dataset.sort_values(['Chromosome', 'PositionVCF'])\n",
    "cleaned_ClinVar_dataset = cleaned_ClinVar_dataset.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GRCh38'], dtype=object)"
      ]
     },
     "execution_count": 96,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_ClinVar_dataset[\"Assembly\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "metadata": {},
   "outputs": [],
   "source": [
    "# this works as database (with the applied filters) from where to retrieve variants for the variants_pipeline.sh\n",
    "cleaned_ClinVar_dataset.to_csv('../data/clinvar/cleaned_ClinVar_dataset.csv', index=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Name</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>NM_198576.4(AGRN):c.11G&gt;C (p.Arg4Pro)</td>\n",
       "      <td>1</td>\n",
       "      <td>G</td>\n",
       "      <td>1020183</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>NM_198576.4(AGRN):c.125A&gt;C (p.Glu42Ala)</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>1020297</td>\n",
       "      <td>C</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>NM_198576.4(AGRN):c.494C&gt;T (p.Pro165Leu)</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1035307</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>NM_198576.4(AGRN):c.773C&gt;T (p.Thr258Ile)</td>\n",
       "      <td>1</td>\n",
       "      <td>C</td>\n",
       "      <td>1041218</td>\n",
       "      <td>T</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>NM_198576.4(AGRN):c.1058A&gt;G (p.Gln353Arg)</td>\n",
       "      <td>1</td>\n",
       "      <td>A</td>\n",
       "      <td>1041583</td>\n",
       "      <td>G</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        Name Chromosome ReferenceAlleleVCF  \\\n",
       "0      NM_198576.4(AGRN):c.11G>C (p.Arg4Pro)          1                  G   \n",
       "1    NM_198576.4(AGRN):c.125A>C (p.Glu42Ala)          1                  A   \n",
       "2   NM_198576.4(AGRN):c.494C>T (p.Pro165Leu)          1                  C   \n",
       "3   NM_198576.4(AGRN):c.773C>T (p.Thr258Ile)          1                  C   \n",
       "4  NM_198576.4(AGRN):c.1058A>G (p.Gln353Arg)          1                  A   \n",
       "\n",
       "   PositionVCF AlternateAlleleVCF  \n",
       "0      1020183                  C  \n",
       "1      1020297                  C  \n",
       "2      1035307                  T  \n",
       "3      1041218                  T  \n",
       "4      1041583                  G  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_ClinVar_dataset[['Name','Chromosome','ReferenceAlleleVCF','PositionVCF','AlternateAlleleVCF']].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Now that the dataset is cleaned, we convert it into VCF format for input to the VEP tool, so we can obtain predictions from the chosen pathogenicity predictors."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 88,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "VCF file created at: ../data/clinvar/cleaned_Clinvar_dataset_inputVEP.vcf\n"
     ]
    }
   ],
   "source": [
    "create_vcf(cleaned_ClinVar_dataset[['Chromosome', 'PositionVCF', 'RS# (dbSNP)', \n",
    "                                    'ReferenceAlleleVCF', 'AlternateAlleleVCF']],\n",
    "                                    \"../data/clinvar/cleaned_Clinvar_dataset_inputVEP.vcf\")"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Command run in the script"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#    /home/aitanadiaz/ensembl-vep/./vep -i \"$input_vcf\" -o \"$output_file\" --offline \\\n",
    "#        --assembly $assembly \\\n",
    "#        --symbol --transcript_version --ccds --protein --uniprot --canonical \\\n",
    "#        --hgvs --fasta /home/aitanadiaz/ensembl-vep/Homo_sapiens.GRCh38.dna.primary_assembly.fa.gz \\\n",
    "#        --af --af_1kg --af_gnomade --af_gnomadg --max_af \\\n",
    "#        --sift b --polyphen b \\\n",
    "#        --plugin AlphaMissense,file=/home/aitanadiaz/ensembl-vep/plugins/AlphaMissense_${assembly}.tsv.gz \\\n",
    "#        --plugin Blosum62 \\\n",
    "#        --plugin CADD,snv=/home/aitanadiaz/ensembl-vep/plugins/whole_genome_SNVs_${assembly}.tsv.gz \\\n",
    "#        --plugin ClinPred,file=/home/aitanadiaz/ensembl-vep/plugins/ClinPred_${assembly}_tabbed.tsv.gz \\\n",
    "#        --plugin dbNSFP,/home/aitanadiaz/ensembl-vep/plugins/dbNSFP5.1a.grch38.gz,VEST4_score,VEST4_rankscore,BayesDel_addAF_pred,BayesDel_addAF_score \\\n",
    "#        --plugin EVE,file=/home/aitanadiaz/ensembl-vep/plugins/EVE/eve_merged.vcf.gz \\\n",
    "#        --plugin PrimateAI,/home/aitanadiaz/ensembl-vep/plugins/PrimateAI_scores_v0.2_${assembly}_sorted.tsv.bgz \\\n",
    "#        --plugin REVEL,file=/home/aitanadiaz/ensembl-vep/plugins/new_tabbed_revel_${assembly}.tsv.gz\n",
    "#        #--plugin BayesDel,file=/home/aitanadiaz/ensembl-vep/plugins/BayesDel_170824_addAF/BayesDel_170824_addAF_all_scores.txt.gz \\\n",
    "#        #--plugin VARITY,file=/home/aitanadiaz/ensembl-vep/plugins/varity_all_predictions.tsv.gz"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Read VEP output (predictions added!)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Notice we are obtaining predictions only for the following predictors:\n",
    "- __CADD__\n",
    "- __EVE__\n",
    "- __AlphaMissense__\n",
    "- __BayesDel__\n",
    "- __REVEL__\n",
    "- __VEST4__\n",
    "\n",
    "Missing ones will be run and add after parsing VEP results."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "VEP_output = pd.read_csv(\n",
    "    '../data/clinvar/cleaned_Clinvar_dataset_outputVEP.txt', \n",
    "    sep='\\t', \n",
    "    comment='#', \n",
    "    header=None\n",
    ")\n",
    "\n",
    "# actual header\n",
    "with open('../data/clinvar/cleaned_Clinvar_dataset_outputVEP.txt') as f:\n",
    "    for line in f:\n",
    "        if line.startswith(\"#Uploaded_variation\"):  \n",
    "            columns = line.strip(\"#\").strip().split(\"\\t\")\n",
    "            break\n",
    "\n",
    "VEP_output.columns = columns"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>539283387</td>\n",
       "      <td>1:1020183</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>64</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>R/P</td>\n",
       "      <td>cGg/cCg</td>\n",
       "      <td>rs539283387</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=tolerated_low_confidence(0.22);PolyPhen=unknown(0);HGVSc=ENST00000379370.7:c.11G&gt;C;HGVSp=ENSP00000368678.2:p.Arg4Pro;AF=0.0096;AFR_AF=0.0348;AMR_AF=0.0029;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0009945;gnomADe_AFR_AF=0.04268;gnomADe_AMR_AF=0.002747;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=3.934e-05;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.002163;gnomADe_NFE_AF=3.783e-05;gnomADe_REMAINING_AF=0.002043;gnomADe_SAS_AF=4.178e-05;gnomADg_AF=0.01198;gnomADg_AFR_AF=0.04195;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003163;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001334;gnomADg_REMAINING_AF=0.008621;gnomADg_SAS_AF=0;MAX_AF=0.04268;MAX_AF_POPS=gnomADe_AFR;CLIN_SIG=benign;PHENO=1;BLOSUM62=-2;CADD_PHRED=11.57;CADD_RAW=1.122195;ClinPred=0.00471446055936884;PrimateAI=0.819412112236;REVEL=0.130;VEST4_rankscore=0.04419;VEST4_score=0.071</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>539283387</td>\n",
       "      <td>1:1020183</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000620552.4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>61</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>rs539283387</td>\n",
       "      <td>IMPACT=MODIFIER;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;ENSP=ENSP00000484607;TREMBL=A0A087X208.70;UNIPARC=UPI0004E4CB7F;HGVSc=ENST00000620552.4:c.-404G&gt;C;AF=0.0096;AFR_AF=0.0348;AMR_AF=0.0029;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0009945;gnomADe_AFR_AF=0.04268;gnomADe_AMR_AF=0.002747;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=3.934e-05;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.002163;gnomADe_NFE_AF=3.783e-05;gnomADe_REMAINING_AF=0.002043;gnomADe_SAS_AF=4.178e-05;gnomADg_AF=0.01198;gnomADg_AFR_AF=0.04195;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003163;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001334;gnomADg_REMAINING_AF=0.008621;gnomADg_SAS_AF=0;MAX_AF=0.04268;MAX_AF_POPS=gnomADe_AFR;CLIN_SIG=benign;PHENO=1;CADD_PHRED=11.57;CADD_RAW=1.122195;PrimateAI=0.819412112236</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>757604648</td>\n",
       "      <td>1:1020297</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>178</td>\n",
       "      <td>125</td>\n",
       "      <td>42</td>\n",
       "      <td>E/A</td>\n",
       "      <td>gAg/gCg</td>\n",
       "      <td>rs757604648</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=deleterious_low_confidence(0.02);PolyPhen=probably_damaging(0.996);HGVSc=ENST00000379370.7:c.125A&gt;C;HGVSp=ENSP00000368678.2:p.Glu42Ala;gnomADe_AF=8.23e-05;gnomADe_AFR_AF=3.741e-05;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0.002266;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.699e-06;gnomADe_REMAINING_AF=0.0001103;gnomADe_SAS_AF=0.0004259;gnomADg_AF=0.0001846;gnomADg_AFR_AF=2.414e-05;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.0001311;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0.004673;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.474e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=0.004673;MAX_AF_POPS=gnomADg_EAS;CLIN_SIG=benign,likely_benign;PHENO=1;BLOSUM62=-1;CADD_PHRED=24.8;CADD_RAW=4.272024;ClinPred=0.169498920597853;PrimateAI=0.942414879799;REVEL=0.270;VEST4_rankscore=0.31372;VEST4_score=0.279</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>757604648</td>\n",
       "      <td>1:1020297</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000620552.4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>rs757604648</td>\n",
       "      <td>IMPACT=MODIFIER;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;ENSP=ENSP00000484607;TREMBL=A0A087X208.70;UNIPARC=UPI0004E4CB7F;HGVSc=ENST00000620552.4:c.-290A&gt;C;gnomADe_AF=8.23e-05;gnomADe_AFR_AF=3.741e-05;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0.002266;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.699e-06;gnomADe_REMAINING_AF=0.0001103;gnomADe_SAS_AF=0.0004259;gnomADg_AF=0.0001846;gnomADg_AFR_AF=2.414e-05;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.0001311;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0.004673;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.474e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=0.004673;MAX_AF_POPS=gnomADg_EAS;CLIN_SIG=benign,likely_benign;PHENO=1;CADD_PHRED=24.8;CADD_RAW=4.272024;PrimateAI=0.942414879799</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>140954236</td>\n",
       "      <td>1:1035307</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>547</td>\n",
       "      <td>494</td>\n",
       "      <td>165</td>\n",
       "      <td>P/L</td>\n",
       "      <td>cCt/cTt</td>\n",
       "      <td>rs140954236,COSV99062194</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=tolerated_low_confidence(0.06);PolyPhen=possibly_damaging(0.64);HGVSc=ENST00000379370.7:c.494C&gt;T;HGVSp=ENSP00000368678.2:p.Pro165Leu;AF=0.0030;AFR_AF=0.0106;AMR_AF=0.0014;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0001917;gnomADe_AFR_AF=0.006959;gnomADe_AMR_AF=0.0003578;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.0005201;gnomADe_NFE_AF=1.799e-06;gnomADe_REMAINING_AF=0.0004306;gnomADe_SAS_AF=0;gnomADg_AF=0.001937;gnomADg_AFR_AF=0.006855;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.000392;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.47e-05;gnomADg_REMAINING_AF=0.001418;gnomADg_SAS_AF=0;MAX_AF=0.0106;MAX_AF_POPS=AFR;CLIN_SIG=likely_benign;SOMATIC=0,1;PHENO=1,1;BLOSUM62=-3;CADD_PHRED=15.58;CADD_RAW=1.757188;ClinPred=0.0527309163900522;PrimateAI=0.515500366688;REVEL=0.324;VEST4_rankscore=0.28880;VEST4_score=0.226,0.257,.,.</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Uploaded_variation   Location Allele             Gene            Feature Feature_type          Consequence cDNA_position CDS_position Protein_position Amino_acids   Codons        Existing_variation                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                  Extra\n",
       "0           539283387  1:1020183      C  ENSG00000188157  ENST00000379370.7   Transcript     missense_variant            64           11                4         R/P  cGg/cCg               rs539283387                            IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=tolerated_low_confidence(0.22);PolyPhen=unknown(0);HGVSc=ENST00000379370.7:c.11G>C;HGVSp=ENSP00000368678.2:p.Arg4Pro;AF=0.0096;AFR_AF=0.0348;AMR_AF=0.0029;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0009945;gnomADe_AFR_AF=0.04268;gnomADe_AMR_AF=0.002747;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=3.934e-05;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.002163;gnomADe_NFE_AF=3.783e-05;gnomADe_REMAINING_AF=0.002043;gnomADe_SAS_AF=4.178e-05;gnomADg_AF=0.01198;gnomADg_AFR_AF=0.04195;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003163;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001334;gnomADg_REMAINING_AF=0.008621;gnomADg_SAS_AF=0;MAX_AF=0.04268;MAX_AF_POPS=gnomADe_AFR;CLIN_SIG=benign;PHENO=1;BLOSUM62=-2;CADD_PHRED=11.57;CADD_RAW=1.122195;ClinPred=0.00471446055936884;PrimateAI=0.819412112236;REVEL=0.130;VEST4_rankscore=0.04419;VEST4_score=0.071\n",
       "1           539283387  1:1020183      C  ENSG00000188157  ENST00000620552.4   Transcript  5_prime_UTR_variant            61            -                -           -        -               rs539283387                                                                                                                                                                                                                                                                           IMPACT=MODIFIER;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;ENSP=ENSP00000484607;TREMBL=A0A087X208.70;UNIPARC=UPI0004E4CB7F;HGVSc=ENST00000620552.4:c.-404G>C;AF=0.0096;AFR_AF=0.0348;AMR_AF=0.0029;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0009945;gnomADe_AFR_AF=0.04268;gnomADe_AMR_AF=0.002747;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=3.934e-05;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.002163;gnomADe_NFE_AF=3.783e-05;gnomADe_REMAINING_AF=0.002043;gnomADe_SAS_AF=4.178e-05;gnomADg_AF=0.01198;gnomADg_AFR_AF=0.04195;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003163;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001334;gnomADg_REMAINING_AF=0.008621;gnomADg_SAS_AF=0;MAX_AF=0.04268;MAX_AF_POPS=gnomADe_AFR;CLIN_SIG=benign;PHENO=1;CADD_PHRED=11.57;CADD_RAW=1.122195;PrimateAI=0.819412112236\n",
       "2           757604648  1:1020297      C  ENSG00000188157  ENST00000379370.7   Transcript     missense_variant           178          125               42         E/A  gAg/gCg               rs757604648                                                                       IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=deleterious_low_confidence(0.02);PolyPhen=probably_damaging(0.996);HGVSc=ENST00000379370.7:c.125A>C;HGVSp=ENSP00000368678.2:p.Glu42Ala;gnomADe_AF=8.23e-05;gnomADe_AFR_AF=3.741e-05;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0.002266;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.699e-06;gnomADe_REMAINING_AF=0.0001103;gnomADe_SAS_AF=0.0004259;gnomADg_AF=0.0001846;gnomADg_AFR_AF=2.414e-05;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.0001311;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0.004673;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.474e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=0.004673;MAX_AF_POPS=gnomADg_EAS;CLIN_SIG=benign,likely_benign;PHENO=1;BLOSUM62=-1;CADD_PHRED=24.8;CADD_RAW=4.272024;ClinPred=0.169498920597853;PrimateAI=0.942414879799;REVEL=0.270;VEST4_rankscore=0.31372;VEST4_score=0.279\n",
       "3           757604648  1:1020297      C  ENSG00000188157  ENST00000620552.4   Transcript  5_prime_UTR_variant           175            -                -           -        -               rs757604648                                                                                                                                                                                                                                                                                                                                      IMPACT=MODIFIER;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;ENSP=ENSP00000484607;TREMBL=A0A087X208.70;UNIPARC=UPI0004E4CB7F;HGVSc=ENST00000620552.4:c.-290A>C;gnomADe_AF=8.23e-05;gnomADe_AFR_AF=3.741e-05;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0.002266;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.699e-06;gnomADe_REMAINING_AF=0.0001103;gnomADe_SAS_AF=0.0004259;gnomADg_AF=0.0001846;gnomADg_AFR_AF=2.414e-05;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.0001311;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0.004673;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.474e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=0.004673;MAX_AF_POPS=gnomADg_EAS;CLIN_SIG=benign,likely_benign;PHENO=1;CADD_PHRED=24.8;CADD_RAW=4.272024;PrimateAI=0.942414879799\n",
       "4           140954236  1:1035307      T  ENSG00000188157  ENST00000379370.7   Transcript     missense_variant           547          494              165         P/L  cCt/cTt  rs140954236,COSV99062194  IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=tolerated_low_confidence(0.06);PolyPhen=possibly_damaging(0.64);HGVSc=ENST00000379370.7:c.494C>T;HGVSp=ENSP00000368678.2:p.Pro165Leu;AF=0.0030;AFR_AF=0.0106;AMR_AF=0.0014;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0001917;gnomADe_AFR_AF=0.006959;gnomADe_AMR_AF=0.0003578;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.0005201;gnomADe_NFE_AF=1.799e-06;gnomADe_REMAINING_AF=0.0004306;gnomADe_SAS_AF=0;gnomADg_AF=0.001937;gnomADg_AFR_AF=0.006855;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.000392;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.47e-05;gnomADg_REMAINING_AF=0.001418;gnomADg_SAS_AF=0;MAX_AF=0.0106;MAX_AF_POPS=AFR;CLIN_SIG=likely_benign;SOMATIC=0,1;PHENO=1,1;BLOSUM62=-3;CADD_PHRED=15.58;CADD_RAW=1.757188;ClinPred=0.0527309163900522;PrimateAI=0.515500366688;REVEL=0.324;VEST4_rankscore=0.28880;VEST4_score=0.226,0.257,.,."
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VEP_output.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0    IMPACT=MODERATE;STRAND=1;SYMBOL=AGRN;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:329;CANONICAL=YES;CCDS=CCDS30551.1;ENSP=ENSP00000368678;SWISSPROT=O00468.212;UNIPARC=UPI00001D7C8B;UNIPROT_ISOFORM=O00468-6;SIFT=tolerated_low_confidence(0.22);PolyPhen=unknown(0);HGVSc=ENST00000379370.7:c.11G>C;HGVSp=ENSP00000368678.2:p.Arg4Pro;AF=0.0096;AFR_AF=0.0348;AMR_AF=0.0029;EAS_AF=0;EUR_AF=0;SAS_AF=0;gnomADe_AF=0.0009945;gnomADe_AFR_AF=0.04268;gnomADe_AMR_AF=0.002747;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=3.934e-05;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.002163;gnomADe_NFE_AF=3.783e-05;gnomADe_REMAINING_AF=0.002043;gnomADe_SAS_AF=4.178e-05;gnomADg_AF=0.01198;gnomADg_AFR_AF=0.04195;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003163;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001334;gnomADg_REMAINING_AF=0.008621;gnomADg_SAS_AF=0;MAX_AF=0.04268;MAX_AF_POPS=gnomADe_AFR;CLIN_SIG=benign;PHENO=1;BLOSUM62=-2;CADD_PHRED=11.57;CADD_RAW=1.122195;ClinPred=0.00471446055936884;PrimateAI=0.819412112236;REVEL=0.130;VEST4_rankscore=0.04419;VEST4_score=0.071\n",
       "Name: Extra, dtype: object"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VEP_output[\"Extra\"].head(1)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Uploaded_variation', 'Location', 'Allele', 'Gene', 'Feature',\n",
       "       'Feature_type', 'Consequence', 'cDNA_position', 'CDS_position',\n",
       "       'Protein_position', 'Amino_acids', 'Codons', 'Existing_variation',\n",
       "       'Extra'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "VEP_output.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "After obtaining the VEP output, we need to **parse** it to extract **relevant** information. The `Extra` column in the VEP output contains multiple annotations in a single field, so we extract these as separate formal columns. Additionally, predictor outputs (e.g., SIFT, PolyPhen) are split into distinct columns for labels and scores.  \n",
    "\n",
    "This step ensures that the data is structured correctly for analysis. Once parsed, we will merge this processed VEP output with our original dataset, using common columns to retain all the initial information while incorporating the VEP predictions.  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "import parsing_ClinVar as parse\n",
    "\n",
    "input_file = '../data/clinvar/cleaned_Clinvar_dataset_outputVEP.txt'  \n",
    "output_file = '../data/clinvar/cleaned_Clinvar_dataset_parsed.csv'  \n",
    "\n",
    "parse.parse_vep_output(input_file, output_file)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "output_file = '../data/clinvar/cleaned_Clinvar_dataset_parsed.csv' "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "df = pd.read_csv(output_file)  "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Canonical</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>Swissprot</th>\n",
       "      <th>Uniparc</th>\n",
       "      <th>Uniprot_isoform</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>SIFT_label</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>PolyPhen_label</th>\n",
       "      <th>PolyPhen_score</th>\n",
       "      <th>BayesDel_label</th>\n",
       "      <th>CADD_PHRED_score</th>\n",
       "      <th>CADD_RAW_score</th>\n",
       "      <th>ClinPred_score</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>EVE_label</th>\n",
       "      <th>EVE_score</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>AM_label</th>\n",
       "      <th>AM_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>rs539283387</td>\n",
       "      <td>1:1020183</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>YES</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>O00468.212</td>\n",
       "      <td>UPI00001D7C8B</td>\n",
       "      <td>O00468-6</td>\n",
       "      <td>64</td>\n",
       "      <td>11</td>\n",
       "      <td>4</td>\n",
       "      <td>R/P</td>\n",
       "      <td>cGg/cCg</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>HGNC:329</td>\n",
       "      <td>tolerated_low_confidence</td>\n",
       "      <td>0.22</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.57</td>\n",
       "      <td>1.122195</td>\n",
       "      <td>0.004714</td>\n",
       "      <td>0.071</td>\n",
       "      <td>0.04419</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.130</td>\n",
       "      <td>0.819412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>rs539283387</td>\n",
       "      <td>1:1020183</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000620552.4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UPI0004E4CB7F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>61</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>HGNC:329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>11.57</td>\n",
       "      <td>1.122195</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.819412</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>rs757604648</td>\n",
       "      <td>1:1020297</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>YES</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>O00468.212</td>\n",
       "      <td>UPI00001D7C8B</td>\n",
       "      <td>O00468-6</td>\n",
       "      <td>178</td>\n",
       "      <td>125</td>\n",
       "      <td>42</td>\n",
       "      <td>E/A</td>\n",
       "      <td>gAg/gCg</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>HGNC:329</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>0.02</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.996</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.80</td>\n",
       "      <td>4.272024</td>\n",
       "      <td>0.169499</td>\n",
       "      <td>0.279</td>\n",
       "      <td>0.31372</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.270</td>\n",
       "      <td>0.942415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>rs757604648</td>\n",
       "      <td>1:1020297</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000620552.4</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>NaN</td>\n",
       "      <td>5_prime_UTR_variant</td>\n",
       "      <td>NaN</td>\n",
       "      <td>UPI0004E4CB7F</td>\n",
       "      <td>NaN</td>\n",
       "      <td>175</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>-</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>HGNC:329</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>24.80</td>\n",
       "      <td>4.272024</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.942415</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>rs140954236,COSV99062194</td>\n",
       "      <td>1:1035307</td>\n",
       "      <td>ENSG00000188157</td>\n",
       "      <td>ENST00000379370.7</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>YES</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>O00468.212</td>\n",
       "      <td>UPI00001D7C8B</td>\n",
       "      <td>O00468-6</td>\n",
       "      <td>547</td>\n",
       "      <td>494</td>\n",
       "      <td>165</td>\n",
       "      <td>P/L</td>\n",
       "      <td>cCt/cTt</td>\n",
       "      <td>AGRN</td>\n",
       "      <td>HGNC:329</td>\n",
       "      <td>tolerated_low_confidence</td>\n",
       "      <td>0.06</td>\n",
       "      <td>possibly_damaging</td>\n",
       "      <td>0.640</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15.58</td>\n",
       "      <td>1.757188</td>\n",
       "      <td>0.052731</td>\n",
       "      <td>0.257</td>\n",
       "      <td>0.28880</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.324</td>\n",
       "      <td>0.515500</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         Existing_variation   Location             Gene            Feature Feature_type Canonical          Consequence   Swissprot        Uniparc Uniprot_isoform cDNA_position CDS_position Protein_position Amino_acids   Codons GeneSymbol   HGNC_ID                  SIFT_label  SIFT_score     PolyPhen_label  PolyPhen_score  BayesDel_label  CADD_PHRED_score  CADD_RAW_score  ClinPred_score  VEST4_score  VEST4_rankscore EVE_label  EVE_score  REVEL_score  PrimateAI_score AM_label  AM_score\n",
       "0               rs539283387  1:1020183  ENSG00000188157  ENST00000379370.7   Transcript       YES     missense_variant  O00468.212  UPI00001D7C8B        O00468-6            64           11                4         R/P  cGg/cCg       AGRN  HGNC:329    tolerated_low_confidence        0.22            unknown           0.000             NaN             11.57        1.122195        0.004714        0.071          0.04419       NaN        NaN        0.130         0.819412      NaN       NaN\n",
       "1               rs539283387  1:1020183  ENSG00000188157  ENST00000620552.4   Transcript       NaN  5_prime_UTR_variant         NaN  UPI0004E4CB7F             NaN            61            -                -           -        -       AGRN  HGNC:329                         NaN         NaN                NaN             NaN             NaN             11.57        1.122195             NaN        0.000              NaN       NaN        NaN          NaN         0.819412      NaN       NaN\n",
       "2               rs757604648  1:1020297  ENSG00000188157  ENST00000379370.7   Transcript       YES     missense_variant  O00468.212  UPI00001D7C8B        O00468-6           178          125               42         E/A  gAg/gCg       AGRN  HGNC:329  deleterious_low_confidence        0.02  probably_damaging           0.996             NaN             24.80        4.272024        0.169499        0.279          0.31372       NaN        NaN        0.270         0.942415      NaN       NaN\n",
       "3               rs757604648  1:1020297  ENSG00000188157  ENST00000620552.4   Transcript       NaN  5_prime_UTR_variant         NaN  UPI0004E4CB7F             NaN           175            -                -           -        -       AGRN  HGNC:329                         NaN         NaN                NaN             NaN             NaN             24.80        4.272024             NaN        0.000              NaN       NaN        NaN          NaN         0.942415      NaN       NaN\n",
       "4  rs140954236,COSV99062194  1:1035307  ENSG00000188157  ENST00000379370.7   Transcript       YES     missense_variant  O00468.212  UPI00001D7C8B        O00468-6           547          494              165         P/L  cCt/cTt       AGRN  HGNC:329    tolerated_low_confidence        0.06  possibly_damaging           0.640             NaN             15.58        1.757188        0.052731        0.257          0.28880       NaN        NaN        0.324         0.515500      NaN       NaN"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "VEP output contains multiple rows for the same variant due to different transcript versions. As in this case, 2,559 variants were found in the VEP output. However, to maintain consistency with our original dataset, we need to carefully select specific columns for merging. This ensures that we accurately add only the pathogenicity predictions without introducing redundant or erroneous rows of variants."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2559"
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df[df[\"GeneSymbol\"] == \"BRCA1\"].shape[0]"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "From the original ClinVar dataset, we retrieved 82 variants for the BRCA1 case. After merging, we expect to retain the same number of variants, with the only difference being the addition of prediction columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "82"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cleaned_ClinVar_dataset[cleaned_ClinVar_dataset[\"GeneSymbol\"] == \"BRCA1\"].shape[0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Index(['Existing_variation', 'Location', 'Gene', 'Feature', 'Feature_type',\n",
       "       'Canonical', 'Consequence', 'Swissprot', 'Uniparc', 'Uniprot_isoform',\n",
       "       'cDNA_position', 'CDS_position', 'Protein_position', 'Amino_acids',\n",
       "       'Codons', 'GeneSymbol', 'HGNC_ID', 'SIFT_label', 'SIFT_score',\n",
       "       'PolyPhen_label', 'PolyPhen_score', 'BayesDel_label',\n",
       "       'CADD_PHRED_score', 'CADD_RAW_score', 'ClinPred_score', 'VEST4_score',\n",
       "       'VEST4_rankscore', 'EVE_label', 'EVE_score', 'REVEL_score',\n",
       "       'PrimateAI_score', 'AM_label', 'AM_score'],\n",
       "      dtype='object')"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df.columns"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "An important step is to calculate the coverage for each predictor in the dataset. With this we aim to determine the percentage of non NaN values for each predictor score column and output a table sorted by coverage percentage.\n",
    "\n",
    "Function is defined now for later usage."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "def check_coverage(df):\n",
    "    predictors = [i for i in df.columns if '_score' in i]\n",
    "    adding = []\n",
    "    predictor_groups = {'CADD': ['CADD_RAW_score', 'CADD_PHRED_score']}\n",
    "\n",
    "    for predictor in predictors:\n",
    "        if predictor in predictor_groups.get('CADD', []):\n",
    "            coverage_type = 'RAW' if 'RAW' in predictor else 'PHRED'\n",
    "            predictor_label = f'CADD ({coverage_type})'\n",
    "        else:\n",
    "            predictor_label = predictor.split('_')[0]\n",
    "            if predictor_label == 'AM':\n",
    "                predictor_label = 'AlphaMissense' \n",
    "\n",
    "        tmp = df[df[predictor].notna()]  \n",
    "        coverage = round(100 * len(tmp) / len(df), 2)\n",
    "        adding.append([predictor_label, coverage])\n",
    "\n",
    "    table = pd.DataFrame(adding, columns=['Predictor', 'Coverage'])\n",
    "    table = table.sort_values('Coverage', ascending=False).reset_index(drop=True)\n",
    "    return table"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Final step: merge VEP output with ClinVar original dataset"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "A Python script was used to merge ClinVar data with VEP predictions by first filtering ClinVar variants for the specified gene and then matching them with VEP entries based on *Feature*, *Existing_variation*, and *Codons*. If multiple matches exist, the row with the least missing values is selected. This ensures that only the relevant pathogenicity predictions are added without redundancy. Unmatched variants are logged, and the final dataset retains the original ClinVar structure with the additional prediction columns."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {
    "vscode": {
     "languageId": "shellscript"
    }
   },
   "outputs": [],
   "source": [
    "#    python3 merging_ClinVar.py"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "49187"
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds = pd.read_csv('../data/clinvar/cleaned_ClinVar_with_preds.csv')\n",
    "len(clinvar_with_preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_with_preds = clinvar_with_preds.sort_values(['GeneSymbol'])\n",
    "clinvar_with_preds = clinvar_with_preds.reset_index(drop=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>Variant (3-letter)</th>\n",
       "      <th>Variant</th>\n",
       "      <th>LastEvaluated (Year)</th>\n",
       "      <th>Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "      <th>PositionVCF_dashed</th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>gnomADg_AF</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>BinaryClinicalSignificance</th>\n",
       "      <th>SIFT_label</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>PolyPhen_label</th>\n",
       "      <th>PolyPhen_score</th>\n",
       "      <th>BayesDel_label</th>\n",
       "      <th>BayesDel_score</th>\n",
       "      <th>CADD_PHRED_score</th>\n",
       "      <th>CADD_RAW_score</th>\n",
       "      <th>ClinPred_score</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>EVE_label</th>\n",
       "      <th>EVE_score</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>AM_label</th>\n",
       "      <th>AM_score</th>\n",
       "      <th>Uniprot_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>317931</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_015665.6(AAAS):c.1597G&gt;A (p.Gly533Arg)</td>\n",
       "      <td>8086</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>HGNC:13666</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>34451260</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000343022|RCV000886859</td>\n",
       "      <td>MONDO:MONDO:0009279,MedGen:C0271742,OMIM:231550,Orphanet:869|MedGen:C3661900</td>\n",
       "      <td>Glucocorticoid deficiency with achalasia|not provided</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>53307533</td>\n",
       "      <td>53307533</td>\n",
       "      <td>12q13.13</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>3</td>\n",
       "      <td>ClinGen:CA6598796</td>\n",
       "      <td>2</td>\n",
       "      <td>309718</td>\n",
       "      <td>53307533</td>\n",
       "      <td>C</td>\n",
       "      <td>T</td>\n",
       "      <td>Gly533Arg</td>\n",
       "      <td>G533R</td>\n",
       "      <td>2024</td>\n",
       "      <td>3.445126e+07</td>\n",
       "      <td>12:53307533</td>\n",
       "      <td>T</td>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>ENST00000209873.9</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>1742</td>\n",
       "      <td>1597</td>\n",
       "      <td>533</td>\n",
       "      <td>G/R</td>\n",
       "      <td>Ggg/Agg</td>\n",
       "      <td>rs34451260</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=AAAS;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:13666;CCDS=CCDS8856.1;ENSP=ENSP00000209873;SWISSPROT=Q9NRG9.206;UNIPARC=UPI0000039E40;UNIPROT_ISOFORM=Q9NRG9-1;HGVSc=ENST00000209873.9:c.1597G&gt;A;HGVSp=ENSP00000209873.4:p.Gly533Arg;AF=0.0124;AFR_AF=0.0439;AMR_AF=0.0043;EAS_AF=0;EUR_AF=0.001;SAS_AF=0;gnomADe_AF=0.0007928;gnomADe_AFR_AF=0.02712;gnomADe_AMR_AF=0.001879;gnomADe_ASJ_AF=3.827e-05;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.001214;gnomADe_NFE_AF=2.428e-05;gnomADe_REMAINING_AF=0.002053;gnomADe_SAS_AF=9.275e-05;gnomADg_AF=0.007867;gnomADg_AFR_AF=0.02713;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003529;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001029;gnomADg_REMAINING_AF=0.004748;gnomADg_SAS_AF=0;MAX_AF=0.0439;MAX_AF_POPS=AFR;CLIN_SIG=benign,likely_benign;PHENO=1</td>\n",
       "      <td>53307533</td>\n",
       "      <td>0.000793</td>\n",
       "      <td>0.007867</td>\n",
       "      <td>0.000793</td>\n",
       "      <td>ENSP00000209873.4:p.Gly533Arg</td>\n",
       "      <td>B</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>0.00</td>\n",
       "      <td>unknown</td>\n",
       "      <td>0.000</td>\n",
       "      <td>T</td>\n",
       "      <td>-0.362112</td>\n",
       "      <td>20.9</td>\n",
       "      <td>2.951810</td>\n",
       "      <td>0.062144</td>\n",
       "      <td>0.420</td>\n",
       "      <td>0.45743</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.257</td>\n",
       "      <td>0.526710</td>\n",
       "      <td>likely_benign</td>\n",
       "      <td>0.1192</td>\n",
       "      <td>Q9NRG9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>20084</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_015665.6(AAAS):c.787T&gt;C (p.Ser263Pro)</td>\n",
       "      <td>8086</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>HGNC:13666</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>121918550</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000005348|RCV000311283|RCV000415076|RCV000624696</td>\n",
       "      <td>MONDO:MONDO:0009279,MedGen:C0271742,OMIM:231550,Orphanet:869|MedGen:C3661900|Human Phenotype Ontology:HP:0002313,Human Phenotype Ontology:HP:0007191,MedGen:C0037771;Human Phenotype Ontology:HP:0001282,Human Phenotype Ontology:HP:0001347,Human Phenotype Ontology:HP:0006820,Human Phenotype Ontology:HP:0007184,Human Phenotype Ontology:HP:0007318,MONDO:MONDO:0007774,MedGen:C0151889,OMIM:145290;Human Phenotype Ontology:HP:0001352,Human Phenotype Ontology:HP:0003487,MedGen:C0034935|MeSH:D030342,MedGen:C0950123</td>\n",
       "      <td>Glucocorticoid deficiency with achalasia|not provided|Spastic paraparesis;Hyperreflexia;Babinski sign|Inborn genetic diseases</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>53309624</td>\n",
       "      <td>53309624</td>\n",
       "      <td>12q13.13</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>11</td>\n",
       "      <td>ClinGen:CA117228,UniProtKB:Q9NRG9#VAR_012806,OMIM:605378.0007</td>\n",
       "      <td>3</td>\n",
       "      <td>5045</td>\n",
       "      <td>53309624</td>\n",
       "      <td>A</td>\n",
       "      <td>G</td>\n",
       "      <td>Ser263Pro</td>\n",
       "      <td>S263P</td>\n",
       "      <td>2024</td>\n",
       "      <td>1.219186e+08</td>\n",
       "      <td>12:53309624</td>\n",
       "      <td>G</td>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>ENST00000209873.9</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>932</td>\n",
       "      <td>787</td>\n",
       "      <td>263</td>\n",
       "      <td>S/P</td>\n",
       "      <td>Tca/Cca</td>\n",
       "      <td>rs121918550,CM010150</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=AAAS;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:13666;CCDS=CCDS8856.1;ENSP=ENSP00000209873;SWISSPROT=Q9NRG9.206;UNIPARC=UPI0000039E40;UNIPROT_ISOFORM=Q9NRG9-1;HGVSc=ENST00000209873.9:c.787T&gt;C;HGVSp=ENSP00000209873.4:p.Ser263Pro;AF=0.0002;AFR_AF=0;AMR_AF=0;EAS_AF=0;EUR_AF=0.001;SAS_AF=0;gnomADe_AF=7.801e-05;gnomADe_AFR_AF=2.988e-05;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0.0005994;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.925e-05;gnomADe_REMAINING_AF=6.627e-05;gnomADe_SAS_AF=0;gnomADg_AF=5.908e-05;gnomADg_AFR_AF=0;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0.0002824;gnomADg_MID_AF=0;gnomADg_NFE_AF=8.819e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=0.001;MAX_AF_POPS=EUR;CLIN_SIG=pathogenic/likely_pathogenic,pathogenic;SOMATIC=0,1;PHENO=1,1</td>\n",
       "      <td>53309624</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>0.000059</td>\n",
       "      <td>0.000078</td>\n",
       "      <td>ENSP00000209873.4:p.Ser263Pro</td>\n",
       "      <td>P</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.00</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.992</td>\n",
       "      <td>D</td>\n",
       "      <td>0.258310</td>\n",
       "      <td>29.2</td>\n",
       "      <td>5.211912</td>\n",
       "      <td>0.980945</td>\n",
       "      <td>0.983</td>\n",
       "      <td>0.99270</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.951397</td>\n",
       "      <td>0.844</td>\n",
       "      <td>0.718980</td>\n",
       "      <td>likely_pathogenic</td>\n",
       "      <td>0.9852</td>\n",
       "      <td>Q9NRG9</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>1312858</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_015665.6(AAAS):c.500C&gt;T (p.Ala167Val)</td>\n",
       "      <td>8086</td>\n",
       "      <td>AAAS</td>\n",
       "      <td>HGNC:13666</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>1017700992</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV003557743|RCV004765890</td>\n",
       "      <td>MedGen:C3661900|MONDO:MONDO:0009279,MedGen:C0271742,OMIM:231550,Orphanet:869</td>\n",
       "      <td>not provided|Glucocorticoid deficiency with achalasia</td>\n",
       "      <td>germline</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000012.12</td>\n",
       "      <td>12</td>\n",
       "      <td>53314796</td>\n",
       "      <td>53314796</td>\n",
       "      <td>12q13.13</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>2</td>\n",
       "      <td>ClinGen:CA237335767</td>\n",
       "      <td>2</td>\n",
       "      <td>2735891</td>\n",
       "      <td>53314796</td>\n",
       "      <td>G</td>\n",
       "      <td>A</td>\n",
       "      <td>Ala167Val</td>\n",
       "      <td>A167V</td>\n",
       "      <td>2024</td>\n",
       "      <td>1.017701e+09</td>\n",
       "      <td>12:53314796</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000094914</td>\n",
       "      <td>ENST00000209873.9</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>645</td>\n",
       "      <td>500</td>\n",
       "      <td>167</td>\n",
       "      <td>A/V</td>\n",
       "      <td>gCa/gTa</td>\n",
       "      <td>rs1017700992,CM065949</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=AAAS;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:13666;CCDS=CCDS8856.1;ENSP=ENSP00000209873;SWISSPROT=Q9NRG9.206;UNIPARC=UPI0000039E40;UNIPROT_ISOFORM=Q9NRG9-1;HGVSc=ENST00000209873.9:c.500C&gt;T;HGVSp=ENSP00000209873.4:p.Ala167Val;gnomADe_AF=3.421e-06;gnomADe_AFR_AF=0;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=4.497e-06;gnomADe_REMAINING_AF=0;gnomADe_SAS_AF=0;MAX_AF=4.497e-06;MAX_AF_POPS=gnomADe_NFE;CLIN_SIG=pathogenic;SOMATIC=0,1;PHENO=1,1</td>\n",
       "      <td>53314796</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.000003</td>\n",
       "      <td>ENSP00000209873.4:p.Ala167Val</td>\n",
       "      <td>P</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.03</td>\n",
       "      <td>probably_damaging</td>\n",
       "      <td>0.969</td>\n",
       "      <td>D</td>\n",
       "      <td>0.571933</td>\n",
       "      <td>34.0</td>\n",
       "      <td>5.972388</td>\n",
       "      <td>0.998261</td>\n",
       "      <td>0.980</td>\n",
       "      <td>0.99026</td>\n",
       "      <td>Pathogenic</td>\n",
       "      <td>0.853707</td>\n",
       "      <td>0.920</td>\n",
       "      <td>0.760144</td>\n",
       "      <td>likely_pathogenic</td>\n",
       "      <td>0.8614</td>\n",
       "      <td>Q9NRG9</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   #AlleleID                       Type                                       Name  GeneID GeneSymbol     HGNC_ID          ClinicalSignificance  ClinSigSimple  RS# (dbSNP) nsv/esv (dbVar)                                         RCVaccession                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                   PhenotypeIDS                                                                                                                  PhenotypeList            Origin OriginSimple Assembly ChromosomeAccession Chromosome     Start      Stop Cytogenetic                                          ReviewStatus  NumberSubmitters                                                       OtherIDs  SubmitterCategories  VariationID  PositionVCF ReferenceAlleleVCF AlternateAlleleVCF Variant (3-letter) Variant  LastEvaluated (Year)  Uploaded_variation     Location Allele             Gene            Feature Feature_type       Consequence  cDNA_position  CDS_position  Protein_position Amino_acids   Codons     Existing_variation                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                       Extra  PositionVCF_dashed  gnomADe_AF  gnomADg_AF  gnomAD_AF                          HGVSp BinaryClinicalSignificance                  SIFT_label  SIFT_score     PolyPhen_label  PolyPhen_score BayesDel_label  BayesDel_score  CADD_PHRED_score  CADD_RAW_score  ClinPred_score  VEST4_score  VEST4_rankscore   EVE_label  EVE_score  REVEL_score  PrimateAI_score           AM_label  AM_score Uniprot_acc\n",
       "0     317931  single nucleotide variant  NM_015665.6(AAAS):c.1597G>A (p.Gly533Arg)    8086       AAAS  HGNC:13666          Benign/Likely benign              0     34451260               -                            RCV000343022|RCV000886859                                                                                                                                                                                                                                                                                                                                                                                                                                                   MONDO:MONDO:0009279,MedGen:C0271742,OMIM:231550,Orphanet:869|MedGen:C3661900                                                                          Glucocorticoid deficiency with achalasia|not provided          germline     germline   GRCh38        NC_000012.12         12  53307533  53307533    12q13.13  criteria provided, multiple submitters, no conflicts                 3                                              ClinGen:CA6598796                    2       309718     53307533                  C                  T          Gly533Arg   G533R                  2024        3.445126e+07  12:53307533      T  ENSG00000094914  ENST00000209873.9   Transcript  missense_variant           1742          1597               533         G/R  Ggg/Agg             rs34451260  IMPACT=MODERATE;STRAND=-1;SYMBOL=AAAS;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:13666;CCDS=CCDS8856.1;ENSP=ENSP00000209873;SWISSPROT=Q9NRG9.206;UNIPARC=UPI0000039E40;UNIPROT_ISOFORM=Q9NRG9-1;HGVSc=ENST00000209873.9:c.1597G>A;HGVSp=ENSP00000209873.4:p.Gly533Arg;AF=0.0124;AFR_AF=0.0439;AMR_AF=0.0043;EAS_AF=0;EUR_AF=0.001;SAS_AF=0;gnomADe_AF=0.0007928;gnomADe_AFR_AF=0.02712;gnomADe_AMR_AF=0.001879;gnomADe_ASJ_AF=3.827e-05;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0.001214;gnomADe_NFE_AF=2.428e-05;gnomADe_REMAINING_AF=0.002053;gnomADe_SAS_AF=9.275e-05;gnomADg_AF=0.007867;gnomADg_AFR_AF=0.02713;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0.003529;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=0.0001029;gnomADg_REMAINING_AF=0.004748;gnomADg_SAS_AF=0;MAX_AF=0.0439;MAX_AF_POPS=AFR;CLIN_SIG=benign,likely_benign;PHENO=1            53307533    0.000793    0.007867   0.000793  ENSP00000209873.4:p.Gly533Arg                          B  deleterious_low_confidence        0.00            unknown           0.000              T       -0.362112              20.9        2.951810        0.062144        0.420          0.45743         NaN        NaN        0.257         0.526710      likely_benign    0.1192      Q9NRG9\n",
       "1      20084  single nucleotide variant   NM_015665.6(AAAS):c.787T>C (p.Ser263Pro)    8086       AAAS  HGNC:13666  Pathogenic/Likely pathogenic              1    121918550               -  RCV000005348|RCV000311283|RCV000415076|RCV000624696  MONDO:MONDO:0009279,MedGen:C0271742,OMIM:231550,Orphanet:869|MedGen:C3661900|Human Phenotype Ontology:HP:0002313,Human Phenotype Ontology:HP:0007191,MedGen:C0037771;Human Phenotype Ontology:HP:0001282,Human Phenotype Ontology:HP:0001347,Human Phenotype Ontology:HP:0006820,Human Phenotype Ontology:HP:0007184,Human Phenotype Ontology:HP:0007318,MONDO:MONDO:0007774,MedGen:C0151889,OMIM:145290;Human Phenotype Ontology:HP:0001352,Human Phenotype Ontology:HP:0003487,MedGen:C0034935|MeSH:D030342,MedGen:C0950123  Glucocorticoid deficiency with achalasia|not provided|Spastic paraparesis;Hyperreflexia;Babinski sign|Inborn genetic diseases  germline;unknown     germline   GRCh38        NC_000012.12         12  53309624  53309624    12q13.13  criteria provided, multiple submitters, no conflicts                11  ClinGen:CA117228,UniProtKB:Q9NRG9#VAR_012806,OMIM:605378.0007                    3         5045     53309624                  A                  G          Ser263Pro   S263P                  2024        1.219186e+08  12:53309624      G  ENSG00000094914  ENST00000209873.9   Transcript  missense_variant            932           787               263         S/P  Tca/Cca   rs121918550,CM010150           IMPACT=MODERATE;STRAND=-1;SYMBOL=AAAS;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:13666;CCDS=CCDS8856.1;ENSP=ENSP00000209873;SWISSPROT=Q9NRG9.206;UNIPARC=UPI0000039E40;UNIPROT_ISOFORM=Q9NRG9-1;HGVSc=ENST00000209873.9:c.787T>C;HGVSp=ENSP00000209873.4:p.Ser263Pro;AF=0.0002;AFR_AF=0;AMR_AF=0;EAS_AF=0;EUR_AF=0.001;SAS_AF=0;gnomADe_AF=7.801e-05;gnomADe_AFR_AF=2.988e-05;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0.0005994;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.925e-05;gnomADe_REMAINING_AF=6.627e-05;gnomADe_SAS_AF=0;gnomADg_AF=5.908e-05;gnomADg_AFR_AF=0;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0.0002824;gnomADg_MID_AF=0;gnomADg_NFE_AF=8.819e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=0.001;MAX_AF_POPS=EUR;CLIN_SIG=pathogenic/likely_pathogenic,pathogenic;SOMATIC=0,1;PHENO=1,1            53309624    0.000078    0.000059   0.000078  ENSP00000209873.4:p.Ser263Pro                          P                 deleterious        0.00  probably_damaging           0.992              D        0.258310              29.2        5.211912        0.980945        0.983          0.99270  Pathogenic   0.951397        0.844         0.718980  likely_pathogenic    0.9852      Q9NRG9\n",
       "2    1312858  single nucleotide variant   NM_015665.6(AAAS):c.500C>T (p.Ala167Val)    8086       AAAS  HGNC:13666                    Pathogenic              1   1017700992               -                            RCV003557743|RCV004765890                                                                                                                                                                                                                                                                                                                                                                                                                                                   MedGen:C3661900|MONDO:MONDO:0009279,MedGen:C0271742,OMIM:231550,Orphanet:869                                                                          not provided|Glucocorticoid deficiency with achalasia          germline     germline   GRCh38        NC_000012.12         12  53314796  53314796    12q13.13  criteria provided, multiple submitters, no conflicts                 2                                            ClinGen:CA237335767                    2      2735891     53314796                  G                  A          Ala167Val   A167V                  2024        1.017701e+09  12:53314796      A  ENSG00000094914  ENST00000209873.9   Transcript  missense_variant            645           500               167         A/V  gCa/gTa  rs1017700992,CM065949                                                                                                                                                                                                                                                                                                                                    IMPACT=MODERATE;STRAND=-1;SYMBOL=AAAS;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:13666;CCDS=CCDS8856.1;ENSP=ENSP00000209873;SWISSPROT=Q9NRG9.206;UNIPARC=UPI0000039E40;UNIPROT_ISOFORM=Q9NRG9-1;HGVSc=ENST00000209873.9:c.500C>T;HGVSp=ENSP00000209873.4:p.Ala167Val;gnomADe_AF=3.421e-06;gnomADe_AFR_AF=0;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=4.497e-06;gnomADe_REMAINING_AF=0;gnomADe_SAS_AF=0;MAX_AF=4.497e-06;MAX_AF_POPS=gnomADe_NFE;CLIN_SIG=pathogenic;SOMATIC=0,1;PHENO=1,1            53314796    0.000003         NaN   0.000003  ENSP00000209873.4:p.Ala167Val                          P                 deleterious        0.03  probably_damaging           0.969              D        0.571933              34.0        5.972388        0.998261        0.980          0.99026  Pathogenic   0.853707        0.920         0.760144  likely_pathogenic    0.8614      Q9NRG9"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "2156"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds['GeneSymbol'].nunique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneSymbol\n",
       "DNAH11       1023\n",
       "KMT2D         808\n",
       "NEB           593\n",
       "ADGRV1        586\n",
       "SACS          495\n",
       "DNAH5         444\n",
       "TTN           322\n",
       "COL7A1        313\n",
       "ABCA4         301\n",
       "ATRX          295\n",
       "DMD           287\n",
       "KMT2B         284\n",
       "CACNA1H       280\n",
       "APOB          278\n",
       "MACF1         253\n",
       "CHD7          245\n",
       "KAT6A         240\n",
       "RAI1          234\n",
       "NOTCH1        233\n",
       "SETBP1        226\n",
       "USH2A         223\n",
       "FLNA          209\n",
       "LDLR          206\n",
       "COL6A3        205\n",
       "TSC2          201\n",
       "DYNC1H1       192\n",
       "FBN2          182\n",
       "EP300         182\n",
       "ANKRD11       181\n",
       "COL5A1        179\n",
       "ATP7A         175\n",
       "ARID1B        173\n",
       "COL4A5        173\n",
       "SETX          168\n",
       "COL11A1       164\n",
       "NBAS          161\n",
       "MECP2         159\n",
       "MTOR          156\n",
       "SON           152\n",
       "TP53          151\n",
       "FLNB          148\n",
       "PKD1          146\n",
       "GRIN2A        145\n",
       "ARID1A        143\n",
       "CREBBP        142\n",
       "ATP7B         140\n",
       "L1CAM         140\n",
       "ABL1          138\n",
       "MED13L        137\n",
       "COL2A1        136\n",
       "PTCH1         134\n",
       "COL12A1       130\n",
       "NSD1          128\n",
       "NOTCH3        127\n",
       "COL11A2       125\n",
       "GAA           125\n",
       "LAMA5         124\n",
       "ABCA12        122\n",
       "CACNA1A       122\n",
       "ALPL          122\n",
       "COL6A1        119\n",
       "COL4A3        119\n",
       "PKHD1         117\n",
       "DYSF          117\n",
       "FBN1          116\n",
       "COL4A4        115\n",
       "ENG           115\n",
       "SLC12A3       114\n",
       "PAH           114\n",
       "INF2          113\n",
       "MYO7A         111\n",
       "RYR1          109\n",
       "COL1A1        108\n",
       "KMT2A         106\n",
       "SLC26A4       106\n",
       "NPC1          103\n",
       "FLNC          101\n",
       "RP1L1         101\n",
       "MYO15A        101\n",
       "KIF1A         100\n",
       "SCN1A         100\n",
       "DSP           100\n",
       "KCNT1          98\n",
       "BRCA2          98\n",
       "ABCD1          97\n",
       "GCDH           97\n",
       "KANSL1         96\n",
       "GCK            96\n",
       "SETD2          96\n",
       "IFIH1          93\n",
       "PIEZO1         91\n",
       "OTOF           90\n",
       "ARSA           90\n",
       "CACNA1G        89\n",
       "TNXB           89\n",
       "TCF20          89\n",
       "CRB1           88\n",
       "SETD5          87\n",
       "LAMA2          86\n",
       "HCFC1          86\n",
       "RNF213         85\n",
       "RBM20          85\n",
       "KCNB1          83\n",
       "TYR            83\n",
       "CDH1           83\n",
       "BRCA1          82\n",
       "CARD11         81\n",
       "MMUT           80\n",
       "CDK13          80\n",
       "GLDC           80\n",
       "MYH9           79\n",
       "GRIN2B         78\n",
       "SPTAN1         78\n",
       "GALNS          78\n",
       "PRR12          77\n",
       "KCNQ1          77\n",
       "OCA2           77\n",
       "CNGA3          77\n",
       "DNAH1          77\n",
       "PDGFRB         77\n",
       "CPS1           77\n",
       "MSH2           76\n",
       "CAPN3          75\n",
       "CLCN1          75\n",
       "BICD2          75\n",
       "SLC22A5        74\n",
       "SPTBN2         74\n",
       "ACADVL         74\n",
       "F8             74\n",
       "COL5A2         73\n",
       "COL27A1        73\n",
       "CIC            73\n",
       "ABCB11         73\n",
       "VWF            72\n",
       "ABCC6          71\n",
       "EPHB4          71\n",
       "NPHS1          71\n",
       "SMARCA2        70\n",
       "GLB1           70\n",
       "MYH7           70\n",
       "RPE65          70\n",
       "LRP2           69\n",
       "BPTF           69\n",
       "CACNA1C        69\n",
       "NF1            69\n",
       "POLG           69\n",
       "DNAH9          68\n",
       "COL1A2         68\n",
       "DHCR7          68\n",
       "NEDD4L         68\n",
       "IQSEC2         68\n",
       "CDKL5          68\n",
       "CACNA1S        68\n",
       "GALT           67\n",
       "MED12          67\n",
       "CSF1R          66\n",
       "NAGLU          66\n",
       "ALMS1          66\n",
       "COL6A2         66\n",
       "COL4A2         66\n",
       "CDH23          66\n",
       "BEST1          66\n",
       "TGM1           65\n",
       "DYNC2H1        65\n",
       "SMPD1          65\n",
       "KMT2E          65\n",
       "KIF5A          64\n",
       "AGXT           64\n",
       "ACADM          64\n",
       "FN1            64\n",
       "TRIO           62\n",
       "FANCA          62\n",
       "ATM            62\n",
       "SETD1B         62\n",
       "ASL            61\n",
       "RAG1           61\n",
       "BTD            61\n",
       "PMM2           60\n",
       "KCNQ5          60\n",
       "PLEC           60\n",
       "CUBN           59\n",
       "DCHS1          59\n",
       "SYNGAP1        59\n",
       "HSPG2          59\n",
       "HUWE1          59\n",
       "DNHD1          59\n",
       "RYR2           59\n",
       "ABCC8          59\n",
       "SLC6A1         59\n",
       "GBA1           58\n",
       "GALC           58\n",
       "ABCA1          58\n",
       "DYRK1A         57\n",
       "GJB2           57\n",
       "COPA           57\n",
       "SCN5A          56\n",
       "LAMA3          55\n",
       "WFS1           55\n",
       "SCN3A          55\n",
       "HNF1A          54\n",
       "IDS            54\n",
       "IDUA           53\n",
       "CHD2           53\n",
       "ETFDH          53\n",
       "CCDC88C        53\n",
       "GABBR2         53\n",
       "KCNH1          52\n",
       "ALPK1          52\n",
       "SETD1A         52\n",
       "HECW2          52\n",
       "FRAS1          51\n",
       "CAMK2B         51\n",
       "LAMA1          51\n",
       "LRP5           51\n",
       "KDM6B          51\n",
       "RANBP2         50\n",
       "ERCC6          50\n",
       "TG             50\n",
       "JAG1           50\n",
       "FOXG1          49\n",
       "ASH1L          49\n",
       "SCN4A          49\n",
       "ASXL2          49\n",
       "IFT140         49\n",
       "AR             48\n",
       "ARSB           48\n",
       "CPLANE1        48\n",
       "BMPR2          48\n",
       "EYS            48\n",
       "RPGR           48\n",
       "KCNH5          48\n",
       "SCN11A         47\n",
       "ASS1           47\n",
       "OPA1           47\n",
       "OTOG           47\n",
       "AGRN           47\n",
       "APC            47\n",
       "SGSH           46\n",
       "SCN2A          46\n",
       "ITGB4          46\n",
       "NOD2           45\n",
       "EPAS1          45\n",
       "FH             44\n",
       "NIPBL          44\n",
       "KCNQ2          44\n",
       "PHKA2          44\n",
       "PHEX           44\n",
       "SLC1A2         44\n",
       "LRRK2          44\n",
       "SLC6A8         43\n",
       "COL3A1         43\n",
       "FOXP1          43\n",
       "SATB2          43\n",
       "CBS            43\n",
       "TNC            43\n",
       "DICER1         41\n",
       "KATNIP         41\n",
       "RHO            41\n",
       "HCN1           40\n",
       "KDM6A          40\n",
       "VPS13D         40\n",
       "TRPM1          40\n",
       "FBXO11         40\n",
       "CYFIP2         40\n",
       "EDA            40\n",
       "PTPN11         40\n",
       "MYH14          40\n",
       "TGM6           40\n",
       "CHD8           40\n",
       "CHAT           39\n",
       "G6PC1          39\n",
       "EIF2B5         39\n",
       "PACS1          39\n",
       "ABCA3          39\n",
       "ACTN2          39\n",
       "CYP11B1        39\n",
       "CYP17A1        39\n",
       "TCF4           39\n",
       "COL4A1         39\n",
       "CFTR           39\n",
       "RET            39\n",
       "FGFR3          38\n",
       "F9             38\n",
       "F5             38\n",
       "NTRK1          38\n",
       "HGD            38\n",
       "DUOX2          38\n",
       "KDM5C          37\n",
       "GUCY2D         37\n",
       "ADA            37\n",
       "UBR1           37\n",
       "MYBPC3         37\n",
       "CLCN7          37\n",
       "PCDH12         37\n",
       "EPG5           37\n",
       "TPP1           36\n",
       "PRPH2          36\n",
       "ACAN           36\n",
       "RB1            36\n",
       "LMNA           36\n",
       "PC             36\n",
       "ADCY5          36\n",
       "AHI1           36\n",
       "AIRE           35\n",
       "DNM1           35\n",
       "PRF1           35\n",
       "PLCG2          35\n",
       "PCCB           35\n",
       "GNE            35\n",
       "SPTA1          35\n",
       "RECQL4         35\n",
       "RDH12          34\n",
       "FREM2          34\n",
       "CC2D2A         34\n",
       "PEX1           34\n",
       "AVPR2          34\n",
       "PLA2G6         34\n",
       "ITGB3          34\n",
       "MSH6           34\n",
       "F11            34\n",
       "SCN9A          34\n",
       "G6PD           34\n",
       "KIF22          34\n",
       "WDFY3          33\n",
       "CASR           33\n",
       "BCKDHB         33\n",
       "JAK3           33\n",
       "CYP21A2        33\n",
       "PANK2          33\n",
       "ALDH7A1        33\n",
       "LPL            33\n",
       "BCKDHA         33\n",
       "CACNA1F        33\n",
       "KCNH2          33\n",
       "CYP27A1        33\n",
       "MYO5B          33\n",
       "GABRB2         33\n",
       "CYBB           32\n",
       "FKRP           32\n",
       "HOGA1          32\n",
       "GLA            32\n",
       "F7             32\n",
       "CDON           32\n",
       "MAN2B1         32\n",
       "CHD3           32\n",
       "SDHB           32\n",
       "SAMD9          32\n",
       "SRD5A2         32\n",
       "TAF1           32\n",
       "ASPA           32\n",
       "IGHMBP2        32\n",
       "FRMPD4         32\n",
       "TPO            31\n",
       "PDE6B          31\n",
       "TERT           31\n",
       "ALK            31\n",
       "TBCD           31\n",
       "CLCNKB         31\n",
       "ADAMTS13       31\n",
       "PROC           31\n",
       "CRB2           31\n",
       "FREM1          31\n",
       "ITGA2B         31\n",
       "MMACHC         31\n",
       "NACC1          31\n",
       "SHANK1         31\n",
       "ACAT1          31\n",
       "SKI            31\n",
       "VPS13B         31\n",
       "ACE            31\n",
       "SGCA           31\n",
       "RAG2           31\n",
       "COL17A1        30\n",
       "OTOGL          30\n",
       "VHL            30\n",
       "HEXA           30\n",
       "SPAST          30\n",
       "COL4A6         30\n",
       "STXBP1         30\n",
       "SOS1           30\n",
       "SLC4A11        30\n",
       "ALDH5A1        30\n",
       "IL2RG          30\n",
       "MTHFR          30\n",
       "OPLAH          29\n",
       "EVC2           29\n",
       "GRM6           29\n",
       "USP9X          29\n",
       "SERPINC1       29\n",
       "MUTYH          29\n",
       "TMPRSS3        29\n",
       "MCCC2          29\n",
       "RUNX1          29\n",
       "SPG11          29\n",
       "PYGM           28\n",
       "LDB3           28\n",
       "RS1            28\n",
       "CDKN2A         28\n",
       "RP1            28\n",
       "TMEM67         28\n",
       "PIK3R2         28\n",
       "RTTN           28\n",
       "SCN8A          28\n",
       "VPS13A         28\n",
       "GRIN1          28\n",
       "SEMA6B         28\n",
       "TRPV4          28\n",
       "IVD            28\n",
       "MVK            28\n",
       "DHX37          28\n",
       "PHIP           27\n",
       "CYP1B1         27\n",
       "CCDC40         27\n",
       "EVC            27\n",
       "FRMD7          27\n",
       "CBL            27\n",
       "BRWD3          27\n",
       "STAT3          27\n",
       "TCF3           27\n",
       "TONSL          27\n",
       "TELO2          27\n",
       "PIEZO2         27\n",
       "HSD17B4        27\n",
       "CHD4           27\n",
       "DOK7           27\n",
       "CNGB1          27\n",
       "DSG2           27\n",
       "ALDH3A2        26\n",
       "ACAD9          26\n",
       "GSN            26\n",
       "HNF4A          26\n",
       "TBC1D24        26\n",
       "NEU1           26\n",
       "SZT2           26\n",
       "MAPK8IP3       26\n",
       "SLC45A2        26\n",
       "KIAA0586       26\n",
       "SERPING1       26\n",
       "SH3TC2         26\n",
       "SNRNP200       26\n",
       "PTEN           26\n",
       "PCCA           26\n",
       "ACVRL1         26\n",
       "TBX5           26\n",
       "HR             26\n",
       "ATP1A2         26\n",
       "COL10A1        26\n",
       "OTC            25\n",
       "WAS            25\n",
       "SLC26A3        25\n",
       "MYOC           25\n",
       "INPP5E         25\n",
       "ATP8B1         25\n",
       "POLE           25\n",
       "GBE1           25\n",
       "ARSL           25\n",
       "AIFM1          25\n",
       "ALG1           25\n",
       "PIGG           25\n",
       "CPT1A          25\n",
       "CTNS           25\n",
       "AARS2          25\n",
       "AARS1          25\n",
       "ABCC2          24\n",
       "ODAD1          24\n",
       "DDC            24\n",
       "PCDH15         24\n",
       "SLC35A2        24\n",
       "KCNV2          24\n",
       "STAG1          24\n",
       "PRKN           24\n",
       "MAST3          24\n",
       "MLC1           24\n",
       "MET            24\n",
       "CDHR1          24\n",
       "GNPTAB         24\n",
       "ACADS          24\n",
       "BRPF1          24\n",
       "CASZ1          24\n",
       "CHRNE          24\n",
       "EZH2           24\n",
       "TRPM4          24\n",
       "MCCC1          24\n",
       "ECHS1          24\n",
       "MTM1           24\n",
       "DNMT3A         24\n",
       "PKLR           24\n",
       "LAMB3          24\n",
       "DNAH10         23\n",
       "NR2E3          23\n",
       "ZFPM2          23\n",
       "KCNJ1          23\n",
       "COQ8A          23\n",
       "ERCC2          23\n",
       "WNT10A         23\n",
       "PRRT2          23\n",
       "PTS            23\n",
       "GABRB3         23\n",
       "ATP2B3         23\n",
       "SLC16A2        23\n",
       "KRT17          23\n",
       "LMF1           23\n",
       "BMPR1B         23\n",
       "CPT2           23\n",
       "DBT            23\n",
       "AMT            23\n",
       "SLC37A4        23\n",
       "GREB1L         23\n",
       "DNAH3          22\n",
       "HBB            22\n",
       "PLD1           22\n",
       "ROR2           22\n",
       "WDR45          22\n",
       "UMOD           22\n",
       "ANXA11         22\n",
       "TECTA          22\n",
       "GLRA1          22\n",
       "ATR            22\n",
       "DNAAF5         22\n",
       "NLRP12         22\n",
       "ADAMTSL4       22\n",
       "ZNF341         22\n",
       "CARD14         22\n",
       "CYP27B1        22\n",
       "IL12RB1        22\n",
       "GAMT           22\n",
       "FGFR1          22\n",
       "FMO3           22\n",
       "IFT172         22\n",
       "SLC6A5         22\n",
       "SMARCAL1       22\n",
       "CAPN5          22\n",
       "SI             22\n",
       "PEX6           22\n",
       "RP2            22\n",
       "LAMB2          22\n",
       "CUL7           21\n",
       "PCARE          21\n",
       "DCTN1          21\n",
       "SOX9           21\n",
       "FOXP3          21\n",
       "BLM            21\n",
       "TH             21\n",
       "ARX            21\n",
       "DYNC2I1        21\n",
       "SLC4A1         21\n",
       "DOCK2          21\n",
       "LIPA           21\n",
       "EXT2           21\n",
       "BRAT1          21\n",
       "SLC2A1         21\n",
       "SLC38A8        21\n",
       "PNPLA6         21\n",
       "CYP11B2        21\n",
       "CNGB3          21\n",
       "CEP164         21\n",
       "MYH3           21\n",
       "PPT1           21\n",
       "HEXB           21\n",
       "ANLN           21\n",
       "HCN4           21\n",
       "GFM1           21\n",
       "ZBTB20         21\n",
       "CTC1           20\n",
       "PCSK9          20\n",
       "MCPH1          20\n",
       "AQP2           20\n",
       "NPHP4          20\n",
       "ODAD2          20\n",
       "FARS2          20\n",
       "LRBA           20\n",
       "FANCD2         20\n",
       "KIF1C          20\n",
       "TSHR           20\n",
       "ANO5           20\n",
       "TTR            20\n",
       "GNAO1          20\n",
       "TEK            20\n",
       "WDR62          20\n",
       "TNNI3K         20\n",
       "TNNI3          20\n",
       "DSPP           20\n",
       "LRPPRC         20\n",
       "TULP1          20\n",
       "GJB1           20\n",
       "RTEL1          20\n",
       "MID1           20\n",
       "MFN2           20\n",
       "ADSL           20\n",
       "SPG7           20\n",
       "BCS1L          20\n",
       "KIAA0753       20\n",
       "SOD1           20\n",
       "APP            19\n",
       "MERTK          19\n",
       "OAT            19\n",
       "ARMC9          19\n",
       "DSCAM          19\n",
       "ROBO3          19\n",
       "DNAI1          19\n",
       "CYP24A1        19\n",
       "LRRC56         19\n",
       "LSS            19\n",
       "PNPLA1         19\n",
       "FLT4           19\n",
       "ANKLE2         19\n",
       "TRDN           19\n",
       "GRHPR          19\n",
       "GANAB          19\n",
       "IFITM5         19\n",
       "GDAP1          19\n",
       "TTC7A          19\n",
       "TK2            19\n",
       "ENPP1          19\n",
       "KRT16          19\n",
       "FANCI          19\n",
       "SLC3A1         19\n",
       "SLCO2A1        19\n",
       "ADA2           19\n",
       "NPHS2          19\n",
       "TYMP           19\n",
       "KCNQ3          19\n",
       "CLPB           19\n",
       "CRX            19\n",
       "HLCS           19\n",
       "EIF2AK4        19\n",
       "CFI            18\n",
       "CYP4V2         18\n",
       "ARHGEF18       18\n",
       "TET2           18\n",
       "GNAS           18\n",
       "ABCD4          18\n",
       "TRIP12         18\n",
       "DCLRE1C        18\n",
       "GATA1          18\n",
       "PDHA1          18\n",
       "ADGRG1         18\n",
       "ATN1           18\n",
       "SLC39A4        18\n",
       "GATA2          18\n",
       "TCIRG1         18\n",
       "SH3BP2         18\n",
       "SH3PXD2B       18\n",
       "CDAN1          18\n",
       "HDAC8          18\n",
       "LOXHD1         18\n",
       "ABCG8          18\n",
       "SUMF1          18\n",
       "SEC24D         18\n",
       "CP             18\n",
       "CARMIL2        18\n",
       "FAH            18\n",
       "MPL            18\n",
       "GFAP           18\n",
       "SOX10          18\n",
       "CDT1           18\n",
       "DNM2           18\n",
       "GEMIN5         18\n",
       "ABCB4          17\n",
       "CFH            17\n",
       "CEP152         17\n",
       "XPC            17\n",
       "ANKRD17        17\n",
       "LARS2          17\n",
       "POLR1A         17\n",
       "IL7R           17\n",
       "ITGB2          17\n",
       "IRF2BPL        17\n",
       "ASAH1          17\n",
       "STAR           17\n",
       "ITPR1          17\n",
       "NR3C2          17\n",
       "CTSK           17\n",
       "CYP4F22        17\n",
       "PCDH19         17\n",
       "DEAF1          17\n",
       "PALB2          17\n",
       "RBM10          17\n",
       "POMGNT1        17\n",
       "PPP2R1A        17\n",
       "POMT1          17\n",
       "RARS2          17\n",
       "MTR            17\n",
       "HMBS           17\n",
       "EDAR           17\n",
       "TNFRSF1A       17\n",
       "SLC5A6         17\n",
       "SLC2A10        17\n",
       "SLC34A3        17\n",
       "SERPINA1       17\n",
       "FGFR2          17\n",
       "CANT1          17\n",
       "SELENON        17\n",
       "PIGQ           17\n",
       "BRF1           17\n",
       "SHROOM2        17\n",
       "LIG4           17\n",
       "PEPD           17\n",
       "JAK2           16\n",
       "CSPP1          16\n",
       "AFG2A          16\n",
       "CEP290         16\n",
       "MUSK           16\n",
       "SLC7A9         16\n",
       "BBS10          16\n",
       "SPINK5         16\n",
       "SPTB           16\n",
       "MSTO1          16\n",
       "CLCN2          16\n",
       "FAM20C         16\n",
       "SLC40A1        16\n",
       "SYNJ1          16\n",
       "PKP2           16\n",
       "KIZ            16\n",
       "RPGRIP1        16\n",
       "PKD2           16\n",
       "PDE6A          16\n",
       "MYT1L          16\n",
       "BRIP1          16\n",
       "SKIC3          16\n",
       "GARS1          16\n",
       "TNNT2          16\n",
       "HSD3B2         16\n",
       "TJP2           16\n",
       "GNRHR          16\n",
       "TSEN54         15\n",
       "TTBK2          15\n",
       "SLC12A1        15\n",
       "GPR143         15\n",
       "WDR81          15\n",
       "XYLT1          15\n",
       "DARS2          15\n",
       "GCM2           15\n",
       "AMH            15\n",
       "DHTKD1         15\n",
       "BRAF           15\n",
       "ALAS2          15\n",
       "TBXAS1         15\n",
       "TEX14          15\n",
       "TGFBR2         15\n",
       "TMC1           15\n",
       "ALOX12B        15\n",
       "HSD17B3        15\n",
       "BMP1           15\n",
       "SLC29A3        15\n",
       "FOXN1          15\n",
       "KCNJ11         15\n",
       "SOX6           15\n",
       "RAPSN          15\n",
       "SLC6A19        15\n",
       "SMC1A          15\n",
       "CHST6          15\n",
       "FA2H           15\n",
       "KRT10          15\n",
       "KRT5           15\n",
       "LCAT           15\n",
       "LZTR1          15\n",
       "RNF216         15\n",
       "PNKD           15\n",
       "IL10RA         15\n",
       "MBTPS1         15\n",
       "EPRS1          15\n",
       "CEP120         15\n",
       "NMNAT1         15\n",
       "ASNS           15\n",
       "IGF1R          15\n",
       "GUSB           15\n",
       "HADHA          15\n",
       "HBA1           15\n",
       "WDR19          15\n",
       "MIPEP          15\n",
       "ELANE          15\n",
       "EMC1           15\n",
       "MOCOS          15\n",
       "MSN            15\n",
       "MAPT           15\n",
       "CPA1           14\n",
       "SERAC1         14\n",
       "SLC12A2        14\n",
       "SLC19A3        14\n",
       "ZMYND11        14\n",
       "GDF2           14\n",
       "GNB1           14\n",
       "WDR37          14\n",
       "WWOX           14\n",
       "TRMU           14\n",
       "AGL            14\n",
       "PCYT1A         14\n",
       "TRAPPC12       14\n",
       "GATA4          14\n",
       "TTLL5          14\n",
       "PDE6C          14\n",
       "CLCN4          14\n",
       "PYGL           14\n",
       "SLC9A3         14\n",
       "BBS2           14\n",
       "CR2            14\n",
       "ATP13A2        14\n",
       "NDUFV1         14\n",
       "LRP6           14\n",
       "CHRNB2         14\n",
       "PNPT1          14\n",
       "RPGRIP1L       14\n",
       "SCNN1B         14\n",
       "LEPR           14\n",
       "KRT2           14\n",
       "PLOD1          14\n",
       "PRKAG2         14\n",
       "MEN1           14\n",
       "HGSNAT         14\n",
       "GYS2           14\n",
       "MORC2          14\n",
       "MYH11          14\n",
       "EBF3           14\n",
       "HADHB          14\n",
       "RFX6           14\n",
       "RDH5           14\n",
       "CFAP410        14\n",
       "EXT1           14\n",
       "STS            14\n",
       "ORC1           14\n",
       "CTR9           14\n",
       "UNC45B         14\n",
       "ACSF3          13\n",
       "NBEAL2         13\n",
       "TGFBI          13\n",
       "TFR2           13\n",
       "THRB           13\n",
       "NEK1           13\n",
       "STIM1          13\n",
       "HBA2           13\n",
       "TGM5           13\n",
       "GHR            13\n",
       "TMEM231        13\n",
       "TMEM43         13\n",
       "GJA8           13\n",
       "MAT1A          13\n",
       "MBTPS2         13\n",
       "BBS12          13\n",
       "SMARCC2        13\n",
       "ATP6V0A4       13\n",
       "MEF2C          13\n",
       "GMPPB          13\n",
       "HJV            13\n",
       "CAPN15         13\n",
       "SEC23B         13\n",
       "CASK           13\n",
       "KARS1          13\n",
       "MAN1B1         13\n",
       "PIGN           13\n",
       "CYP11A1        13\n",
       "KIF12          13\n",
       "KCNQ4          13\n",
       "FUCA1          13\n",
       "PSEN1          13\n",
       "PRSS1          13\n",
       "RFT1           13\n",
       "KRT1           13\n",
       "CLDN16         13\n",
       "KLHL40         13\n",
       "ADAMTS10       13\n",
       "WRN            13\n",
       "TTC21B         13\n",
       "ZMYND10        13\n",
       "GPI            13\n",
       "GP1BA          13\n",
       "POMT2          13\n",
       "SLC34A1        13\n",
       "SLC26A2        13\n",
       "COQ8B          13\n",
       "COMP           13\n",
       "RHBDF2         13\n",
       "JPH2           13\n",
       "SLC25A13       13\n",
       "BTK            13\n",
       "SIN3A          13\n",
       "GRHL3          13\n",
       "TRAF7          13\n",
       "MPI            13\n",
       "STXBP2         12\n",
       "OPHN1          12\n",
       "TRAPPC11       12\n",
       "HARS1          12\n",
       "SPRED1         12\n",
       "SOX11          12\n",
       "GBA2           12\n",
       "GABRG2         12\n",
       "BARD1          12\n",
       "NUP214         12\n",
       "MATN3          12\n",
       "NLGN4X         12\n",
       "SOX17          12\n",
       "CYP26B1        12\n",
       "MKKS           12\n",
       "EARS2          12\n",
       "NPHP3          12\n",
       "MLPH           12\n",
       "DLL3           12\n",
       "TP63           12\n",
       "ARHGEF9        12\n",
       "TBCK           12\n",
       "TCF12          12\n",
       "HPS5           12\n",
       "GJA3           12\n",
       "TBK1           12\n",
       "IARS2          12\n",
       "DDX11          12\n",
       "NR5A1          12\n",
       "NGLY1          12\n",
       "MMAB           12\n",
       "ECEL1          12\n",
       "NCF2           12\n",
       "DNAI2          12\n",
       "DONSON         12\n",
       "MTRR           12\n",
       "VARS2          12\n",
       "TRPV3          12\n",
       "AGA            12\n",
       "AFG3L2         12\n",
       "MOV10L1        12\n",
       "ELAC2          12\n",
       "CDC14A         12\n",
       "SBF1           12\n",
       "CDH2           12\n",
       "ABCG5          12\n",
       "FBXO7          12\n",
       "PTCHD1         12\n",
       "CLN3           12\n",
       "RGS9           12\n",
       "KRT6C          12\n",
       "UROS           12\n",
       "LARS1          12\n",
       "SLC20A2        12\n",
       "KCND2          12\n",
       "EPHA2          12\n",
       "ERCC5          12\n",
       "PGAP3          12\n",
       "COQ4           12\n",
       "LETM1          12\n",
       "CSF3R          12\n",
       "PLG            12\n",
       "PIGT           12\n",
       "CTSC           12\n",
       "LRSAM1         12\n",
       "P2RX2          12\n",
       "CTCF           12\n",
       "GALNT3         12\n",
       "AMHR2          12\n",
       "TMPRSS6        12\n",
       "STK11          12\n",
       "GABRA1         12\n",
       "CLCN5          12\n",
       "KRT14          12\n",
       "FOXE3          12\n",
       "PRNP           12\n",
       "TWNK           12\n",
       "PAX1           12\n",
       "MC2R           11\n",
       "MCM9           11\n",
       "SQSTM1         11\n",
       "HPS1           11\n",
       "TBCE           11\n",
       "NLRP3          11\n",
       "GHRHR          11\n",
       "TLE6           11\n",
       "HPS3           11\n",
       "STING1         11\n",
       "GDF5           11\n",
       "DES            11\n",
       "ABCB6          11\n",
       "WDR4           11\n",
       "DPYS           11\n",
       "MYOT           11\n",
       "MTTP           11\n",
       "DNMT3B         11\n",
       "EEF1A2         11\n",
       "TBX4           11\n",
       "TET3           11\n",
       "NPRL3          11\n",
       "ATP2B2         11\n",
       "ATP1A1         11\n",
       "SOX5           11\n",
       "SMARCA4        11\n",
       "SLC9A6         11\n",
       "NUP93          11\n",
       "DDX41          11\n",
       "OTOA           11\n",
       "ETHE1          11\n",
       "PROS1          11\n",
       "PPP1R15B       11\n",
       "PPA2           11\n",
       "EYA1           11\n",
       "PREPL          11\n",
       "POLR1C         11\n",
       "POLR3A         11\n",
       "LBR            11\n",
       "FAS            11\n",
       "CABP4          11\n",
       "POC1B          11\n",
       "SLC25A15       11\n",
       "PIDD1          11\n",
       "BRSK2          11\n",
       "LMX1B          11\n",
       "PDZD7          11\n",
       "MAK            11\n",
       "CSNK2A1        11\n",
       "PIGO           11\n",
       "ITGA6          11\n",
       "F2             11\n",
       "FANCC          11\n",
       "COLQ           11\n",
       "FECH           11\n",
       "RSPH1          11\n",
       "CD40LG         11\n",
       "CCNO           11\n",
       "CHCHD10        11\n",
       "CEP78          11\n",
       "UROD           11\n",
       "ZBTB18         11\n",
       "CEACAM16       11\n",
       "SDHA           11\n",
       "CLN6           11\n",
       "CLN8           11\n",
       "QDPR           11\n",
       "PYCR1          11\n",
       "PSTPIP1        11\n",
       "FBP1           11\n",
       "AIPL1          11\n",
       "ALDH18A1       11\n",
       "MPZ            11\n",
       "MTFMT          11\n",
       "MTO1           11\n",
       "ELP1           11\n",
       "MOGS           11\n",
       "EIF2B2         11\n",
       "TUBB1          11\n",
       "TSPEAR         11\n",
       "GPAA1          11\n",
       "TREM2          11\n",
       "TUBB3          11\n",
       "GTPBP3         11\n",
       "ACAD8          11\n",
       "UGT1A1         11\n",
       "AASS           10\n",
       "UNC13D         10\n",
       "TRMT1          10\n",
       "USH1C          10\n",
       "ADD3           10\n",
       "ACTN4          10\n",
       "ZCCHC8         10\n",
       "ADAMTSL2       10\n",
       "MYCN           10\n",
       "ALDOB          10\n",
       "PSAP           10\n",
       "CHST14         10\n",
       "CKAP2L         10\n",
       "PRSS56         10\n",
       "KRAS           10\n",
       "CHST3          10\n",
       "FLCN           10\n",
       "VPS11          10\n",
       "ILDR1          10\n",
       "IGFALS         10\n",
       "PADI6          10\n",
       "AVP            10\n",
       "PAX3           10\n",
       "CXCR2          10\n",
       "LMOD3          10\n",
       "LPIN2          10\n",
       "ITGA3          10\n",
       "LIFR           10\n",
       "PNKP           10\n",
       "FOXRED1        10\n",
       "SLC25A38       10\n",
       "SLC2A2         10\n",
       "PNPLA2         10\n",
       "PNP            10\n",
       "LHCGR          10\n",
       "CHD5           10\n",
       "RBP3           10\n",
       "RIT1           10\n",
       "FUS            10\n",
       "RPS6KA3        10\n",
       "KCNK4          10\n",
       "CENPJ          10\n",
       "ZNF408         10\n",
       "STRA6          10\n",
       "AP3D1          10\n",
       "HMGCL          10\n",
       "ALG6           10\n",
       "TF             10\n",
       "ALS2           10\n",
       "WDR35          10\n",
       "DYNC2I2        10\n",
       "DYNC2LI1       10\n",
       "AAAS           10\n",
       "MYO1A          10\n",
       "NARS1          10\n",
       "NDUFAF5        10\n",
       "DLD            10\n",
       "TFE3           10\n",
       "NPR2           10\n",
       "TMTC3          10\n",
       "TCF7L2         10\n",
       "NLRP7          10\n",
       "HSPB1          10\n",
       "HTRA1          10\n",
       "AQP5           10\n",
       "ARG1           10\n",
       "CAD            10\n",
       "PIK3CA         10\n",
       "PIGB           10\n",
       "MAGED2         10\n",
       "INPP5K         10\n",
       "CCN6           10\n",
       "BLTP1          10\n",
       "SLC5A1         10\n",
       "OSGEP          10\n",
       "MC4R           10\n",
       "MEFV           10\n",
       "NUP107         10\n",
       "CYP7B1         10\n",
       "PUM1            9\n",
       "CLCC1           9\n",
       "FHOD3           9\n",
       "PROP1           9\n",
       "PRUNE1          9\n",
       "PTPRC           9\n",
       "ACP5            9\n",
       "CNGA1           9\n",
       "ALG12           9\n",
       "AFF3            9\n",
       "ADAR            9\n",
       "EIF2B4          9\n",
       "WARS2           9\n",
       "GUCY2C          9\n",
       "TYRP1           9\n",
       "GRN             9\n",
       "GPD1            9\n",
       "GRIA3           9\n",
       "TRIT1           9\n",
       "WT1             9\n",
       "EIF2AK3         9\n",
       "MPV17           9\n",
       "MTHFD1          9\n",
       "AICDA           9\n",
       "FLVCR1          9\n",
       "COG4            9\n",
       "KCNJ2           9\n",
       "SMAD6           9\n",
       "GALK1           9\n",
       "GATM            9\n",
       "IFT74           9\n",
       "ATP13A3         9\n",
       "ATL1            9\n",
       "MAF             9\n",
       "ATP6V1B1        9\n",
       "SFTPB           9\n",
       "SLC17A5         9\n",
       "SKIC2           9\n",
       "SLC19A2         9\n",
       "PIK3CD          9\n",
       "ERCC4           9\n",
       "ERF             9\n",
       "ESCO2           9\n",
       "INSR            9\n",
       "POLR2A          9\n",
       "RMND1           9\n",
       "CERKL           9\n",
       "CERS3           9\n",
       "FGA             9\n",
       "FZD4            9\n",
       "RTN4IP1         9\n",
       "POLH            9\n",
       "PNPO            9\n",
       "FAM20A          9\n",
       "L2HGDH          9\n",
       "COQ2            9\n",
       "SLC34A2         9\n",
       "KCNC1           9\n",
       "BCHE            9\n",
       "GLDN            9\n",
       "HSD11B2         9\n",
       "ARSG            9\n",
       "ARMC5           9\n",
       "NRAS            9\n",
       "NPHP1           9\n",
       "TBX1            9\n",
       "TARDBP          9\n",
       "ETV6            9\n",
       "BBS7            9\n",
       "BAP1            9\n",
       "SLC5A5          9\n",
       "NSMCE3          9\n",
       "NYX             9\n",
       "CYP2U1          9\n",
       "MAP3K7          9\n",
       "SLC52A2         9\n",
       "TGFBR1          9\n",
       "GGCX            9\n",
       "ALOXE3          9\n",
       "SUOX            9\n",
       "SUCLA2          9\n",
       "TNFRSF11A       9\n",
       "ANO3            9\n",
       "GLUD1           9\n",
       "TMEM126B        9\n",
       "DGUOK           9\n",
       "NHLRC1          9\n",
       "MME             9\n",
       "EBP             9\n",
       "MYH8            9\n",
       "DPH1            9\n",
       "WNT5A           8\n",
       "GPIHBP1         8\n",
       "GRIA4           8\n",
       "GSR             8\n",
       "GSS             8\n",
       "F13A1           8\n",
       "ZBTB24          8\n",
       "TRPC6           8\n",
       "KLHL41          8\n",
       "GNPTG           8\n",
       "ROS1            8\n",
       "RNASEL          8\n",
       "PINK1           8\n",
       "MMP2            8\n",
       "MMADHC          8\n",
       "NDUFA9          8\n",
       "TMEM260         8\n",
       "DYRK1B          8\n",
       "DPYD            8\n",
       "NBN             8\n",
       "NARS2           8\n",
       "B3GALT6         8\n",
       "EFEMP2          8\n",
       "MEI1            8\n",
       "NCKAP1L         8\n",
       "KRT9            8\n",
       "RARS1           8\n",
       "PUS1            8\n",
       "PUF60           8\n",
       "PEX7            8\n",
       "APRT            8\n",
       "ARL6            8\n",
       "PGAP2           8\n",
       "RLBP1           8\n",
       "RIPK4           8\n",
       "RIC1            8\n",
       "CEP83           8\n",
       "GJC2            8\n",
       "SPTLC2          8\n",
       "SPTLC1          8\n",
       "HPS6            8\n",
       "NIPAL4          8\n",
       "SPATA7          8\n",
       "SEPTIN9         8\n",
       "GNB4            8\n",
       "HK1             8\n",
       "AK2             8\n",
       "AGBL5           8\n",
       "PMS2            8\n",
       "ELP2            8\n",
       "VARS1           8\n",
       "VCP             8\n",
       "EGFR            8\n",
       "PMS1            8\n",
       "POMGNT2         8\n",
       "POLR3B          8\n",
       "PRMT7           8\n",
       "PMPCB           8\n",
       "YARS1           8\n",
       "AFG2B           8\n",
       "SCNN1A          8\n",
       "DDX3X           8\n",
       "FDXR            8\n",
       "CYB5R3          8\n",
       "CD55            8\n",
       "CYBA            8\n",
       "CYP19A1         8\n",
       "MED17           8\n",
       "CUL3            8\n",
       "C7              8\n",
       "KCNA2           8\n",
       "IYD             8\n",
       "SLC39A8         8\n",
       "MCOLN1          8\n",
       "B4GALNT1        8\n",
       "DBR1            8\n",
       "CA5A            8\n",
       "BBS1            8\n",
       "LRP4            8\n",
       "SPINT2          8\n",
       "ATP1A3          8\n",
       "RSPH4A          8\n",
       "KCNMA1          8\n",
       "LOX             8\n",
       "LMNB1           8\n",
       "RXYLT1          8\n",
       "RUNX2           8\n",
       "IL11RA          8\n",
       "GALE            8\n",
       "MANBA           8\n",
       "DIS3L2          8\n",
       "EXOSC3          8\n",
       "CHRNA1          8\n",
       "SEPSECS         8\n",
       "SDR9C7          8\n",
       "CBLIF           8\n",
       "CASQ2           8\n",
       "TUBB2A          8\n",
       "MRE11           8\n",
       "ACTN1           8\n",
       "CCDC22          8\n",
       "CAPN1           8\n",
       "SAMHD1          8\n",
       "CDH3            8\n",
       "CHEK2           7\n",
       "PFKM            7\n",
       "ARL13B          7\n",
       "SYN1            7\n",
       "DPAGT1          7\n",
       "TFRC            7\n",
       "TFG             7\n",
       "CRPPA           7\n",
       "CPOX            7\n",
       "RNASEH2B        7\n",
       "ESX1            7\n",
       "CRELD1          7\n",
       "APOE            7\n",
       "GJB3            7\n",
       "SURF1           7\n",
       "STUB1           7\n",
       "GJB6            7\n",
       "STRC            7\n",
       "MTPAP           7\n",
       "MSR1            7\n",
       "RAB3GAP2        7\n",
       "TSPAN12         7\n",
       "GNAI1           7\n",
       "GRM7            7\n",
       "GRIK2           7\n",
       "XPA             7\n",
       "MOCS1           7\n",
       "TSFM            7\n",
       "TTI1            7\n",
       "F12             7\n",
       "LARGE1          7\n",
       "FANCE           7\n",
       "FGG             7\n",
       "FLAD1           7\n",
       "WASHC5          7\n",
       "WDR26           7\n",
       "ACSL4           7\n",
       "RAB27A          7\n",
       "KIF5C           7\n",
       "TAT             7\n",
       "MFSD8           7\n",
       "PYROXD1         7\n",
       "QRSL1           7\n",
       "PIBF1           7\n",
       "RAD51C          7\n",
       "PUS3            7\n",
       "KRT12           7\n",
       "MTMR2           7\n",
       "MMAA            7\n",
       "TGFB2           7\n",
       "PAX2            7\n",
       "GATA6           7\n",
       "ERCC8           7\n",
       "CDKN1C          7\n",
       "BGN             7\n",
       "C1QBP           7\n",
       "C3              7\n",
       "SLC2A9          7\n",
       "CA4             7\n",
       "SLC45A1         7\n",
       "SLC27A4         7\n",
       "D2HGDH          7\n",
       "CTSF            7\n",
       "SLC1A4          7\n",
       "FBXL4           7\n",
       "KIF21A          7\n",
       "SLC25A20        7\n",
       "LITAF           7\n",
       "CHRND           7\n",
       "IFT80           7\n",
       "IFT52           7\n",
       "MARS1           7\n",
       "MAP2K2          7\n",
       "SLC7A7          7\n",
       "MAOA            7\n",
       "SMAD3           7\n",
       "DKC1            7\n",
       "PGM1            7\n",
       "NKX2-5          7\n",
       "GATAD2B         7\n",
       "GCH1            7\n",
       "CASQ1           7\n",
       "SGCB            7\n",
       "CARS1           7\n",
       "CCND2           7\n",
       "TUBB4A          7\n",
       "UBA5            7\n",
       "UBE3B           7\n",
       "SLC12A6         7\n",
       "SHH             7\n",
       "SGCG            7\n",
       "CCNF            7\n",
       "VAC14           7\n",
       "SLC11A2         7\n",
       "SDHC            7\n",
       "CDC45           7\n",
       "CUL4B           7\n",
       "POLG2           7\n",
       "POR             7\n",
       "POC1A           7\n",
       "POGLUT1         7\n",
       "HPDL            7\n",
       "NALCN           7\n",
       "NAGS            7\n",
       "NADSYN1         7\n",
       "AMN             7\n",
       "PLPBP           7\n",
       "PLOD2           7\n",
       "HIBCH           7\n",
       "TP53RK          7\n",
       "AHCY            7\n",
       "TRAPPC9         7\n",
       "ALDH1A3         7\n",
       "EIF2B3          7\n",
       "ANO10           7\n",
       "TPM1            7\n",
       "EFNB1           7\n",
       "VDR             6\n",
       "AIMP1           6\n",
       "HOXD13          6\n",
       "GH1             6\n",
       "HMGCS2          6\n",
       "HMMR            6\n",
       "HARS2           6\n",
       "HCCS            6\n",
       "HEPACAM         6\n",
       "ALG8            6\n",
       "ALX4            6\n",
       "PMP22           6\n",
       "PROM1           6\n",
       "PRICKLE1        6\n",
       "LFNG            6\n",
       "CNTNAP1         6\n",
       "COASY           6\n",
       "ACTG1           6\n",
       "TUBB2B          6\n",
       "GNPAT           6\n",
       "YARS2           6\n",
       "SLC13A5         6\n",
       "SGPL1           6\n",
       "SHOX            6\n",
       "CCBE1           6\n",
       "SDHD            6\n",
       "SCARF2          6\n",
       "SCN1B           6\n",
       "SATB1           6\n",
       "CDH11           6\n",
       "DNAAF3          6\n",
       "NKX2-1          6\n",
       "BBS4            6\n",
       "BIN1            6\n",
       "SLC39A7         6\n",
       "B4GAT1          6\n",
       "SLC25A46        6\n",
       "DDHD2           6\n",
       "OXCT1           6\n",
       "MECR            6\n",
       "SLC22A12        6\n",
       "BSND            6\n",
       "CTNND1          6\n",
       "CTSA            6\n",
       "PAPSS2          6\n",
       "P2RY12          6\n",
       "PADI3           6\n",
       "KCNJ5           6\n",
       "G6PC3           6\n",
       "GABRA5          6\n",
       "RRAS2           6\n",
       "MAFB            6\n",
       "EPM2A           6\n",
       "LPAR6           6\n",
       "SPOUT1          6\n",
       "ATF6            6\n",
       "SMO             6\n",
       "SMS             6\n",
       "SMCHD1          6\n",
       "IFT27           6\n",
       "IFT56           6\n",
       "BMP15           6\n",
       "SLC5A2          6\n",
       "SLC6A3          6\n",
       "NR4A2           6\n",
       "NR2F1           6\n",
       "DLG4            6\n",
       "LRAT            6\n",
       "PGM3            6\n",
       "CFAP300         6\n",
       "RNASEH2A        6\n",
       "RCBTB1          6\n",
       "REEP6           6\n",
       "CRADD           6\n",
       "CRYAA           6\n",
       "CRYGD           6\n",
       "COQ6            6\n",
       "COX10           6\n",
       "PLCB3           6\n",
       "PHF8            6\n",
       "APTX            6\n",
       "IDH1            6\n",
       "HPGD            6\n",
       "GLE1            6\n",
       "STAMBP          6\n",
       "B3GAT3          6\n",
       "DHODH           6\n",
       "NF2             6\n",
       "NFIB            6\n",
       "NEFL            6\n",
       "NEK10           6\n",
       "ATP2A2          6\n",
       "ATP5F1A         6\n",
       "ATP6V0A2        6\n",
       "MYO6            6\n",
       "DOLK            6\n",
       "MYCBP2          6\n",
       "TIMP3           6\n",
       "TLK2            6\n",
       "TGFB3           6\n",
       "THOC6           6\n",
       "MEGF10          6\n",
       "MFSD2A          6\n",
       "TAFAZZIN        6\n",
       "TALDO1          6\n",
       "TANGO2          6\n",
       "PHGDH           6\n",
       "PHYH            6\n",
       "QARS1           6\n",
       "KIT             6\n",
       "KLF1            6\n",
       "FLII            6\n",
       "CHRNG           6\n",
       "FGD4            6\n",
       "PTH1R           6\n",
       "FARSB           6\n",
       "LARP7           6\n",
       "FANCG           6\n",
       "MSH5            6\n",
       "WNT10B          6\n",
       "GPAT2           6\n",
       "ACADSB          6\n",
       "GNB2            5\n",
       "GYG1            5\n",
       "ZP1             5\n",
       "ABCC9           5\n",
       "ABAT            5\n",
       "ZAP70           5\n",
       "UBQLN2          5\n",
       "USH1G           5\n",
       "VAPB            5\n",
       "ACO2            5\n",
       "JAG2            5\n",
       "KCND3           5\n",
       "IQSEC1          5\n",
       "IRF6            5\n",
       "DDR2            5\n",
       "ITPA            5\n",
       "SIX3            5\n",
       "SLC24A5         5\n",
       "OPN1LW          5\n",
       "OPTN            5\n",
       "DARS1           5\n",
       "SIGMAR1         5\n",
       "FKBP10          5\n",
       "FGF3            5\n",
       "IMPG2           5\n",
       "HNRNPK          5\n",
       "PSMB8           5\n",
       "GFPT1           5\n",
       "GHSR            5\n",
       "ALG9            5\n",
       "TNFRSF13B       5\n",
       "HFE             5\n",
       "HINT1           5\n",
       "PPOX            5\n",
       "PROKR2          5\n",
       "TECRL           5\n",
       "MINPP1          5\n",
       "MITF            5\n",
       "PDHB            5\n",
       "CTLA4           5\n",
       "MLH1            5\n",
       "MMP13           5\n",
       "NCF4            5\n",
       "NDP             5\n",
       "NDST1           5\n",
       "MEOX1           5\n",
       "ECM1            5\n",
       "MFRP            5\n",
       "TCTN1           5\n",
       "TCTN2           5\n",
       "TDP1            5\n",
       "TBX19           5\n",
       "IBA57           5\n",
       "C19orf12        5\n",
       "MDH2            5\n",
       "MED25           5\n",
       "RAD51           5\n",
       "RAD51D          5\n",
       "RALA            5\n",
       "SLC46A1         5\n",
       "NR3C1           5\n",
       "SORD            5\n",
       "SPOP            5\n",
       "IMPDH1          5\n",
       "DNAAF11         5\n",
       "SLC26A5         5\n",
       "IFNGR1          5\n",
       "GAD1            5\n",
       "B4GALT7         5\n",
       "SNX27           5\n",
       "IDH2            5\n",
       "APOA1           5\n",
       "GK              5\n",
       "GLRA2           5\n",
       "HRAS            5\n",
       "TAF6            5\n",
       "HSD3B7          5\n",
       "ST3GAL5         5\n",
       "STAC3           5\n",
       "STAT2           5\n",
       "STX11           5\n",
       "SUCLG1          5\n",
       "PAK3            5\n",
       "PARK7           5\n",
       "PAX5            5\n",
       "PAX6            5\n",
       "PAX8            5\n",
       "CCT5            5\n",
       "CCDC115         5\n",
       "CELA2A          5\n",
       "RLIM            5\n",
       "RNASEH1         5\n",
       "PEX10           5\n",
       "PEX16           5\n",
       "THSD1           5\n",
       "DNAJB6          5\n",
       "TMEM240         5\n",
       "MVD             5\n",
       "ALG3            5\n",
       "CRYBB1          5\n",
       "CRYGC           5\n",
       "TMEM199         5\n",
       "TGDS            5\n",
       "CLRN1           5\n",
       "PIGL            5\n",
       "PIGV            5\n",
       "PLCD1           5\n",
       "RAD21           5\n",
       "ACY1            5\n",
       "MMP21           5\n",
       "MMP9            5\n",
       "XRCC1           5\n",
       "WDR73           5\n",
       "TPCN2           5\n",
       "TPK1            5\n",
       "VPS41           5\n",
       "AGPAT2          5\n",
       "ACTG2           5\n",
       "TRIP13          5\n",
       "TRNT1           5\n",
       "TSHZ3           5\n",
       "ACTA1           5\n",
       "MRPL50          5\n",
       "MRPS22          5\n",
       "MSH3            5\n",
       "COCH            5\n",
       "COLEC11         5\n",
       "POU1F1          5\n",
       "CSF2RA          5\n",
       "PLP1            5\n",
       "PMPCA           5\n",
       "DGAT1           5\n",
       "PPIB            5\n",
       "CLDN19          5\n",
       "PRMT9           5\n",
       "NEK9            5\n",
       "IFT81           5\n",
       "SMAD4           5\n",
       "IDH3A           5\n",
       "GP9             5\n",
       "GPC6            5\n",
       "DEGS1           5\n",
       "NEK2            5\n",
       "B3GLCT          5\n",
       "WBP2            4\n",
       "ATP2B1          4\n",
       "AUH             4\n",
       "NFU1            4\n",
       "NGF             4\n",
       "NHLRC2          4\n",
       "SLC4A4          4\n",
       "KPTN            4\n",
       "KRT86           4\n",
       "CLN5            4\n",
       "FIG4            4\n",
       "POU3F4          4\n",
       "PPARG           4\n",
       "UQCRC2          4\n",
       "ACTA2           4\n",
       "MRPS34          4\n",
       "ALG11           4\n",
       "ACTB            4\n",
       "VSX2            4\n",
       "LGI1            4\n",
       "TRIM32          4\n",
       "MRPL44          4\n",
       "AKR1D1          4\n",
       "TNNT3           4\n",
       "VPS45           4\n",
       "ADAT3           4\n",
       "ADPRS           4\n",
       "ADSS1           4\n",
       "PHKG2           4\n",
       "PJVK            4\n",
       "MYORG           4\n",
       "NAGA            4\n",
       "RBBP8           4\n",
       "TGFB1           4\n",
       "DPM1            4\n",
       "TMEM70          4\n",
       "REEP1           4\n",
       "RHAG            4\n",
       "MYL2            4\n",
       "PEX12           4\n",
       "PEX5            4\n",
       "CHMP2B          4\n",
       "RRM2B           4\n",
       "KCNJ10          4\n",
       "KCTD7           4\n",
       "ERCC3           4\n",
       "ESRRB           4\n",
       "KERA            4\n",
       "KIAA0930        4\n",
       "KIF21B          4\n",
       "EXOSC9          4\n",
       "MAN2C1          4\n",
       "GNA11           4\n",
       "ARV1            4\n",
       "GAN             4\n",
       "GBF1            4\n",
       "GCNT2           4\n",
       "B9D1            4\n",
       "IKBKG           4\n",
       "DMRT1           4\n",
       "DNAAF4          4\n",
       "ETFA            4\n",
       "ETFB            4\n",
       "B9D2            4\n",
       "BBS5            4\n",
       "BCKDK           4\n",
       "PYCR2           4\n",
       "QRICH1          4\n",
       "LIPH            4\n",
       "LPIN1           4\n",
       "EPB42           4\n",
       "ERBB3           4\n",
       "HTRA2           4\n",
       "NUS1            4\n",
       "FOXL2           4\n",
       "TDRD1           4\n",
       "TDRD12          4\n",
       "NCDN            4\n",
       "ITPR3           4\n",
       "NDUFA10         4\n",
       "EFEMP1          4\n",
       "PBX1            4\n",
       "C1QB            4\n",
       "PNLDC1          4\n",
       "PPP3CA          4\n",
       "PRKAR1B         4\n",
       "PRKG1           4\n",
       "PRPF3           4\n",
       "GFI1B           4\n",
       "AMACR           4\n",
       "H3-3A           4\n",
       "ANKH            4\n",
       "AP4M1           4\n",
       "HMGCS1          4\n",
       "FGF8            4\n",
       "LCA5            4\n",
       "PSAT1           4\n",
       "FKTN            4\n",
       "KLHL3           4\n",
       "KLK11           4\n",
       "CIB2            4\n",
       "SMC3            4\n",
       "ORC4            4\n",
       "DAAM2           4\n",
       "DAB1            4\n",
       "SHQ1            4\n",
       "DCX             4\n",
       "DDOST           4\n",
       "SLC10A1         4\n",
       "SLC18A2         4\n",
       "SLC25A1         4\n",
       "FOLR1           4\n",
       "FOXI1           4\n",
       "ISCU            4\n",
       "UBAP1           4\n",
       "UBE3A           4\n",
       "UBIAD1          4\n",
       "UGDH            4\n",
       "ACOX1           4\n",
       "ABHD5           4\n",
       "ZMYM3           3\n",
       "YY1             3\n",
       "GNAT1           3\n",
       "GNB5            3\n",
       "ACTL6B          3\n",
       "MPLKIP          3\n",
       "MPO             3\n",
       "ALDOA           3\n",
       "EGR2            3\n",
       "EIF3F           3\n",
       "AGPS            3\n",
       "AIP             3\n",
       "AK1             3\n",
       "AGO1            3\n",
       "GPX4            3\n",
       "TSHB            3\n",
       "TTPA            3\n",
       "LGI3            3\n",
       "LHFPL5          3\n",
       "UPB1            3\n",
       "TWIST1          3\n",
       "GTF2E2          3\n",
       "KCNC2           3\n",
       "ITGAV           3\n",
       "BSCL2           3\n",
       "SGCE            3\n",
       "DDB2            3\n",
       "DDX23           3\n",
       "CYP2R1          3\n",
       "SCO2            3\n",
       "CD8A            3\n",
       "CDCA7           3\n",
       "PCBD1           3\n",
       "CWF19L1         3\n",
       "CYC1            3\n",
       "TFAP2A          3\n",
       "MICU1           3\n",
       "MKS1            3\n",
       "NSF             3\n",
       "OGDHL           3\n",
       "NRL             3\n",
       "SYT1            3\n",
       "TBX2            3\n",
       "HYAL2           3\n",
       "GJA1            3\n",
       "HPRT1           3\n",
       "TIA1            3\n",
       "P3H1            3\n",
       "PAK1            3\n",
       "PAX9            3\n",
       "KIAA0825        3\n",
       "DDX6            3\n",
       "OXGR1           3\n",
       "CD36            3\n",
       "FXN             3\n",
       "FXYD2           3\n",
       "KCNJ13          3\n",
       "KCNN4           3\n",
       "CAV3            3\n",
       "SAR1B           3\n",
       "SBDS            3\n",
       "SCO1            3\n",
       "CFAP418         3\n",
       "CFHR1           3\n",
       "RORB            3\n",
       "RNASEH2C        3\n",
       "RNF170          3\n",
       "MYL4            3\n",
       "TMEM230         3\n",
       "TMEM63B         3\n",
       "TMIE            3\n",
       "TMEM138         3\n",
       "TMEM218         3\n",
       "TFAP2B          3\n",
       "THPO            3\n",
       "DUOXA2          3\n",
       "RBCK1           3\n",
       "PIK3R1          3\n",
       "PLCZ1           3\n",
       "COX15           3\n",
       "RAX2            3\n",
       "PIGK            3\n",
       "CLP1            3\n",
       "LIPT1           3\n",
       "PDXK            3\n",
       "MCIDAS          3\n",
       "EXOSC5          3\n",
       "BHLHA9          3\n",
       "NKX6-2          3\n",
       "SLC35A3         3\n",
       "IKZF1           3\n",
       "GDF1            3\n",
       "KISS1R          3\n",
       "PTF1A           3\n",
       "LEP             3\n",
       "PRPF8           3\n",
       "ANTXR2          3\n",
       "GFER            3\n",
       "AMPD2           3\n",
       "ANG             3\n",
       "HAAO            3\n",
       "CNNM4           3\n",
       "EYA4            3\n",
       "FAR1            3\n",
       "FARSA           3\n",
       "FASLG           3\n",
       "PROK2           3\n",
       "PRKRA           3\n",
       "POU4F3          3\n",
       "KYNU            3\n",
       "CLDN14          3\n",
       "SLC4A3          3\n",
       "SLC9A1          3\n",
       "DENND4A         3\n",
       "DHCR24          3\n",
       "NEUROD1         3\n",
       "NEUROG3         3\n",
       "NDUFS7          3\n",
       "NDUFS8          3\n",
       "NECTIN4         3\n",
       "ATP6V1E1        3\n",
       "GRK1            3\n",
       "GRXCR1          3\n",
       "ZMPSTE24        3\n",
       "TUBG1           2\n",
       "TUBA4A          2\n",
       "UFC1            2\n",
       "TUSC3           2\n",
       "GP1BB           2\n",
       "GOSR2           2\n",
       "VRK1            2\n",
       "GNRH1           2\n",
       "TREX1           2\n",
       "MSH4            2\n",
       "MRPS14          2\n",
       "LIM2            2\n",
       "NDUFS3          2\n",
       "DHDDS           2\n",
       "NIPA1           2\n",
       "NDUFS1          2\n",
       "SPR             2\n",
       "AVIL            2\n",
       "GRWD1           2\n",
       "WDR47           2\n",
       "SFTPC           2\n",
       "C1QC            2\n",
       "DCDC2           2\n",
       "NUDT2           2\n",
       "FTL             2\n",
       "OPA3            2\n",
       "OGT             2\n",
       "ARF1            2\n",
       "DHPS            2\n",
       "NDUFV2          2\n",
       "SNF8            2\n",
       "SNX10           2\n",
       "IL36RN          2\n",
       "IHH             2\n",
       "BPGM            2\n",
       "DHX16           2\n",
       "SLC25A4         2\n",
       "SLC25A22        2\n",
       "SMPX            2\n",
       "SMN1            2\n",
       "STT3A           2\n",
       "SSBP1           2\n",
       "GIPC3           2\n",
       "HSF2BP          2\n",
       "MCM7            2\n",
       "BOLA3           2\n",
       "MAP2K1          2\n",
       "MAPRE2          2\n",
       "NPTX1           2\n",
       "NPC2            2\n",
       "NLGN3           2\n",
       "SLC35C1         2\n",
       "FOXP4           2\n",
       "TBL1XR1         2\n",
       "ARL3            2\n",
       "KDSR            2\n",
       "FBXW7           2\n",
       "SF3B1           2\n",
       "MYH10           2\n",
       "TMEM216         2\n",
       "DNAJC30         2\n",
       "RBP4            2\n",
       "CHRNB1          2\n",
       "CHKA            2\n",
       "MYMK            2\n",
       "THAP1           2\n",
       "SCFD2           2\n",
       "CDC42           2\n",
       "EEFSEC          2\n",
       "EED             2\n",
       "CSRP3           2\n",
       "PDE4DIP         2\n",
       "SCUBE3          2\n",
       "TBX22           2\n",
       "NDUFA13         2\n",
       "MIP             2\n",
       "ISCA2           2\n",
       "FOXE1           2\n",
       "INS             2\n",
       "IPO8            2\n",
       "BTG4            2\n",
       "SIX1            2\n",
       "SHMT2           2\n",
       "SGCD            2\n",
       "AGO2            2\n",
       "ACP4            2\n",
       "F10             2\n",
       "POLA2           2\n",
       "TPM3            2\n",
       "TPM2            2\n",
       "WNT1            2\n",
       "WLS             2\n",
       "MRAS            2\n",
       "USB1            2\n",
       "ELF4            2\n",
       "MPC1            2\n",
       "VPS51           2\n",
       "ALAD            2\n",
       "AGR2            2\n",
       "PTRHD1          2\n",
       "NAA10           2\n",
       "NAXE            2\n",
       "HIKESHI         2\n",
       "HNMT            2\n",
       "PSKH1           2\n",
       "PSMC3           2\n",
       "LDHD            2\n",
       "PRPS1           2\n",
       "PITX1           2\n",
       "PI4KA           2\n",
       "CPA6            2\n",
       "COX20           2\n",
       "CRYBB2          2\n",
       "TLCD3B          2\n",
       "CASP8           2\n",
       "XPNPEP3         2\n",
       "FGF23           2\n",
       "FHL1            2\n",
       "PHACTR1         2\n",
       "PEX2            2\n",
       "CLTC            2\n",
       "TASP1           1\n",
       "MED23           1\n",
       "NT5C3A          1\n",
       "MAPK1           1\n",
       "NDUFS2          1\n",
       "SRY             1\n",
       "SLURP1          1\n",
       "SMAD2           1\n",
       "NMNAT2          1\n",
       "ST3GAL3         1\n",
       "SVBP            1\n",
       "NDUFS4          1\n",
       "GRID1           1\n",
       "NDUFS6          1\n",
       "NDUFB3          1\n",
       "GREM2           1\n",
       "UGP2            1\n",
       "U2AF2           1\n",
       "TRIM24          1\n",
       "GSTZ1           1\n",
       "GTSF1           1\n",
       "ACTC1           1\n",
       "UCHL1           1\n",
       "U2AF1           1\n",
       "IER3IP1         1\n",
       "SMARCB1         1\n",
       "SOCS1           1\n",
       "C1QTNF5         1\n",
       "NUBPL           1\n",
       "RNF220          1\n",
       "RBM5            1\n",
       "TMEM147         1\n",
       "TMEM107         1\n",
       "MXI1            1\n",
       "MYL3            1\n",
       "TMCO1           1\n",
       "RPSA            1\n",
       "SFTPA2          1\n",
       "SEPHS1          1\n",
       "FBXW11          1\n",
       "CD27            1\n",
       "SAG             1\n",
       "THG1L           1\n",
       "CALM2           1\n",
       "CAMK2D          1\n",
       "RRM1            1\n",
       "BVES            1\n",
       "SGO1            1\n",
       "NCF1            1\n",
       "KCNA3           1\n",
       "ECE1            1\n",
       "TERB2           1\n",
       "CSNK2B          1\n",
       "EDARADD         1\n",
       "CD59            1\n",
       "CDK4            1\n",
       "CDK19           1\n",
       "CDK9            1\n",
       "CDK8            1\n",
       "METTL5          1\n",
       "CHN1            1\n",
       "DPH5            1\n",
       "TNFRSF13C       1\n",
       "NAA20           1\n",
       "DPM3            1\n",
       "HBD             1\n",
       "H3-3B           1\n",
       "CLPP            1\n",
       "LMAN2L          1\n",
       "LINGO1          1\n",
       "PEX26           1\n",
       "PEX19           1\n",
       "RAC2            1\n",
       "RAB34           1\n",
       "COX6B1          1\n",
       "CIAO1           1\n",
       "AP2S1           1\n",
       "TNNC1           1\n",
       "TPI1            1\n",
       "CLDN9           1\n",
       "POLR1D          1\n",
       "PRICKLE2        1\n",
       "PRKCG           1\n",
       "KRT81           1\n",
       "KLHL7           1\n",
       "PLN             1\n",
       "FBLN5           1\n",
       "MON1A           1\n",
       "VKORC1          1\n",
       "MPZL2           1\n",
       "WNT7A           1\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds['GeneSymbol'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>#AlleleID</th>\n",
       "      <th>Type</th>\n",
       "      <th>Name</th>\n",
       "      <th>GeneID</th>\n",
       "      <th>GeneSymbol</th>\n",
       "      <th>HGNC_ID</th>\n",
       "      <th>ClinicalSignificance</th>\n",
       "      <th>ClinSigSimple</th>\n",
       "      <th>RS# (dbSNP)</th>\n",
       "      <th>nsv/esv (dbVar)</th>\n",
       "      <th>RCVaccession</th>\n",
       "      <th>PhenotypeIDS</th>\n",
       "      <th>PhenotypeList</th>\n",
       "      <th>Origin</th>\n",
       "      <th>OriginSimple</th>\n",
       "      <th>Assembly</th>\n",
       "      <th>ChromosomeAccession</th>\n",
       "      <th>Chromosome</th>\n",
       "      <th>Start</th>\n",
       "      <th>Stop</th>\n",
       "      <th>Cytogenetic</th>\n",
       "      <th>ReviewStatus</th>\n",
       "      <th>NumberSubmitters</th>\n",
       "      <th>OtherIDs</th>\n",
       "      <th>SubmitterCategories</th>\n",
       "      <th>VariationID</th>\n",
       "      <th>PositionVCF</th>\n",
       "      <th>ReferenceAlleleVCF</th>\n",
       "      <th>AlternateAlleleVCF</th>\n",
       "      <th>Variant (3-letter)</th>\n",
       "      <th>Variant</th>\n",
       "      <th>LastEvaluated (Year)</th>\n",
       "      <th>Uploaded_variation</th>\n",
       "      <th>Location</th>\n",
       "      <th>Allele</th>\n",
       "      <th>Gene</th>\n",
       "      <th>Feature</th>\n",
       "      <th>Feature_type</th>\n",
       "      <th>Consequence</th>\n",
       "      <th>cDNA_position</th>\n",
       "      <th>CDS_position</th>\n",
       "      <th>Protein_position</th>\n",
       "      <th>Amino_acids</th>\n",
       "      <th>Codons</th>\n",
       "      <th>Existing_variation</th>\n",
       "      <th>Extra</th>\n",
       "      <th>PositionVCF_dashed</th>\n",
       "      <th>gnomADe_AF</th>\n",
       "      <th>gnomADg_AF</th>\n",
       "      <th>gnomAD_AF</th>\n",
       "      <th>HGVSp</th>\n",
       "      <th>BinaryClinicalSignificance</th>\n",
       "      <th>SIFT_label</th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>PolyPhen_label</th>\n",
       "      <th>PolyPhen_score</th>\n",
       "      <th>BayesDel_label</th>\n",
       "      <th>BayesDel_score</th>\n",
       "      <th>CADD_PHRED_score</th>\n",
       "      <th>CADD_RAW_score</th>\n",
       "      <th>ClinPred_score</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>EVE_label</th>\n",
       "      <th>EVE_score</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>AM_label</th>\n",
       "      <th>AM_score</th>\n",
       "      <th>Uniprot_acc</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>6567</th>\n",
       "      <td>46226</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_007294.4(BRCA1):c.5434C&gt;G (p.Pro1812Ala)</td>\n",
       "      <td>672</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HGNC:1100</td>\n",
       "      <td>Pathogenic/Likely pathogenic</td>\n",
       "      <td>1</td>\n",
       "      <td>1800751</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000031251|RCV000496797|RCV000484398|RCV001390965|RCV000574861</td>\n",
       "      <td>MONDO:MONDO:0011450,MedGen:C2676676,OMIM:604370,Orphanet:145|MedGen:CN169374|MedGen:C3661900|MONDO:MONDO:0003582,MeSH:D061325,MedGen:C0677776,Orphanet:145|MONDO:MONDO:0015356,MeSH:D009386,MedGen:C0027672,Orphanet:140162</td>\n",
       "      <td>Breast-ovarian cancer, familial, susceptibility to, 1|not specified|not provided|Hereditary breast ovarian cancer syndrome|Hereditary cancer-predisposing syndrome</td>\n",
       "      <td>germline;inherited;not applicable</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>43047676</td>\n",
       "      <td>43047676</td>\n",
       "      <td>17q21.31</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>15</td>\n",
       "      <td>ClinGen:CA003596</td>\n",
       "      <td>2</td>\n",
       "      <td>37670</td>\n",
       "      <td>43047676</td>\n",
       "      <td>G</td>\n",
       "      <td>C</td>\n",
       "      <td>Pro1812Ala</td>\n",
       "      <td>P1812A</td>\n",
       "      <td>2022</td>\n",
       "      <td>1800751.0</td>\n",
       "      <td>17:43047676</td>\n",
       "      <td>C</td>\n",
       "      <td>ENSG00000012048</td>\n",
       "      <td>ENST00000357654.9</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5547</td>\n",
       "      <td>5434</td>\n",
       "      <td>1812</td>\n",
       "      <td>P/A</td>\n",
       "      <td>Cca/Gca</td>\n",
       "      <td>rs1800751,CM032862</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=BRCA1;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:1100;CCDS=CCDS11453.1;ENSP=ENSP00000350283;SWISSPROT=P38398.275;UNIPARC=UPI0000126AC8;UNIPROT_ISOFORM=P38398-1;HGVSc=ENST00000357654.9:c.5434C&gt;G;HGVSp=ENSP00000350283.3:p.Pro1812Ala;gnomADe_AF=6.84e-07;gnomADe_AFR_AF=0;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=8.993e-07;gnomADe_REMAINING_AF=0;gnomADe_SAS_AF=0;MAX_AF=8.993e-07;MAX_AF_POPS=gnomADe_NFE;CLIN_SIG=uncertain_significance,pathogenic,pathogenic/likely_pathogenic;SOMATIC=0,1;PHENO=1,1</td>\n",
       "      <td>43047676</td>\n",
       "      <td>6.840000e-07</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6.840000e-07</td>\n",
       "      <td>ENSP00000350283.3:p.Pro1812Ala</td>\n",
       "      <td>P</td>\n",
       "      <td>deleterious</td>\n",
       "      <td>0.01</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.031</td>\n",
       "      <td>D</td>\n",
       "      <td>0.097989</td>\n",
       "      <td>21.90</td>\n",
       "      <td>3.234223</td>\n",
       "      <td>0.986117</td>\n",
       "      <td>0.714</td>\n",
       "      <td>0.71542</td>\n",
       "      <td>Uncertain</td>\n",
       "      <td>0.428204</td>\n",
       "      <td>0.639</td>\n",
       "      <td>0.516515</td>\n",
       "      <td>likely_benign</td>\n",
       "      <td>0.171</td>\n",
       "      <td>P38398</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6568</th>\n",
       "      <td>70304</td>\n",
       "      <td>single nucleotide variant</td>\n",
       "      <td>NM_007294.4(BRCA1):c.5585A&gt;T (p.His1862Leu)</td>\n",
       "      <td>672</td>\n",
       "      <td>BRCA1</td>\n",
       "      <td>HGNC:1100</td>\n",
       "      <td>Benign/Likely benign</td>\n",
       "      <td>0</td>\n",
       "      <td>80357183</td>\n",
       "      <td>-</td>\n",
       "      <td>RCV000049059|RCV000112707|RCV000774923|RCV001356964|RCV003237433|RCV003607227</td>\n",
       "      <td>MONDO:MONDO:0003582,MeSH:D061325,MedGen:C0677776,Orphanet:145|MONDO:MONDO:0011450,MedGen:C2676676,OMIM:604370,Orphanet:145|MONDO:MONDO:0015356,MeSH:D009386,MedGen:C0027672,Orphanet:140162|MONDO:MONDO:0007254,MedGen:C0006142|MedGen:C3661900|MONDO:MONDO:0016419,MedGen:C0346153,OMIM:114480,Orphanet:227535</td>\n",
       "      <td>Hereditary breast ovarian cancer syndrome|Breast-ovarian cancer, familial, susceptibility to, 1|Hereditary cancer-predisposing syndrome|Malignant tumor of breast|not provided|Familial cancer of breast</td>\n",
       "      <td>germline;unknown</td>\n",
       "      <td>germline</td>\n",
       "      <td>GRCh38</td>\n",
       "      <td>NC_000017.11</td>\n",
       "      <td>17</td>\n",
       "      <td>43045685</td>\n",
       "      <td>43045685</td>\n",
       "      <td>17q21.31</td>\n",
       "      <td>criteria provided, multiple submitters, no conflicts</td>\n",
       "      <td>8</td>\n",
       "      <td>ClinGen:CA003734,UniProtKB:P38398#VAR_070519</td>\n",
       "      <td>2</td>\n",
       "      <td>55637</td>\n",
       "      <td>43045685</td>\n",
       "      <td>T</td>\n",
       "      <td>A</td>\n",
       "      <td>His1862Leu</td>\n",
       "      <td>H1862L</td>\n",
       "      <td>2024</td>\n",
       "      <td>80357183.0</td>\n",
       "      <td>17:43045685</td>\n",
       "      <td>A</td>\n",
       "      <td>ENSG00000012048</td>\n",
       "      <td>ENST00000357654.9</td>\n",
       "      <td>Transcript</td>\n",
       "      <td>missense_variant</td>\n",
       "      <td>5698</td>\n",
       "      <td>5585</td>\n",
       "      <td>1862</td>\n",
       "      <td>H/L</td>\n",
       "      <td>cAc/cTc</td>\n",
       "      <td>rs80357183</td>\n",
       "      <td>IMPACT=MODERATE;STRAND=-1;SYMBOL=BRCA1;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:1100;CCDS=CCDS11453.1;ENSP=ENSP00000350283;SWISSPROT=P38398.275;UNIPARC=UPI0000126AC8;UNIPROT_ISOFORM=P38398-1;HGVSc=ENST00000357654.9:c.5585A&gt;T;HGVSp=ENSP00000350283.3:p.His1862Leu;gnomADe_AF=4.79e-06;gnomADe_AFR_AF=0;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.297e-06;gnomADe_REMAINING_AF=0;gnomADe_SAS_AF=0;gnomADg_AF=6.571e-06;gnomADg_AFR_AF=0;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.47e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=1.47e-05;MAX_AF_POPS=gnomADg_NFE;CLIN_SIG=uncertain_significance,benign,likely_benign;PHENO=1</td>\n",
       "      <td>43045685</td>\n",
       "      <td>4.790000e-06</td>\n",
       "      <td>0.000007</td>\n",
       "      <td>4.790000e-06</td>\n",
       "      <td>ENSP00000350283.3:p.His1862Leu</td>\n",
       "      <td>B</td>\n",
       "      <td>deleterious_low_confidence</td>\n",
       "      <td>0.01</td>\n",
       "      <td>benign</td>\n",
       "      <td>0.013</td>\n",
       "      <td>D</td>\n",
       "      <td>0.220853</td>\n",
       "      <td>10.58</td>\n",
       "      <td>1.021643</td>\n",
       "      <td>0.253849</td>\n",
       "      <td>0.308</td>\n",
       "      <td>0.34533</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>0.553</td>\n",
       "      <td>0.296682</td>\n",
       "      <td>likely_benign</td>\n",
       "      <td>0.094</td>\n",
       "      <td>P38398</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      #AlleleID                       Type                                         Name  GeneID GeneSymbol    HGNC_ID          ClinicalSignificance  ClinSigSimple  RS# (dbSNP) nsv/esv (dbVar)                                                                   RCVaccession                                                                                                                                                                                                                                                                                                     PhenotypeIDS                                                                                                                                                                                             PhenotypeList                             Origin OriginSimple Assembly ChromosomeAccession Chromosome     Start      Stop Cytogenetic                                          ReviewStatus  NumberSubmitters                                      OtherIDs  SubmitterCategories  VariationID  PositionVCF ReferenceAlleleVCF AlternateAlleleVCF Variant (3-letter) Variant  LastEvaluated (Year)  Uploaded_variation     Location Allele             Gene            Feature Feature_type       Consequence  cDNA_position  CDS_position  Protein_position Amino_acids   Codons  Existing_variation                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                                     Extra  PositionVCF_dashed    gnomADe_AF  gnomADg_AF     gnomAD_AF                           HGVSp BinaryClinicalSignificance                  SIFT_label  SIFT_score PolyPhen_label  PolyPhen_score BayesDel_label  BayesDel_score  CADD_PHRED_score  CADD_RAW_score  ClinPred_score  VEST4_score  VEST4_rankscore  EVE_label  EVE_score  REVEL_score  PrimateAI_score       AM_label  AM_score Uniprot_acc\n",
       "6567      46226  single nucleotide variant  NM_007294.4(BRCA1):c.5434C>G (p.Pro1812Ala)     672      BRCA1  HGNC:1100  Pathogenic/Likely pathogenic              1      1800751               -               RCV000031251|RCV000496797|RCV000484398|RCV001390965|RCV000574861                                                                                      MONDO:MONDO:0011450,MedGen:C2676676,OMIM:604370,Orphanet:145|MedGen:CN169374|MedGen:C3661900|MONDO:MONDO:0003582,MeSH:D061325,MedGen:C0677776,Orphanet:145|MONDO:MONDO:0015356,MeSH:D009386,MedGen:C0027672,Orphanet:140162                                        Breast-ovarian cancer, familial, susceptibility to, 1|not specified|not provided|Hereditary breast ovarian cancer syndrome|Hereditary cancer-predisposing syndrome  germline;inherited;not applicable     germline   GRCh38        NC_000017.11         17  43047676  43047676    17q21.31  criteria provided, multiple submitters, no conflicts                15                              ClinGen:CA003596                    2        37670     43047676                  G                  C         Pro1812Ala  P1812A                  2022           1800751.0  17:43047676      C  ENSG00000012048  ENST00000357654.9   Transcript  missense_variant           5547          5434              1812         P/A  Cca/Gca  rs1800751,CM032862                                                                                                                                                                            IMPACT=MODERATE;STRAND=-1;SYMBOL=BRCA1;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:1100;CCDS=CCDS11453.1;ENSP=ENSP00000350283;SWISSPROT=P38398.275;UNIPARC=UPI0000126AC8;UNIPROT_ISOFORM=P38398-1;HGVSc=ENST00000357654.9:c.5434C>G;HGVSp=ENSP00000350283.3:p.Pro1812Ala;gnomADe_AF=6.84e-07;gnomADe_AFR_AF=0;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=8.993e-07;gnomADe_REMAINING_AF=0;gnomADe_SAS_AF=0;MAX_AF=8.993e-07;MAX_AF_POPS=gnomADe_NFE;CLIN_SIG=uncertain_significance,pathogenic,pathogenic/likely_pathogenic;SOMATIC=0,1;PHENO=1,1            43047676  6.840000e-07         NaN  6.840000e-07  ENSP00000350283.3:p.Pro1812Ala                          P                 deleterious        0.01         benign           0.031              D        0.097989             21.90        3.234223        0.986117        0.714          0.71542  Uncertain   0.428204        0.639         0.516515  likely_benign     0.171      P38398\n",
       "6568      70304  single nucleotide variant  NM_007294.4(BRCA1):c.5585A>T (p.His1862Leu)     672      BRCA1  HGNC:1100          Benign/Likely benign              0     80357183               -  RCV000049059|RCV000112707|RCV000774923|RCV001356964|RCV003237433|RCV003607227  MONDO:MONDO:0003582,MeSH:D061325,MedGen:C0677776,Orphanet:145|MONDO:MONDO:0011450,MedGen:C2676676,OMIM:604370,Orphanet:145|MONDO:MONDO:0015356,MeSH:D009386,MedGen:C0027672,Orphanet:140162|MONDO:MONDO:0007254,MedGen:C0006142|MedGen:C3661900|MONDO:MONDO:0016419,MedGen:C0346153,OMIM:114480,Orphanet:227535  Hereditary breast ovarian cancer syndrome|Breast-ovarian cancer, familial, susceptibility to, 1|Hereditary cancer-predisposing syndrome|Malignant tumor of breast|not provided|Familial cancer of breast                   germline;unknown     germline   GRCh38        NC_000017.11         17  43045685  43045685    17q21.31  criteria provided, multiple submitters, no conflicts                 8  ClinGen:CA003734,UniProtKB:P38398#VAR_070519                    2        55637     43045685                  T                  A         His1862Leu  H1862L                  2024          80357183.0  17:43045685      A  ENSG00000012048  ENST00000357654.9   Transcript  missense_variant           5698          5585              1862         H/L  cAc/cTc          rs80357183  IMPACT=MODERATE;STRAND=-1;SYMBOL=BRCA1;SYMBOL_SOURCE=HGNC;HGNC_ID=HGNC:1100;CCDS=CCDS11453.1;ENSP=ENSP00000350283;SWISSPROT=P38398.275;UNIPARC=UPI0000126AC8;UNIPROT_ISOFORM=P38398-1;HGVSc=ENST00000357654.9:c.5585A>T;HGVSp=ENSP00000350283.3:p.His1862Leu;gnomADe_AF=4.79e-06;gnomADe_AFR_AF=0;gnomADe_AMR_AF=0;gnomADe_ASJ_AF=0;gnomADe_EAS_AF=0;gnomADe_FIN_AF=0;gnomADe_MID_AF=0;gnomADe_NFE_AF=6.297e-06;gnomADe_REMAINING_AF=0;gnomADe_SAS_AF=0;gnomADg_AF=6.571e-06;gnomADg_AFR_AF=0;gnomADg_AMI_AF=0;gnomADg_AMR_AF=0;gnomADg_ASJ_AF=0;gnomADg_EAS_AF=0;gnomADg_FIN_AF=0;gnomADg_MID_AF=0;gnomADg_NFE_AF=1.47e-05;gnomADg_REMAINING_AF=0;gnomADg_SAS_AF=0;MAX_AF=1.47e-05;MAX_AF_POPS=gnomADg_NFE;CLIN_SIG=uncertain_significance,benign,likely_benign;PHENO=1            43045685  4.790000e-06    0.000007  4.790000e-06  ENSP00000350283.3:p.His1862Leu                          B  deleterious_low_confidence        0.01         benign           0.013              D        0.220853             10.58        1.021643        0.253849        0.308          0.34533        NaN        NaN        0.553         0.296682  likely_benign     0.094      P38398"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds[clinvar_with_preds['GeneSymbol'] == \"BRCA1\"].head(2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SIFT_score</th>\n",
       "      <th>PolyPhen_score</th>\n",
       "      <th>CADD_RAW_score</th>\n",
       "      <th>CADD_PHRED_score</th>\n",
       "      <th>ClinPred_score</th>\n",
       "      <th>VEST4_score</th>\n",
       "      <th>VEST4_rankscore</th>\n",
       "      <th>EVE_score</th>\n",
       "      <th>REVEL_score</th>\n",
       "      <th>PrimateAI_score</th>\n",
       "      <th>AM_score</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>count</th>\n",
       "      <td>47562.000000</td>\n",
       "      <td>45581.000000</td>\n",
       "      <td>49187.000000</td>\n",
       "      <td>49187.000000</td>\n",
       "      <td>48977.000000</td>\n",
       "      <td>49187.000000</td>\n",
       "      <td>47239.000000</td>\n",
       "      <td>31381.000000</td>\n",
       "      <td>43752.000000</td>\n",
       "      <td>48568.000000</td>\n",
       "      <td>44992.000000</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>mean</th>\n",
       "      <td>0.146499</td>\n",
       "      <td>0.450637</td>\n",
       "      <td>3.281590</td>\n",
       "      <td>20.868921</td>\n",
       "      <td>0.443935</td>\n",
       "      <td>0.498975</td>\n",
       "      <td>0.529627</td>\n",
       "      <td>0.437151</td>\n",
       "      <td>0.426532</td>\n",
       "      <td>0.558980</td>\n",
       "      <td>0.360627</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>std</th>\n",
       "      <td>0.265088</td>\n",
       "      <td>0.435525</td>\n",
       "      <td>1.814761</td>\n",
       "      <td>8.646250</td>\n",
       "      <td>0.414668</td>\n",
       "      <td>0.330537</td>\n",
       "      <td>0.321720</td>\n",
       "      <td>0.294033</td>\n",
       "      <td>0.318185</td>\n",
       "      <td>0.198613</td>\n",
       "      <td>0.340495</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>min</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>-4.563263</td>\n",
       "      <td>0.001000</td>\n",
       "      <td>0.000017</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.000040</td>\n",
       "      <td>0.012372</td>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.160219</td>\n",
       "      <td>0.026500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25%</th>\n",
       "      <td>0.000000</td>\n",
       "      <td>0.006000</td>\n",
       "      <td>2.037965</td>\n",
       "      <td>16.930000</td>\n",
       "      <td>0.042429</td>\n",
       "      <td>0.190000</td>\n",
       "      <td>0.237220</td>\n",
       "      <td>0.158716</td>\n",
       "      <td>0.144000</td>\n",
       "      <td>0.389804</td>\n",
       "      <td>0.086600</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>50%</th>\n",
       "      <td>0.020000</td>\n",
       "      <td>0.301000</td>\n",
       "      <td>3.625519</td>\n",
       "      <td>23.100000</td>\n",
       "      <td>0.256858</td>\n",
       "      <td>0.468000</td>\n",
       "      <td>0.530340</td>\n",
       "      <td>0.379827</td>\n",
       "      <td>0.335000</td>\n",
       "      <td>0.561070</td>\n",
       "      <td>0.170200</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>75%</th>\n",
       "      <td>0.150000</td>\n",
       "      <td>0.972000</td>\n",
       "      <td>4.727878</td>\n",
       "      <td>26.500000</td>\n",
       "      <td>0.958861</td>\n",
       "      <td>0.832000</td>\n",
       "      <td>0.839950</td>\n",
       "      <td>0.703409</td>\n",
       "      <td>0.731000</td>\n",
       "      <td>0.722608</td>\n",
       "      <td>0.659425</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>max</th>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>12.062742</td>\n",
       "      <td>56.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.999990</td>\n",
       "      <td>0.999872</td>\n",
       "      <td>1.000000</td>\n",
       "      <td>0.978576</td>\n",
       "      <td>1.000000</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "         SIFT_score  PolyPhen_score  CADD_RAW_score  CADD_PHRED_score  ClinPred_score   VEST4_score  VEST4_rankscore     EVE_score   REVEL_score  PrimateAI_score      AM_score\n",
       "count  47562.000000    45581.000000    49187.000000      49187.000000    48977.000000  49187.000000     47239.000000  31381.000000  43752.000000     48568.000000  44992.000000\n",
       "mean       0.146499        0.450637        3.281590         20.868921        0.443935      0.498975         0.529627      0.437151      0.426532         0.558980      0.360627\n",
       "std        0.265088        0.435525        1.814761          8.646250        0.414668      0.330537         0.321720      0.294033      0.318185         0.198613      0.340495\n",
       "min        0.000000        0.000000       -4.563263          0.001000        0.000017      0.000000         0.000040      0.012372      0.000000         0.160219      0.026500\n",
       "25%        0.000000        0.006000        2.037965         16.930000        0.042429      0.190000         0.237220      0.158716      0.144000         0.389804      0.086600\n",
       "50%        0.020000        0.301000        3.625519         23.100000        0.256858      0.468000         0.530340      0.379827      0.335000         0.561070      0.170200\n",
       "75%        0.150000        0.972000        4.727878         26.500000        0.958861      0.832000         0.839950      0.703409      0.731000         0.722608      0.659425\n",
       "max        1.000000        1.000000       12.062742         56.000000        1.000000      1.000000         0.999990      0.999872      1.000000         0.978576      1.000000"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds[['SIFT_score', 'PolyPhen_score', 'CADD_RAW_score', 'CADD_PHRED_score', 'ClinPred_score', 'VEST4_score', 'VEST4_rankscore', 'EVE_score', 'REVEL_score', 'PrimateAI_score', 'AM_score']].describe()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "SIFT_score           1625\n",
       "PolyPhen_score       3606\n",
       "CADD_RAW_score          0\n",
       "CADD_PHRED_score        0\n",
       "ClinPred_score        210\n",
       "VEST4_score             0\n",
       "VEST4_rankscore      1948\n",
       "EVE_score           17806\n",
       "REVEL_score          5435\n",
       "PrimateAI_score       619\n",
       "AM_score             4195\n",
       "dtype: int64"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds[['SIFT_score', 'PolyPhen_score', 'CADD_RAW_score', 'CADD_PHRED_score', 'ClinPred_score', 'VEST4_score', 'VEST4_rankscore', 'EVE_score', 'REVEL_score', 'PrimateAI_score', 'AM_score']].isna().sum()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We use the following function to evaluate the coverage of each predictor in the dataset. Not all variants can have a prediction due to varying reasons, such as the specific type of variant, the prediction model, or the absence of relevant data. This check helps identify if any predictors have a high percentage of missing values (NaNs), indicating they may not be informative for a large portion of the dataset."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Predictor</th>\n",
       "      <th>Coverage</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>VEST4</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CADD (RAW)</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CADD (PHRED)</td>\n",
       "      <td>100.00</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>BayesDel</td>\n",
       "      <td>99.87</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>ClinPred</td>\n",
       "      <td>99.57</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>PrimateAI</td>\n",
       "      <td>98.74</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SIFT</td>\n",
       "      <td>96.70</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>PolyPhen</td>\n",
       "      <td>92.67</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AlphaMissense</td>\n",
       "      <td>91.47</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>REVEL</td>\n",
       "      <td>88.95</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>EVE</td>\n",
       "      <td>63.80</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        Predictor  Coverage\n",
       "0           VEST4    100.00\n",
       "1      CADD (RAW)    100.00\n",
       "2    CADD (PHRED)    100.00\n",
       "3        BayesDel     99.87\n",
       "4        ClinPred     99.57\n",
       "5       PrimateAI     98.74\n",
       "6            SIFT     96.70\n",
       "7        PolyPhen     92.67\n",
       "8   AlphaMissense     91.47\n",
       "9           REVEL     88.95\n",
       "10            EVE     63.80"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "check_coverage(clinvar_with_preds)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BinaryClinicalSignificance\n",
       "B    34405\n",
       "P    14782\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds.BinaryClinicalSignificance.value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "For easier merging with Humsavar and for a better identification of variants, we retrieve a column for UniprotIDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['IMPACT=MODERATE',\n",
       " 'STRAND=-1',\n",
       " 'SYMBOL=AAAS',\n",
       " 'SYMBOL_SOURCE=HGNC',\n",
       " 'HGNC_ID=HGNC:13666',\n",
       " 'CCDS=CCDS8856.1',\n",
       " 'ENSP=ENSP00000209873',\n",
       " 'SWISSPROT=Q9NRG9.206',\n",
       " 'UNIPARC=UPI0000039E40',\n",
       " 'UNIPROT_ISOFORM=Q9NRG9-1',\n",
       " 'HGVSc=ENST00000209873.9:c.1597G>A',\n",
       " 'HGVSp=ENSP00000209873.4:p.Gly533Arg',\n",
       " 'AF=0.0124',\n",
       " 'AFR_AF=0.0439',\n",
       " 'AMR_AF=0.0043',\n",
       " 'EAS_AF=0',\n",
       " 'EUR_AF=0.001',\n",
       " 'SAS_AF=0',\n",
       " 'gnomADe_AF=0.0007928',\n",
       " 'gnomADe_AFR_AF=0.02712',\n",
       " 'gnomADe_AMR_AF=0.001879',\n",
       " 'gnomADe_ASJ_AF=3.827e-05',\n",
       " 'gnomADe_EAS_AF=0',\n",
       " 'gnomADe_FIN_AF=0',\n",
       " 'gnomADe_MID_AF=0.001214',\n",
       " 'gnomADe_NFE_AF=2.428e-05',\n",
       " 'gnomADe_REMAINING_AF=0.002053',\n",
       " 'gnomADe_SAS_AF=9.275e-05',\n",
       " 'gnomADg_AF=0.007867',\n",
       " 'gnomADg_AFR_AF=0.02713',\n",
       " 'gnomADg_AMI_AF=0',\n",
       " 'gnomADg_AMR_AF=0.003529',\n",
       " 'gnomADg_ASJ_AF=0',\n",
       " 'gnomADg_EAS_AF=0',\n",
       " 'gnomADg_FIN_AF=0',\n",
       " 'gnomADg_MID_AF=0',\n",
       " 'gnomADg_NFE_AF=0.0001029',\n",
       " 'gnomADg_REMAINING_AF=0.004748',\n",
       " 'gnomADg_SAS_AF=0',\n",
       " 'MAX_AF=0.0439',\n",
       " 'MAX_AF_POPS=AFR',\n",
       " 'CLIN_SIG=benign,likely_benign',\n",
       " 'PHENO=1']"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds[\"Extra\"].iloc[0].split(\";\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "0                                                ClinGen:CA6598796\n",
       "1    ClinGen:CA117228,UniProtKB:Q9NRG9#VAR_012806,OMIM:605378.0007\n",
       "2                                              ClinGen:CA237335767\n",
       "3                                              ClinGen:CA385043710\n",
       "4                               ClinGen:CA6599181,OMIM:605378.0011\n",
       "Name: OtherIDs, dtype: object"
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "clinvar_with_preds[\"OtherIDs\"].head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "At the end, we extract UniProt IDs from the variant annotations. These IDs can appear in different formats or columns, so we check several places.\n",
    "\n",
    "This finally gives us a way to identify which protein each variant maps to."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "def extract_uniprot_id(extra, other_ids):\n",
    "    # check for UNIPROT_ISOFORM in the 'Extra' column\n",
    "    if pd.notna(extra):\n",
    "        match_isoform = re.search(r'UNIPROT_ISOFORM=([A-Z0-9]+)-1\\b', extra)\n",
    "        if match_isoform:\n",
    "            return match_isoform.group(1)\n",
    "    \n",
    "    # check for UniProtKB in the 'OtherIDs' column (and get part before #)\n",
    "    if pd.notna(other_ids):\n",
    "        match_uniprotkb = re.search(r'UniProtKB:([A-Z0-9]+)(?=#)', other_ids)\n",
    "        if match_uniprotkb:\n",
    "            return match_uniprotkb.group(1)\n",
    "    \n",
    "    # check for SWISSPROT in the 'Extra' column (and get part before dot)\n",
    "    if pd.notna(extra):\n",
    "        match_swiss = re.search(r'SWISSPROT=([A-Z0-9]+)', extra)\n",
    "        if match_swiss:\n",
    "            return match_swiss.group(1).split('.')[0]\n",
    "    \n",
    "    # check for TREMBL in the 'Extra' column (and get part before dot)\n",
    "    if pd.notna(extra):\n",
    "        match_trembl = re.search(r'TREMBL=([A-Z0-9]+)', extra)\n",
    "        if match_trembl:\n",
    "            return match_trembl.group(1).split('.')[0]\n",
    "\n",
    "    return None"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "We load a list of reviewed UniProt IDs which we know for sure that are manually curated and more reliable than automatic ones.\n",
    "\n",
    "We use this list to filter out less trustworthy protein mappings, as some variants map to unreviwed IDs."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "with open('uniprotkb_reviewed_true_AND_organism_id_2025_04_09.list') as f:\n",
    "    reviewed_ids = set(line.strip() for line in f if line.strip())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['A0A024R1R8', 'A0A024RBG1', 'A0A075B6H7', 'A0A075B6H8', 'A0A075B6H9', 'A0A075B6I0', 'A0A075B6I1', 'A0A075B6I3', 'A0A075B6I4', 'A0A075B6I6']\n",
      "20417\n"
     ]
    }
   ],
   "source": [
    "print(sorted(reviewed_ids)[0:10])\n",
    "print(len(reviewed_ids))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "clinvar_with_preds['UniprotID'] = clinvar_with_preds.apply(lambda row: extract_uniprot_id(row['Uniprot_acc'], reviewed_ids), axis=1)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Thus, we filter the dataset to keep only variants that map to reviewed UniProt IDs.\n",
    "\n",
    "And after filtering, we check if any gene still maps to more than one UniProt ID."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [],
   "source": [
    "reviewed_df = clinvar_with_preds[clinvar_with_preds['UniprotID'].isin(reviewed_ids)].copy()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "0\n",
      "2105\n",
      "2105\n"
     ]
    }
   ],
   "source": [
    "print(reviewed_df['UniprotID'].isna().sum())\n",
    "print(reviewed_df['GeneSymbol'].nunique())\n",
    "print(reviewed_df['UniprotID'].nunique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 80,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(46050, 70)"
      ]
     },
     "execution_count": 80,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reviewed_df.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 83,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "BinaryClinicalSignificance\n",
       "B    32051\n",
       "P    13999\n",
       "Name: count, dtype: int64"
      ]
     },
     "execution_count": 83,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "reviewed_df.BinaryClinicalSignificance.value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "GeneSymbol\n",
       "ERCC6    True\n",
       "GNAS     True\n",
       "Name: UniprotID, dtype: bool"
      ]
     },
     "execution_count": 79,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "(reviewed_df.groupby('GeneSymbol')['UniprotID']\n",
    " .nunique()\n",
    " .gt(1)\n",
    " .loc[lambda x: x])"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "These genes are linked to more than one reviewed UniProt ID, which could mean alternative isoforms or distinct protein products that were reviewed separately.\n",
    "\n",
    "It's unusual but not necessarily wrong. As it is something not that relevant for the analysis, we kept both."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 81,
   "metadata": {},
   "outputs": [],
   "source": [
    "np.savetxt('../data/clinvar/uniprot_ids.txt', reviewed_df['UniprotID'].unique(), fmt='%s')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 82,
   "metadata": {},
   "outputs": [],
   "source": [
    "reviewed_df.to_csv('../data/clinvar/cleaned_ClinVar_with_preds.csv', index=0)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "The full procedure outlined in this *.ipynb* serves to explain the methodology behind my thesis (for the data preprocessing step), providing a clear and structured approach for better comprehension. \n",
    "\n",
    "In practice, however, this process is executed within a single script, which calls various Python functions. As described in the report, by inputting a gene name or a list of gene names, the final output is a CSV file with the added predictions. \n",
    "\n",
    "*For more detailed demonstrations and explanations of how the program works, please refer to the __GitHub repository__ linked to my thesis project.*"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "jupyter_env",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
